// @univerjs/slides-ui/lib/es/index.js
var hn = Object.defineProperty;
var pn = (i, e, t) => e in i ? hn(i, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : i[e] = t;
var P = (i, e, t) => pn(i, typeof e != "symbol" ? e + "" : e, t);
import { Inject as X, Injector as Ge, RxDisposable as fe, UniverInstanceType as x, debounce as fn, getColorStyle as Cn, IUniverInstanceService as j, CommandType as D, Tools as vn, useDependency as O, LocaleService as Q, ICommandService as U, generateRandomId as xt, PageElementType as Xe, BasicShapes as _n, Disposable as Ue, DisposableCollection as Ye, toDisposable as Et, FOCUSING_COMMON_DRAWINGS as Sn, IContextService as Ce, createInternalEditorID as In, createIdentifier as Rt, EDITOR_ACTIVATED as Oe, FOCUSING_EDITOR_STANDALONE as Ot, FOCUSING_UNIVER_EDITOR_STANDALONE_SINGLE_MODE as jt, DocumentDataModel as Pn, VerticalAlign as je, HorizontalAlign as yt, WrapStrategy as it, FOCUSING_EDITOR_BUT_HIDDEN as bn, DEFAULT_EMPTY_DOCUMENT_VALUE as Tt, Direction as B, IUndoRedoService as xn, DocumentFlavor as En, FOCUSING_UNIVER_EDITOR as Mt, FORMULA_EDITOR_ACTIVATED as Rn, connectInjector as rt, Plugin as On, merge as jn, mergeOverrideWithDependencies as st, IConfigService as yn } from "@univerjs/core";
import { Viewport as ot, ScrollBar as wt, Slide as Tn, Rect as Nt, Scene as Mn, IRenderManagerService as Y, ObjectType as he, pxToNum as wn, DeviceInputEventType as q, convertTextRotation as at, FIX_ONE_PIXEL_BLUR_OFFSET as re, fixLineWidthByScale as ct } from "@univerjs/engine-render";
import { ObjectProvider as Nn, SLIDE_KEY as se } from "@univerjs/slides";
import { ISidebarService as Dt, ICanvasPopupService as Dn, KeyCode as y, ILayoutService as Un, ILocalFileService as Vn, useObservable as Bn, DISABLE_AUTO_FOCUS_KEY as lt, MenuItemType as ve, getMenuHiddenObservable as _e, RibbonStartGroup as Ln, MetaKeys as An, ComponentManager as $n, BuiltInUIPart as dt, IMenuManagerService as Hn, IUIPartsService as Fn, IShortcutService as zn } from "@univerjs/ui";
import ke, { forwardRef as A, useRef as Ut, createElement as $, useState as W, useEffect as oe, useCallback as ut } from "react";
import { Button as Pe, InputNumber as ue, DropdownLegacy as Vt, ColorPicker as Wn, Scrollbar as Kn } from "@univerjs/design";
import { BehaviorSubject as J, Subject as Bt, filter as Gn, takeUntil as mt } from "rxjs";
import { DocSelectionManagerService as Xn, DocSkeletonManagerService as gt, RichTextEditingMutation as Yn } from "@univerjs/docs";
import { IEditorService as Ze, DocSelectionRenderService as ht, VIEWPORT_KEY as pt, DOCS_COMPONENT_MAIN_LAYER_INDEX as kn, MoveSelectionOperation as Zn, MoveCursorOperation as qn, DOCS_VIEW_KEY as Jn, DeleteLeftCommand as Qn } from "@univerjs/docs-ui";
import { DRAWING_IMAGE_ALLOW_IMAGE_LIST as ei, IImageIoService as ti, getImageSize as ni } from "@univerjs/drawing";
var ii = Object.defineProperty, ri = Object.getOwnPropertyDescriptor, si = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? ri(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && ii(e, t, r), r;
}, Ve = (i, e) => (t, n) => e(t, n, i);
let pe = class extends fe {
  constructor(e, t, n, r) {
    super();
    P(this, "_objectProvider", null);
    P(this, "_refreshThumb", fn(() => {
      this.createThumbs();
    }, 300));
    this._renderContext = e, this._injector = t, this._univerInstanceService = n, this._renderManagerService = r, this._objectProvider = this._injector.createInstance(Nn), this._addNewRender();
  }
  _addNewRender() {
    const { unitId: e, engine: t, scene: n } = this._renderContext, r = this._getCurrUnitModel();
    if (!r) return;
    const s = t.onTransformChange$.subscribeEvent(() => {
      this._scrollToCenter(), s == null || s.unsubscribe();
    });
    t.onTransformChange$.subscribeEvent(() => {
      setTimeout(() => {
        this.createThumbs();
      }, 300);
    });
    const o = new ot(se.VIEW, n, {
      left: 0,
      top: 0,
      bottom: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      explicitViewportHeightSet: !1,
      isWheelPreventDefaultX: !0
    });
    n.attachControl(), n.onMouseWheel$.subscribeEvent((c, g) => {
      const l = c;
      if (l.ctrlKey) {
        const u = Math.abs(l.deltaX);
        let h = u < 40 ? 0.2 : u < 80 ? 0.4 : 0.2;
        h *= l.deltaY > 0 ? -1 : 1, n.scaleX < 1 && (h /= 2), n.scaleX + h > 4 ? n.scale(4, 4) : n.scaleX + h < 0.1 ? n.scale(0.1, 0.1) : (l.deltaY > 0, l.preventDefault());
      } else
        o.onMouseWheel(l, g);
    }), n.onFileLoaded$.subscribeEvent(() => {
      this._refreshThumb();
    }), wt.attachTo(o), this._renderManagerService.setCurrent(e);
    const a = this._createSlide(n);
    this._renderContext.mainComponent = a, this._createSlidePages(r, a), this.createThumbs(), t.runRenderLoop(() => {
      n.render();
    });
  }
  _scrollToCenter() {
    var c;
    const e = (c = this._currentRender()) == null ? void 0 : c.scene, t = e == null ? void 0 : e.getViewport(se.VIEW), n = this._getCenterPositionViewPort(e);
    if (!t || !n) return;
    const { left: r, top: s } = n, { x: o, y: a } = t.transViewportScroll2ScrollValue(r, s);
    t.scrollToBarPos({
      x: o,
      y: a
    });
  }
  _currentRender() {
    return this._renderManagerService.getCurrentTypeOfRenderer(x.UNIVER_SLIDE);
  }
  /**
   * @param mainScene
   */
  _createSlide(e) {
    const t = this._univerInstanceService.getCurrentUnitForType(x.UNIVER_SLIDE), { width: n, height: r } = e, s = t.getPageSize(), { width: o = 100, height: a = 100 } = s, c = new Tn(se.COMPONENT, {
      left: (n - o) / 2,
      top: (r - a) / 2,
      width: o,
      height: a,
      zIndex: 10
    });
    return c.enableSelectedClipElement(), e.addObject(c), c;
  }
  _addBackgroundRect(e, t) {
    const r = this._univerInstanceService.getCurrentUnitForType(x.UNIVER_SLIDE).getPageSize(), { width: s = 0, height: o = 0 } = r, a = new Nt("canvas", {
      left: 0,
      top: 0,
      width: s,
      height: o,
      strokeWidth: 1,
      stroke: "rgba(198,198,198,1)",
      fill: Cn(t) || "rgba(255,255,255,1)",
      zIndex: 0,
      evented: !1
    });
    e.addObject(a, 0);
  }
  _getCenterPositionViewPort(e) {
    if (!e) return { left: 0, top: 0 };
    const { width: t, height: n } = e, r = e.getEngine(), s = (r == null ? void 0 : r.width) || 0, o = (r == null ? void 0 : r.height) || 0;
    return {
      left: (t - s) / 2,
      top: (n - o) / 2
    };
  }
  _thumbSceneRender(e, t) {
    const n = this._renderManagerService.getRenderById(e);
    if (n == null)
      return;
    const { engine: r } = n;
    if (r == null)
      return;
    const { width: s, height: o } = t, { width: a = s, height: c = o } = r, g = r.getCanvas().getContext();
    t.renderToThumb(g, e, a / s, c / o);
  }
  /**
   * CreateScene by pages, and activate first one.
   * @param slideDataModel
   * @param slide
   */
  _createSlidePages(e, t) {
    const n = e.getPages(), r = e.getPageOrder();
    if (!(!n || !r) && r.length !== 0) {
      for (let s = 0, o = r.length; s < o; s++) {
        const a = r[s];
        this.createPageScene(a, n[a]), this._createThumb(a);
      }
      t.activeFirstPage();
    }
  }
  _createThumb(e) {
    this._renderManagerService.createRender(e);
  }
  /**
   * SlideDataModel is UnitModel
   */
  _getCurrUnitModel() {
    return this._renderContext.unit;
  }
  activePage(e) {
    let t = e;
    const n = this._getCurrUnitModel();
    let r;
    if (t)
      r = n.getPage(t);
    else {
      const c = n.getPages(), g = n.getPageOrder();
      if (g == null || c == null)
        return;
      r = c[g[0]], t = r.id;
    }
    const s = this._currentRender();
    if (r == null || s == null || s.mainComponent == null)
      return;
    const { id: o } = r, a = s.mainComponent;
    if (n.setActivePage(r), a != null && a.hasPage(o)) {
      a.changePage(o);
      return;
    }
    this.createPageScene(o, r);
  }
  createThumbs() {
    const t = this._getCurrUnitModel().getPageOrder(), n = this._currentRender();
    if (!(!t || !n) && t.length !== 0)
      for (let r = 0, s = t.length; r < s; r++) {
        const o = t[r];
        this._thumbSceneRender(o, n.mainComponent);
      }
  }
  /**
   * Create scene by page and set to _sceneMap.
   * @param pageId
   * @param page
   */
  createPageScene(e, t) {
    const n = this._renderContext;
    if (!n || !this._objectProvider)
      return;
    const { scene: r, mainComponent: s } = n, o = s, { width: a, height: c } = o, g = new Mn(e, o, {
      width: a,
      height: c
    });
    new ot(`PageViewer_${e}`, g, {
      left: 0,
      top: 0,
      bottom: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      explicitViewportHeightSet: !1
    }).closeClip();
    const { pageElements: u, pageBackgroundFill: h } = t, v = this._objectProvider.convertToRenderObjects(u, r);
    if (!v || !o) return;
    this._addBackgroundRect(g, h), g.addObjects(v), g.initTransformer(), v.forEach((C) => {
      g.attachTransformerTo(C);
    });
    const _ = g.getTransformer();
    return _ == null || _.changeEnd$.subscribe(() => {
      this._thumbSceneRender(e, o);
    }), _ == null || _.clearControl$.subscribe(() => {
      this._thumbSceneRender(e, o);
    }), o.addPageScene(g), g;
  }
  /**
   * Get pageScene from Slide.
   * @param pageId
   * @returns {Scene, Engine, UnitModel} scene & engine & unit from renderContext
   */
  getPageRenderUnit(e) {
    const n = this._renderContext.mainComponent.getSubScenes().get(e), { engine: r, unit: s } = this._renderContext;
    return {
      scene: n,
      engine: r,
      unit: s
    };
  }
  createObjectToPage(e, t) {
    const { scene: n } = this.getPageRenderUnit(t);
    if (!n || !this._objectProvider)
      return;
    const r = this._objectProvider.convertToRenderObject(e, n);
    if (r)
      return n.addObject(r), n.attachTransformerTo(r), n.getLayer().makeDirty(), r;
  }
  setObjectActiveByPage(e, t) {
    const { scene: n } = this.getPageRenderUnit(t);
    if (!n) return;
    const r = n.getTransformer();
    r == null || r.activeAnObject(e);
  }
  removeObjectById(e, t) {
    const { scene: n } = this.getPageRenderUnit(t);
    if (!n) return;
    n.removeObject(e);
    const r = n.getTransformer();
    r == null || r.clearControls();
  }
  appendPage() {
    const e = this._getCurrUnitModel(), t = e.getBlankPage(), n = this._currentRender();
    if (t == null || n == null || n.mainComponent == null)
      return;
    const { id: r } = t, s = n.mainComponent, o = this.createPageScene(r, t);
    s && o && s.addPageScene(o), e.appendPage(t), e.setActivePage(t);
  }
};
pe = si([
  Ve(1, X(Ge)),
  Ve(2, j),
  Ve(3, Y)
], pe);
var oi = Object.defineProperty, ai = Object.getOwnPropertyDescriptor, ci = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? ai(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && oi(e, t, r), r;
}, li = (i, e) => (t, n) => e(t, n, i);
let M = class extends fe {
  constructor(i) {
    super(), this._renderManagerService = i;
  }
  _getSlideRenderControllerFromRenderUnit(i) {
    return this._renderManagerService.getRenderById(i).with(pe);
  }
  createThumbs(i) {
    this._getSlideRenderControllerFromRenderUnit(i).createThumbs();
  }
  activePage(i, e) {
    this._getSlideRenderControllerFromRenderUnit(e).activePage(i);
  }
  getRenderUnitByPageId(i, e) {
    return this._getSlideRenderControllerFromRenderUnit(e).getPageRenderUnit(i);
  }
  createObjectToPage(i, e, t) {
    return this._getSlideRenderControllerFromRenderUnit(t).createObjectToPage(i, e);
  }
  setObjectActiveByPage(i, e, t) {
    return this._getSlideRenderControllerFromRenderUnit(t).setObjectActiveByPage(i, e);
  }
  removeObjectById(i, e, t) {
    this._getSlideRenderControllerFromRenderUnit(t).removeObjectById(i, e);
  }
  /**
   * append blank page
   */
  appendPage(i) {
    this._getSlideRenderControllerFromRenderUnit(i).appendPage();
  }
};
M = ci([
  li(0, Y)
], M);
const di = "slides-ui.config", ft = {}, Lt = {
  id: "slide.operation.delete-element",
  type: D.OPERATION,
  handler: (i, e) => {
    if (!(e != null && e.id)) return !1;
    const t = e.unitId, r = i.get(j).getUnit(t);
    if (!r) return !1;
    const s = r.getActivePage();
    return delete s.pageElements[e.id], r.updatePage(s.id, s), i.get(M).removeObjectById(e.id, s.id, t), !0;
  }
};
var Be = { exports: {} }, me = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ct;
function ui() {
  if (Ct) return me;
  Ct = 1;
  var i = ke, e = Symbol.for("react.element"), t = Symbol.for("react.fragment"), n = Object.prototype.hasOwnProperty, r = i.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, s = { key: !0, ref: !0, __self: !0, __source: !0 };
  function o(a, c, g) {
    var l, u = {}, h = null, v = null;
    g !== void 0 && (h = "" + g), c.key !== void 0 && (h = "" + c.key), c.ref !== void 0 && (v = c.ref);
    for (l in c) n.call(c, l) && !s.hasOwnProperty(l) && (u[l] = c[l]);
    if (a && a.defaultProps) for (l in c = a.defaultProps, c) u[l] === void 0 && (u[l] = c[l]);
    return { $$typeof: e, type: a, key: h, ref: v, props: u, _owner: r.current };
  }
  return me.Fragment = t, me.jsx = o, me.jsxs = o, me;
}
var vt;
function mi() {
  return vt || (vt = 1, Be.exports = ui()), Be.exports;
}
var d = mi();
function At(i) {
  var e, t, n = "";
  if (typeof i == "string" || typeof i == "number") n += i;
  else if (typeof i == "object") if (Array.isArray(i)) {
    var r = i.length;
    for (e = 0; e < r; e++) i[e] && (t = At(i[e])) && (n && (n += " "), n += t);
  } else for (t in i) i[t] && (n && (n += " "), n += t);
  return n;
}
function T() {
  for (var i, e, t = 0, n = "", r = arguments.length; t < r; t++) (i = arguments[t]) && (e = At(i)) && (n && (n += " "), n += e);
  return n;
}
var N = function() {
  return N = Object.assign || function(i) {
    for (var e, t = 1, n = arguments.length; t < n; t++) {
      e = arguments[t];
      for (var r in e) Object.prototype.hasOwnProperty.call(e, r) && (i[r] = e[r]);
    }
    return i;
  }, N.apply(this, arguments);
}, gi = function(i, e) {
  var t = {};
  for (var n in i) Object.prototype.hasOwnProperty.call(i, n) && e.indexOf(n) < 0 && (t[n] = i[n]);
  if (i != null && typeof Object.getOwnPropertySymbols == "function")
    for (var r = 0, n = Object.getOwnPropertySymbols(i); r < n.length; r++)
      e.indexOf(n[r]) < 0 && Object.prototype.propertyIsEnumerable.call(i, n[r]) && (t[n[r]] = i[n[r]]);
  return t;
}, H = A(function(i, e) {
  var t = i.icon, n = i.id, r = i.className, s = i.extend, o = gi(i, ["icon", "id", "className", "extend"]), a = "univerjs-icon univerjs-icon-".concat(n, " ").concat(r || "").trim(), c = Ut("_".concat(fi()));
  return $t(t, "".concat(n), { defIds: t.defIds, idSuffix: c.current }, N({ ref: e, className: a }, o), s);
});
function $t(i, e, t, n, r) {
  return $(i.tag, N(N({ key: e }, hi(i, t, r)), n), (pi(i, t).children || []).map(function(s, o) {
    return $t(s, "".concat(e, "-").concat(i.tag, "-").concat(o), t, void 0, r);
  }));
}
function hi(i, e, t) {
  var n = N({}, i.attrs);
  t != null && t.colorChannel1 && n.fill === "colorChannel1" && (n.fill = t.colorChannel1);
  var r = e.defIds;
  return !r || r.length === 0 || (i.tag === "use" && n["xlink:href"] && (n["xlink:href"] = n["xlink:href"] + e.idSuffix), Object.entries(n).forEach(function(s) {
    var o = s[0], a = s[1];
    typeof a == "string" && (n[o] = a.replace(/url\(#(.*)\)/, "url(#$1".concat(e.idSuffix, ")")));
  })), n;
}
function pi(i, e) {
  var t, n = e.defIds;
  return !n || n.length === 0 ? i : i.tag === "defs" && (!((t = i.children) === null || t === void 0) && t.length) ? N(N({}, i), { children: i.children.map(function(r) {
    return typeof r.attrs.id == "string" && n && n.indexOf(r.attrs.id) > -1 ? N(N({}, r), { attrs: N(N({}, r.attrs), { id: r.attrs.id + e.idSuffix }) }) : r;
  }) }) : i;
}
function fi() {
  return Math.random().toString(36).substring(2, 8);
}
H.displayName = "UniverIcon";
var Ci = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M2.2498 3.65005C2.2498 2.87685 2.87661 2.25005 3.64981 2.25005H7.9998C8.33118 2.25005 8.5998 1.98142 8.5998 1.65005C8.5998 1.31868 8.33118 1.05005 7.9998 1.05005H3.64981C2.21387 1.05005 1.0498 2.21411 1.0498 3.65005V12.35C1.0498 13.786 2.21386 14.95 3.6498 14.95H12.3498C13.7857 14.95 14.9498 13.786 14.9498 12.3501V8.00005C14.9498 7.66868 14.6812 7.40005 14.3498 7.40005C14.0184 7.40005 13.7498 7.66868 13.7498 8.00005V9.23974L12.2385 8.1063C11.7252 7.72129 11.0068 7.7723 10.5531 8.22605L9.00869 9.77041L6.73916 7.8251C6.24387 7.40055 5.5095 7.41278 5.02864 7.85359L2.2498 10.4009V3.65005ZM2.2498 12.0287V12.35C2.2498 13.1232 2.87661 13.75 3.6498 13.75H12.3498C13.123 13.75 13.7498 13.1232 13.7498 12.3501V10.7397L11.5186 9.06631C11.4829 9.03956 11.433 9.04314 11.4016 9.07458L9.92249 10.5537L11.1015 11.5642C11.3531 11.7799 11.3822 12.1587 11.1666 12.4103C10.9509 12.6619 10.5721 12.691 10.3205 12.4753L5.9582 8.7362C5.92384 8.70674 5.87288 8.70758 5.83952 8.73816L2.2498 12.0287Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M11.8097 1.14783C12.1411 1.14783 12.4097 1.41646 12.4097 1.74783V3.297H14.1541C14.4855 3.297 14.7541 3.56563 14.7541 3.897C14.7541 4.22837 14.4855 4.497 14.1541 4.497H12.4097V6.24167C12.4097 6.57304 12.1411 6.84167 11.8097 6.84167C11.4783 6.84167 11.2097 6.57304 11.2097 6.24167V4.497H9.6603C9.32893 4.497 9.0603 4.22837 9.0603 3.897C9.0603 3.56563 9.32893 3.297 9.6603 3.297H11.2097V1.74783C11.2097 1.41646 11.4783 1.14783 11.8097 1.14783Z" } }] }, Ht = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "add-image-single",
    ref: e,
    icon: Ci
  }));
});
Ht.displayName = "AddImageSingle";
var vi = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "colorChannel1", d: "M11.0363 12.2367V14.0367C11.0363 14.3681 11.3049 14.6367 11.6363 14.6367C11.9676 14.6367 12.2363 14.3681 12.2363 14.0367V12.2367H14.0364C14.3677 12.2367 14.6364 11.9681 14.6364 11.6367C14.6364 11.3054 14.3677 11.0367 14.0364 11.0367H12.2363V9.23672C12.2363 8.90535 11.9676 8.63672 11.6363 8.63672C11.3049 8.63672 11.0363 8.90535 11.0363 9.23672V11.0367H9.23635C8.90498 11.0367 8.63635 11.3054 8.63635 11.6367C8.63635 11.9681 8.90498 12.2367 9.23635 12.2367H11.0363Z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M2.56365 1.36377C1.90091 1.36377 1.36365 1.90103 1.36365 2.56377V6.16377C1.36365 6.82651 1.90091 7.36377 2.56365 7.36377H6.16365C6.82639 7.36377 7.36365 6.82651 7.36365 6.16377V2.56377C7.36365 1.90103 6.82639 1.36377 6.16365 1.36377H2.56365zM6.16365 2.56377H2.56365L2.56365 6.16377H6.16365V2.56377zM2.56365 8.63647C1.90091 8.63647 1.36365 9.17373 1.36365 9.83647V13.4365C1.36365 14.0992 1.90091 14.6365 2.56365 14.6365H6.16365C6.82639 14.6365 7.36365 14.0992 7.36365 13.4365V9.83647C7.36365 9.17373 6.82639 8.63647 6.16365 8.63647H2.56365zM6.16365 9.83647H2.56365L2.56365 13.4365H6.16365V9.83647zM9.83635 7.36377C9.17361 7.36377 8.63635 6.82651 8.63635 6.16377V2.56377C8.63635 1.90103 9.17361 1.36377 9.83635 1.36377H13.4364C14.0991 1.36377 14.6364 1.90103 14.6364 2.56377V6.16377C14.6364 6.82651 14.0991 7.36377 13.4364 7.36377H9.83635zM9.83635 6.16377V2.56377L13.4364 2.56377V6.16377H9.83635z", fillRule: "evenodd", clipRule: "evenodd" } }] }, Ft = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "autofill",
    ref: e,
    icon: vi
  }));
});
Ft.displayName = "Autofill";
var _i = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M14.0045 4.4334C14.8881 4.4334 15.6045 3.71705 15.6045 2.8334 15.6045 1.94974 14.8881 1.2334 14.0045 1.2334H3.70449C2.82084 1.2334 2.10449 1.94974 2.10449 2.8334 2.10449 3.71705 2.82084 4.4334 3.70449 4.4334H14.0045zM14.4045 2.8334C14.4045 3.05431 14.2254 3.2334 14.0045 3.2334H3.70449C3.48358 3.2334 3.30449 3.05431 3.30449 2.8334 3.30449 2.61248 3.48358 2.4334 3.70449 2.4334H14.0045C14.2254 2.4334 14.4045 2.61249 14.4045 2.8334zM14.1544 8.5999C15.038 8.5999 15.7544 7.88356 15.7544 6.9999 15.7544 6.11625 15.038 5.3999 14.1544 5.3999H3.85439C2.97074 5.3999 2.25439 6.11625 2.25439 6.9999 2.25439 7.88356 2.97074 8.5999 3.85439 8.5999H14.1544zM14.5544 6.9999C14.5544 7.22082 14.3753 7.3999 14.1544 7.3999H3.85439C3.63348 7.3999 3.45439 7.22082 3.45439 6.9999 3.45439 6.77899 3.63348 6.5999 3.85439 6.5999H14.1544C14.3753 6.5999 14.5544 6.77899 14.5544 6.9999z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M8.57975 14.5902L6.58023 12.5907C6.34591 12.3564 6.34591 11.9765 6.58023 11.7421 6.81454 11.5078 7.19444 11.5078 7.42876 11.7421L8.40449 12.7179V10.1664C8.40449 9.83504 8.67312 9.56641 9.00449 9.56641 9.33586 9.56641 9.60449 9.83504 9.60449 10.1664V12.7179L10.5802 11.7421C10.8145 11.5078 11.1944 11.5078 11.4288 11.7421 11.6631 11.9765 11.6631 12.3564 11.4288 12.5907L9.42923 14.5902M8.57975 14.5902C8.58121 14.5917 8.58268 14.5931 8.58416 14.5946 8.64077 14.6502 8.70566 14.6923 8.77482 14.7209 8.84557 14.7502 8.92314 14.7664 9.00449 14.7664 9.08585 14.7664 9.16342 14.7502 9.23416 14.7209 9.30332 14.6923 9.36821 14.6502 9.42482 14.5946" } }] }, zt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "bottom-single",
    ref: e,
    icon: _i
  }));
});
zt.displayName = "BottomSingle";
var Si = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "g", attrs: { clipPath: "url(#graph-single_clip0_1293_28)" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M6.29541 11.1896C3.83208 10.7489 1.9624 8.59552 1.9624 6.00544C1.9624 3.09673 4.32036 0.73877 7.22907 0.73877C9.93483 0.73877 12.164 2.77918 12.4619 5.40527H13.5621C14.998 5.40527 16.1621 6.56933 16.1621 8.00527V12.6719C16.1621 14.1079 14.998 15.2719 13.5621 15.2719H8.89541C7.45947 15.2719 6.29541 14.1079 6.29541 12.6719V11.1896ZM3.1624 6.00544C3.1624 3.75947 4.98311 1.93877 7.22907 1.93877C9.27112 1.93877 10.9616 3.44388 11.2518 5.40527H8.89541C7.45947 5.40527 6.29541 6.56933 6.29541 8.00527V9.96442C4.49946 9.5425 3.1624 7.93012 3.1624 6.00544ZM7.49541 8.00527C7.49541 7.23207 8.12221 6.60527 8.89541 6.60527H13.5621C14.3353 6.60527 14.9621 7.23207 14.9621 8.00527V12.6719C14.9621 13.4451 14.3353 14.0719 13.5621 14.0719H8.89541C8.12221 14.0719 7.49541 13.4451 7.49541 12.6719V8.00527Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, { tag: "defs", attrs: {}, children: [{ tag: "clipPath", attrs: { id: "graph-single_clip0_1293_28" }, children: [{ tag: "path", attrs: { fill: "#fff", d: "M0 0H16V16H0z", transform: "translate(.9)" } }] }] }], defIds: ["graph-single_clip0_1293_28"] }, Wt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "graph-single",
    ref: e,
    icon: Si
  }));
});
Wt.displayName = "GraphSingle";
var Ii = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M11.3536 6.14645C11.5488 6.34171 11.5488 6.65829 11.3536 6.85355L8.35355 9.85355C8.15829 10.0488 7.84171 10.0488 7.64645 9.85355L4.64645 6.85355C4.45118 6.65829 4.45118 6.34171 4.64645 6.14645C4.84171 5.95118 5.15829 5.95118 5.35355 6.14645L8 8.79289L10.6464 6.14645C10.8417 5.95118 11.1583 5.95118 11.3536 6.14645Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, qe = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "more-down-single",
    ref: e,
    icon: Ii
  }));
});
qe.displayName = "MoreDownSingle";
var Pi = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M1.25 2.96401C1.25 3.84767 1.96634 4.56401 2.85 4.56401H13.15C14.0337 4.56401 14.75 3.84767 14.75 2.96401C14.75 2.08036 14.0337 1.36401 13.15 1.36401H2.85C1.96635 1.36401 1.25 2.08036 1.25 2.96401ZM2.85 3.36401C2.62909 3.36401 2.45 3.18493 2.45 2.96401C2.45 2.7431 2.62909 2.56401 2.85 2.56401H13.15C13.3709 2.56401 13.55 2.7431 13.55 2.96401C13.55 3.18493 13.3709 3.36401 13.15 3.36401H2.85Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M5.57564 11.6118C5.80995 11.3774 6.18985 11.3774 6.42417 11.6118L7.3999 12.5875V6.36951C7.3999 6.03814 7.66853 5.76951 7.9999 5.76951C8.33127 5.76951 8.5999 6.03814 8.5999 6.36951V12.5875L9.57564 11.6118C9.80995 11.3774 10.1899 11.3774 10.4242 11.6118C10.6585 11.8461 10.6585 12.226 10.4242 12.4603L8.4324 14.452C8.32324 14.5655 8.16982 14.6362 7.9999 14.6362C7.82998 14.6362 7.67655 14.5655 7.56739 14.452L5.57564 12.4603C5.34132 12.226 5.34132 11.8461 5.57564 11.6118Z" } }] }, Kt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "move-down-single",
    ref: e,
    icon: Pi
  }));
});
Kt.displayName = "MoveDownSingle";
var bi = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M1.25 13.036C1.25 12.1523 1.96634 11.436 2.85 11.436H13.15C14.0337 11.436 14.75 12.1523 14.75 13.036C14.75 13.9196 14.0337 14.636 13.15 14.636H2.85C1.96635 14.636 1.25 13.9196 1.25 13.036ZM2.85 12.636C2.62909 12.636 2.45 12.8151 2.45 13.036C2.45 13.2569 2.62909 13.436 2.85 13.436H13.15C13.3709 13.436 13.55 13.2569 13.55 13.036C13.55 12.8151 13.3709 12.636 13.15 12.636H2.85Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M5.57564 4.38825C5.80995 4.62256 6.18985 4.62256 6.42417 4.38825L7.3999 3.41251V9.63049C7.3999 9.96186 7.66853 10.2305 7.9999 10.2305C8.33127 10.2305 8.5999 9.96186 8.5999 9.63049V3.41251L9.57564 4.38825C9.80995 4.62256 10.1899 4.62256 10.4242 4.38825C10.6585 4.15393 10.6585 3.77403 10.4242 3.53972L8.4324 1.54796C8.32324 1.43445 8.16982 1.36382 7.9999 1.36382C7.82998 1.36382 7.67655 1.43446 7.56739 1.54797L5.57564 3.53972C5.34132 3.77403 5.34132 4.15393 5.57564 4.38825Z" } }] }, Gt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "move-up-single",
    ref: e,
    icon: bi
  }));
});
Gt.displayName = "MoveUpSingle";
var xi = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "mask", attrs: { id: "mask0_102_1204", style: { maskType: "alpha" }, width: 16, height: 16, x: 0, y: 0, maskUnits: "userSpaceOnUse" }, children: [{ tag: "path", attrs: { fill: "#DCDCDC", d: "M0 0H16V16H0z" } }] }, { tag: "g", attrs: { mask: "url(#mask0_102_1204)" }, children: [{ tag: "rect", attrs: { width: 12.62, height: 2.79, x: 1.69, y: 12.25, fill: "colorChannel1", rx: 1.4 } }, { tag: "path", attrs: { fill: "currentColor", d: "M7.74646 2.1642C7.34284 1.65121 6.85947 1.28205 6.38447 1.09593C5.86863 0.893787 5.15166 0.852126 4.70047 1.42917C4.29842 1.94344 4.33082 2.70249 4.57691 3.50395C4.68808 3.8661 4.86846 4.22254 5.12522 4.56617C5.13695 4.58185 5.14927 4.5968 5.16211 4.61102C5.17152 4.6214 5.18122 4.63139 5.19115 4.64096C5.20459 4.65393 5.21855 4.66618 5.23296 4.6777L3.5064 6.40422C3.26756 6.64306 3.26756 7.03033 3.5064 7.26919L7.23756 11.0004C7.47648 11.2392 7.86375 11.2392 8.1026 11.0004L12.8329 6.27001C13.0717 6.03117 13.0717 5.6439 12.8329 5.40505L9.10172 1.67387C8.86288 1.43503 8.47561 1.43503 8.23676 1.67387L7.74646 2.1642ZM7.46162 4.50638C7.47252 4.57574 7.48163 4.64652 7.4888 4.71868C7.52232 5.05481 7.82194 5.30013 8.15807 5.26661C8.4942 5.23311 8.73954 4.93345 8.70601 4.59732C8.65786 4.11463 8.54251 3.6666 8.37892 3.26169L8.66928 2.97135L11.5354 5.83753L10.6106 6.76237H4.87824L7.40802 4.2326C7.42885 4.32128 7.44684 4.41255 7.46162 4.50638ZM6.87284 3.03784C6.57441 2.62008 6.22261 2.34634 5.93815 2.23487C5.80233 2.18165 5.71528 2.17827 5.67638 2.18335C5.6725 2.18387 5.66929 2.18443 5.6666 2.18497L5.6619 2.18607C5.64166 2.21954 5.53594 2.45979 5.74626 3.1449C5.81241 3.36036 5.92262 3.58602 6.09307 3.81761L6.87284 3.03784Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M12.1932 11.5463C12.6454 11.5463 13.012 11.1797 13.012 10.7275C13.012 10.4475 12.6981 9.92821 12.459 9.5733C12.3299 9.38173 12.0564 9.38174 11.9274 9.57331C11.6883 9.92822 11.3744 10.4475 11.3744 10.7275C11.3744 11.1797 11.741 11.5463 12.1932 11.5463Z" } }] }] }, Xt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "paint-bucket",
    ref: e,
    icon: xi
  }));
});
Xt.displayName = "PaintBucket";
var Ei = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "g", attrs: { fill: "currentColor", clipPath: "url(#text-single_clip0_1293_26)" }, children: [{ tag: "path", attrs: { d: "M2.22891 2.07227C1.89754 2.07227 1.62891 2.34089 1.62891 2.67227C1.62891 3.00364 1.89754 3.27227 2.22891 3.27227H6.29541V14.0056C6.29541 14.337 6.56404 14.6056 6.89541 14.6056C7.22678 14.6056 7.49541 14.337 7.49541 14.0056V3.27227H11.5622C11.8936 3.27227 12.1622 3.00364 12.1622 2.67227C12.1622 2.34089 11.8936 2.07227 11.5622 2.07227H2.22891Z" } }, { tag: "path", attrs: { d: "M10.2289 6.40552C9.89754 6.40552 9.62891 6.67415 9.62891 7.00552C9.62891 7.33689 9.89754 7.60552 10.2289 7.60552H12.2954V14.0055C12.2954 14.3369 12.564 14.6055 12.8954 14.6055C13.2268 14.6055 13.4954 14.3369 13.4954 14.0055V7.60552H15.5622C15.8936 7.60552 16.1622 7.33689 16.1622 7.00552C16.1622 6.67415 15.8936 6.40552 15.5622 6.40552H10.2289Z" } }] }, { tag: "defs", attrs: {}, children: [{ tag: "clipPath", attrs: { id: "text-single_clip0_1293_26" }, children: [{ tag: "path", attrs: { fill: "#fff", d: "M0 0H16V16H0z", transform: "translate(.9)" } }] }] }], defIds: ["text-single_clip0_1293_26"] }, Yt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "text-single",
    ref: e,
    icon: Ei
  }));
});
Yt.displayName = "TextSingle";
var Ri = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M7.82994 1.40913C7.88746 1.35161 7.95376 1.30821 8.02453 1.27893C8.09527 1.24959 8.17285 1.2334 8.2542 1.2334C8.33555 1.2334 8.41313 1.24959 8.48387 1.27893C8.55464 1.30821 8.62094 1.35161 8.67846 1.40913L10.6785 3.40913C10.9128 3.64345 10.9128 4.02335 10.6785 4.25766C10.4441 4.49198 10.0642 4.49198 9.82994 4.25766L8.8542 3.28193V5.8334C8.8542 6.16477 8.58557 6.4334 8.2542 6.4334C7.92283 6.4334 7.6542 6.16477 7.6542 5.8334V3.28193L6.67846 4.25766C6.44415 4.49198 6.06425 4.49198 5.82994 4.25766C5.59562 4.02335 5.59562 3.64345 5.82994 3.40913L7.82994 1.40913Z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.50439 9C1.50439 8.11634 2.22074 7.4 3.10439 7.4H13.4044C14.288 7.4 15.0044 8.11634 15.0044 9 15.0044 9.88366 14.2881 10.6 13.4044 10.6H3.1044C2.22074 10.6 1.50439 9.88366 1.50439 9zM3.10439 8.6C2.88348 8.6 2.70439 8.77909 2.70439 9 2.70439 9.22091 2.88348 9.4 3.1044 9.4H13.4044C13.6253 9.4 13.8044 9.22091 13.8044 9 13.8044 8.77909 13.6253 8.6 13.4044 8.6H3.10439zM1.6543 13.1665C1.6543 12.2828 2.37064 11.5665 3.2543 11.5665H13.5543C14.438 11.5665 15.1543 12.2828 15.1543 13.1665 15.1543 14.0502 14.438 14.7665 13.5543 14.7665H3.2543C2.37064 14.7665 1.6543 14.0502 1.6543 13.1665zM3.2543 12.7665C3.03338 12.7665 2.8543 12.9456 2.8543 13.1665 2.8543 13.3874 3.03338 13.5665 3.2543 13.5665H13.5543C13.7752 13.5665 13.9543 13.3874 13.9543 13.1665 13.9543 12.9456 13.7752 12.7665 13.5543 12.7665H3.2543z", fillRule: "evenodd", clipRule: "evenodd" } }] }, kt = A(function(i, e) {
  return $(H, Object.assign({}, i, {
    id: "topmost-single",
    ref: e,
    icon: Ri
  }));
});
kt.displayName = "TopmostSingle";
const K = {
  id: "slide.operation.update-element",
  type: D.OPERATION,
  handler: (i, e) => {
    const { oKey: t, props: n } = e, r = i.get(j), s = e == null ? void 0 : e.unitId, o = r.getUnit(s);
    if (!o) return !1;
    const a = o.getActivePage();
    return a.pageElements[t] = vn.deepMerge(a.pageElements[t], n), o.updatePage(a.id, a), !0;
  }
}, Oi = "univer-image-common-panel", ji = "univer-image-common-panel-grid", yi = "univer-image-common-panel-border", Ti = "univer-image-common-panel-title", Mi = "univer-image-common-panel-subtitle", wi = "univer-image-common-panel-row", Ni = "univer-image-common-panel-row-vertical", Di = "univer-image-common-panel-column", Ui = "univer-image-common-panel-column-center", Vi = "univer-image-common-panel-inline", Bi = "univer-image-common-panel-span2", Li = "univer-image-common-panel-span3", Ai = "univer-image-common-panel-input", $i = "univer-slide-panel-color-picker", m = {
  imageCommonPanel: Oi,
  imageCommonPanelGrid: ji,
  imageCommonPanelBorder: yi,
  imageCommonPanelTitle: Ti,
  imageCommonPanelSubtitle: Mi,
  imageCommonPanelRow: wi,
  imageCommonPanelRowVertical: Ni,
  imageCommonPanelColumn: Di,
  imageCommonPanelColumnCenter: Ui,
  imageCommonPanelInline: Vi,
  imageCommonPanelSpan2: Bi,
  imageCommonPanelSpan3: Li,
  imageCommonPanelInput: Ai,
  slidePanelColorPicker: $i
};
function Hi(i) {
  const { pageId: e, unitId: t } = i, n = O(Q), r = O(M), s = O(U), o = r.getRenderUnitByPageId(e, t), a = o == null ? void 0 : o.scene;
  if (!a) return null;
  const c = a.getTransformer();
  if (!c) return null;
  const l = c.getSelectedObjectMap().values().next().value;
  if (!l) return null;
  const u = (h) => {
    const v = a.getAllObjects(), [_, C] = v.reduce(([f, I], b) => {
      const E = b.zIndex, w = E < f ? E : f, V = E > I ? E : I;
      return [w, V];
    }, [0, 0]);
    let p = l.zIndex;
    h === 3 ? p = _ - 1 : h === 2 ? p = C + 1 : h === 0 ? p = l.zIndex + 1 : h === 1 && (p = l.zIndex - 1), l.setProps({
      zIndex: p
    }), s.executeCommand(K.id, {
      unitId: t,
      oKey: l == null ? void 0 : l.oKey,
      props: {
        zIndex: p
      }
    });
  };
  return /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelGrid, children: [
    /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelTitle), children: /* @__PURE__ */ d.jsx("div", { children: n.t("image-panel.arrange.title") }) }) }),
    /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelRow, children: [
      /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan2), children: /* @__PURE__ */ d.jsx(Pe, { size: "small", onClick: () => {
        u(
          0
          /* forward */
        );
      }, children: /* @__PURE__ */ d.jsxs("span", { className: m.imageCommonPanelInline, children: [
        /* @__PURE__ */ d.jsx(Gt, {}),
        n.t("image-panel.arrange.forward")
      ] }) }) }),
      /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan2), children: /* @__PURE__ */ d.jsx(Pe, { size: "small", onClick: () => {
        u(
          1
          /* backward */
        );
      }, children: /* @__PURE__ */ d.jsxs("span", { className: m.imageCommonPanelInline, children: [
        /* @__PURE__ */ d.jsx(Kt, {}),
        n.t("image-panel.arrange.backward")
      ] }) }) })
    ] }),
    /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelRow, children: [
      /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan2), children: /* @__PURE__ */ d.jsx(Pe, { size: "small", onClick: () => {
        u(
          2
          /* front */
        );
      }, children: /* @__PURE__ */ d.jsxs("span", { className: m.imageCommonPanelInline, children: [
        /* @__PURE__ */ d.jsx(kt, {}),
        n.t("image-panel.arrange.front")
      ] }) }) }),
      /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan2), children: /* @__PURE__ */ d.jsx(Pe, { size: "small", onClick: () => {
        u(
          3
          /* back */
        );
      }, children: /* @__PURE__ */ d.jsxs("span", { className: m.imageCommonPanelInline, children: [
        /* @__PURE__ */ d.jsx(zt, {}),
        n.t("image-panel.arrange.back")
      ] }) }) })
    ] })
  ] });
}
function Fi(i) {
  const { pageId: e, unitId: t } = i, n = O(Q), r = O(M), s = O(U), o = r.getRenderUnitByPageId(e, t), a = o == null ? void 0 : o.scene;
  if (!a) return null;
  const c = a.getTransformer();
  if (!c) return null;
  const l = c.getSelectedObjectMap().values().next().value;
  if (!l) return null;
  const {
    width: u = 0,
    height: h = 0,
    left: v = 0,
    top: _ = 0,
    angle: C = 0
  } = l, [p, f] = W(u), [I, b] = W(h), [E, w] = W(v), [V, R] = W(_), [G, ce] = W(C), F = (S) => {
    const { objects: le } = S, de = le.values().next().value, {
      width: ln = 0,
      height: dn = 0,
      left: un = 0,
      top: mn = 0,
      angle: gn = 0
    } = de;
    f(ln), b(dn), w(un), R(mn), ce(gn);
  };
  oe(() => {
    const S = c.changeStart$.subscribe((de) => {
      F(de);
    }), le = c.changing$.subscribe((de) => {
      F(de);
    });
    return () => {
      le.unsubscribe(), S.unsubscribe();
    };
  }, []);
  function z(S) {
    !S || !l || (s.executeCommand(K.id, {
      pageId: e,
      oKey: l.oKey,
      props: {
        width: S
      }
    }), l == null || l.resize(S, l.height), f(S), c == null || c.refreshControls());
  }
  function ee(S) {
    !S || !l || (s.executeCommand(K.id, {
      pageId: e,
      oKey: l.oKey,
      props: {
        height: S
      }
    }), l == null || l.resize(l.width, S), b(S), c == null || c.refreshControls());
  }
  function te(S) {
    !S || !l || (s.executeCommand(K.id, {
      pageId: e,
      oKey: l.oKey,
      props: {
        left: S
      }
    }), l == null || l.translate(S, l.top), w(S), c == null || c.refreshControls());
  }
  function Se(S) {
    !S || !l || (s.executeCommand(K.id, {
      pageId: e,
      oKey: l.oKey,
      props: {
        right: S
      }
    }), l == null || l.translate(l.left, S), R(S), c == null || c.refreshControls());
  }
  function Ie(S) {
    !S || !l || (s.executeCommand(K.id, {
      pageId: e,
      oKey: l.oKey,
      props: {
        angle: S
      }
    }), l == null || l.transformByState({
      angle: S
    }), ce(S), c == null || c.refreshControls());
  }
  return /* @__PURE__ */ d.jsxs(
    "div",
    {
      className: T(m.imageCommonPanelGrid, m.imageCommonPanelBorder),
      children: [
        /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelTitle), children: /* @__PURE__ */ d.jsx("div", { children: n.t("image-panel.transform.title") }) }) }),
        /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelRow, children: [
          /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan3), children: /* @__PURE__ */ d.jsxs("label", { children: [
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: n.t("image-panel.transform.width") }) }),
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: /* @__PURE__ */ d.jsx(
              ue,
              {
                className: m.imageCommonPanelInput,
                min: 1,
                value: p,
                onChange: (S) => {
                  z(S);
                }
              }
            ) }) })
          ] }) }),
          /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan3), children: /* @__PURE__ */ d.jsxs("label", { children: [
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: n.t("image-panel.transform.height") }) }),
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: /* @__PURE__ */ d.jsx(
              ue,
              {
                className: m.imageCommonPanelInput,
                min: 1,
                value: I,
                onChange: (S) => {
                  ee(S);
                }
              }
            ) }) })
          ] }) })
        ] }),
        /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelRow, children: [
          /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan3), children: /* @__PURE__ */ d.jsxs("label", { children: [
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: n.t("image-panel.transform.x") }) }),
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: /* @__PURE__ */ d.jsx(
              ue,
              {
                className: m.imageCommonPanelInput,
                precision: 1,
                min: 0,
                value: E,
                onChange: (S) => {
                  te(S);
                }
              }
            ) }) })
          ] }) }),
          /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan3), children: /* @__PURE__ */ d.jsxs("label", { children: [
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: n.t("image-panel.transform.y") }) }),
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: /* @__PURE__ */ d.jsx(
              ue,
              {
                className: m.imageCommonPanelInput,
                precision: 1,
                min: 0,
                value: V,
                onChange: (S) => {
                  Se(S);
                }
              }
            ) }) })
          ] }) }),
          /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan3), children: /* @__PURE__ */ d.jsxs("label", { children: [
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: n.t("image-panel.transform.rotate") }) }),
            /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelColumn, children: /* @__PURE__ */ d.jsx(
              ue,
              {
                className: m.imageCommonPanelInput,
                precision: 1,
                value: G,
                onChange: Ie
              }
            ) }) })
          ] }) })
        ] })
      ]
    }
  );
}
function zi(i) {
  var _, C;
  const { pageId: e, unitId: t } = i, n = O(Q), r = O(M), s = O(U), o = r.getRenderUnitByPageId(e, t), a = o == null ? void 0 : o.scene;
  if (!a) return null;
  const c = a.getTransformer();
  if (!c) return null;
  const l = c.getSelectedObjectMap().values().next().value;
  if (!l) return null;
  const [u, h] = ke.useState((C = (_ = l.fill) == null ? void 0 : _.toString()) != null ? C : "");
  function v(p) {
    l == null || l.setProps({
      fill: p
    }), s.executeCommand(K.id, {
      unitId: t,
      oKey: l == null ? void 0 : l.oKey,
      props: {
        shape: {
          shapeProperties: {
            shapeBackgroundFill: {
              rgb: p
            }
          }
        }
      }
    }), h(p);
  }
  return /* @__PURE__ */ d.jsx(
    "div",
    {
      className: T(m.imageCommonPanelGrid, m.imageCommonPanelBorder),
      children: /* @__PURE__ */ d.jsxs("div", { className: m.imageCommonPanelGrid, children: [
        /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelTitle), children: /* @__PURE__ */ d.jsx("div", { children: n.t("slide.panel.fill.title") }) }) }),
        /* @__PURE__ */ d.jsx("div", { className: m.imageCommonPanelRow, children: /* @__PURE__ */ d.jsx("div", { className: T(m.imageCommonPanelColumn, m.imageCommonPanelSpan2), children: /* @__PURE__ */ d.jsx(
          Vt,
          {
            align: {
              offset: [0, 18]
            },
            overlay: /* @__PURE__ */ d.jsx("section", { className: m.slidePanelColorPicker, children: /* @__PURE__ */ d.jsx(
              Wn,
              {
                value: "#fff",
                onChange: v
              }
            ) }),
            children: /* @__PURE__ */ d.jsxs("a", { className: m.uiPluginSheetsBorderPanelButton, children: [
              /* @__PURE__ */ d.jsx(Xt, { extend: { colorChannel1: u != null ? u : "rgb(var(--primary-color))" } }),
              /* @__PURE__ */ d.jsx("span", { className: m.uiPluginSheetsBorderPanelMoreIcon, children: /* @__PURE__ */ d.jsx(qe, {}) })
            ] })
          }
        ) }) })
      ] })
    }
  );
}
const Wi = "univer-image-common-panel", Ki = "univer-image-common-panel-grid", Gi = "univer-image-common-panel-border", Xi = "univer-image-common-panel-title", Yi = "univer-image-common-panel-subtitle", ki = "univer-image-common-panel-row", Zi = "univer-image-common-panel-row-vertical", qi = "univer-image-common-panel-column", Ji = "univer-image-common-panel-column-center", Qi = "univer-image-common-panel-inline", er = "univer-image-common-panel-span2", tr = "univer-image-common-panel-span3", nr = "univer-image-common-panel-input", ir = {
  imageCommonPanel: Wi,
  imageCommonPanelGrid: Ki,
  imageCommonPanelBorder: Gi,
  imageCommonPanelTitle: Xi,
  imageCommonPanelSubtitle: Yi,
  imageCommonPanelRow: ki,
  imageCommonPanelRowVertical: Zi,
  imageCommonPanelColumn: qi,
  imageCommonPanelColumnCenter: Ji,
  imageCommonPanelInline: Qi,
  imageCommonPanelSpan2: er,
  imageCommonPanelSpan3: tr,
  imageCommonPanelInput: nr
}, Re = "COMPONENT_SLIDE_SIDEBAR";
function rr() {
  var g, l, u;
  const i = O(j), e = O(M), t = i.getCurrentUnitForType(x.UNIVER_SLIDE), n = (g = t == null ? void 0 : t.getActivePage()) == null ? void 0 : g.id;
  if (!n) return null;
  const s = (l = e.getRenderUnitByPageId(n, n).scene) == null ? void 0 : l.getTransformer();
  if (!s) return null;
  const a = s.getSelectedObjectMap().values().next().value;
  if (!a)
    return null;
  const c = ((u = i.getFocusedUnit()) == null ? void 0 : u.getUnitId()) || "";
  return /* @__PURE__ */ d.jsxs("section", { className: ir.imageCommonPanel, children: [
    /* @__PURE__ */ d.jsx(Hi, { pageId: n, unitId: c }),
    /* @__PURE__ */ d.jsx(Fi, { pageId: n, unitId: c }),
    a.objectType === he.RECT && /* @__PURE__ */ d.jsx(zi, { pageId: n, unitId: c })
  ] });
}
const Je = {
  id: "slide.command.insert-float-shape",
  type: D.COMMAND,
  handler: async (i) => {
    var r;
    const e = i.get(U), n = (r = i.get(j).getFocusedUnit()) == null ? void 0 : r.getUnitId();
    return e.executeCommand(Zt.id, { unitId: n });
  }
}, Zt = {
  id: "slide.operation.insert-float-shape",
  type: D.OPERATION,
  handler: async (i, e) => {
    const t = xt(6), n = i.get(j), r = e.unitId, s = n.getUnit(r);
    if (!s) return !1;
    const o = s.getActivePage(), a = Object.values(o.pageElements), c = a != null && a.length ? Math.max(...a.map((h) => h.zIndex)) : 20, g = {
      id: t,
      zIndex: c + 1,
      left: 378,
      top: 142,
      width: 250,
      height: 250,
      title: t,
      description: "",
      type: Xe.SHAPE,
      shape: {
        shapeType: _n.Rect,
        text: "",
        shapeProperties: {
          shapeBackgroundFill: {
            rgb: "rgb(0,0,255)"
          }
        }
      }
    };
    o.pageElements[t] = g, s.updatePage(o.id, o);
    const l = i.get(M), u = l.createObjectToPage(g, o.id, r);
    return u && l.setObjectActiveByPage(u, o.id, r), !0;
  }
}, He = {
  id: "sidebar.operation.slide-shape",
  type: D.COMMAND,
  handler: async (i, e) => {
    const { visible: t, objectType: n } = e, r = i.get(Dt), s = i.get(Q);
    let o = "", a = "";
    return n === he.RECT ? (o = "slide.sidebar.shape", a = Re) : n === he.IMAGE ? (o = "slide.sidebar.image", a = Re) : n === he.RICH_TEXT && (o = "slide.sidebar.text", a = Re), t ? r.open({
      header: { title: s.t(o) },
      children: { label: a },
      onClose: () => {
      },
      width: 360
    }) : r.close(), !0;
  }
}, qt = "COMPONENT_SLIDE_IMAGE_POPUP_MENU";
var sr = Object.defineProperty, or = Object.getOwnPropertyDescriptor, ar = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? or(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && sr(e, t, r), r;
}, be = (i, e) => (t, n) => e(t, n, i);
function cr(i, e) {
  const t = _t(i.left, i.top, e), n = _t(i.right, i.bottom, e);
  return {
    left: t.x,
    top: t.y,
    right: n.x,
    bottom: n.y
  };
}
function _t(i, e, t) {
  const { scaleX: n, scaleY: r } = t.getAncestorScale(), s = t.getViewport(se.VIEW);
  if (!s)
    return {
      x: i,
      y: e
    };
  const { viewportScrollX: o, viewportScrollY: a } = s, c = (i - o) * n, g = (e - a) * r;
  return {
    x: c,
    y: g
  };
}
let ye = class extends Ue {
  constructor(i, e, t, n) {
    super(), this._globalPopupManagerService = i, this._renderManagerService = e, this._univerInstanceService = t, this._commandService = n;
  }
  _createObjectPositionObserver(i, e) {
    const n = (() => {
      var F, z, ee, te;
      const { scene: o, engine: a } = e, { left: c, top: g, width: l, height: u } = i, h = (o.width - ((z = (F = e.mainComponent) == null ? void 0 : F.width) != null ? z : 0)) / 2, v = (o.height - ((te = (ee = e.mainComponent) == null ? void 0 : ee.height) != null ? te : 0)) / 2, _ = {
        left: c,
        right: c + l,
        top: g,
        bottom: g + u
      }, C = a.getCanvasElement(), p = C.getBoundingClientRect(), f = wn(C.style.width), { scaleX: I, scaleY: b } = o.getAncestorScale(), E = cr(_, o), { top: w, left: V, width: R } = p, G = R / f;
      return {
        left: E.left * G * I + V + h,
        right: E.right * G * I + V + h,
        top: E.top * G * b + w + v,
        bottom: E.bottom * G * b + w + v
      };
    })(), r = new J(n), s = new Ye();
    return {
      position: n,
      position$: r,
      disposable: s
    };
  }
  attachPopupToObject(i, e) {
    const n = this._univerInstanceService.getCurrentUnitForType(x.UNIVER_SLIDE).getUnitId(), r = this._renderManagerService.getRenderById(n);
    if (!r)
      return {
        dispose: () => {
        }
      };
    const { position: s, position$: o, disposable: a } = this._createObjectPositionObserver(i, r), c = this._globalPopupManagerService.addPopup({
      ...e,
      unitId: n,
      subUnitId: "default",
      anchorRect: s,
      anchorRect$: o,
      canvasElement: r.engine.getCanvasElement()
    });
    return {
      dispose: () => {
        this._globalPopupManagerService.removePopup(c), o.complete(), a.dispose();
      }
    };
  }
};
ye = ar([
  be(0, X(Dn)),
  be(1, Y),
  be(2, j),
  be(3, U)
], ye);
var lr = Object.defineProperty, dr = Object.getOwnPropertyDescriptor, ur = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? dr(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && lr(e, t, r), r;
}, Z = (i, e) => (t, n) => e(t, n, i);
let Te = class extends fe {
  constructor(e, t, n, r, s, o, a) {
    super();
    P(this, "_initImagePopupMenu", /* @__PURE__ */ new Set());
    this._canvasPopManagerService = e, this._renderManagerService = t, this._univerInstanceService = n, this._contextService = r, this._canvasView = s, this._sidebarService = o, this._commandService = a, this._init();
  }
  _init() {
    this._univerInstanceService.getAllUnitsForType(x.UNIVER_SLIDE).forEach((e) => this._create(e));
  }
  _create(e) {
    if (!e)
      return;
    const t = e.getUnitId();
    this._renderManagerService.has(t) && !this._initImagePopupMenu.has(t) && (this._popupMenuListener(t), this._initImagePopupMenu.add(t));
  }
  _hasCropObject(e) {
  }
  // eslint-disable-next-line max-lines-per-function
  _popupMenuListener(e) {
    var r;
    const t = this._univerInstanceService.getCurrentUnitForType(x.UNIVER_SLIDE), n = (r = t == null ? void 0 : t.getPages()) != null ? r : {};
    Object.keys(n).forEach((s) => {
      var g;
      const a = (g = this._canvasView.getRenderUnitByPageId(s, e).scene) == null ? void 0 : g.getTransformer();
      if (!a) return;
      let c;
      this.disposeWithMe(
        Et(
          a.createControl$.subscribe(() => {
            const l = a.getSelectedObjectMap();
            if (l.size > 1) {
              c == null || c.dispose();
              return;
            }
            const u = l.values().next().value;
            if (!u)
              return;
            const h = u.oKey;
            c == null || c.dispose(), c = this.disposeWithMe(this._canvasPopManagerService.attachPopupToObject(u, {
              componentKey: qt,
              direction: "horizontal",
              offset: [2, 0],
              extraProps: {
                menuItems: this._getMenuItemsByObjectType(u.objectType, h, e)
              }
            })), this._sidebarService.visible && this._commandService.executeCommand(He.id, {
              visible: !0,
              objectType: u.objectType
            });
          })
        )
      ), this.disposeWithMe(
        a.clearControl$.subscribe(() => {
          c == null || c.dispose(), this._contextService.setContextValue(Sn, !1);
        })
      ), this.disposeWithMe(
        a.changing$.subscribe(() => {
          c == null || c.dispose();
          const l = a.getSelectedObjectMap();
          if (l.size > 1) {
            c == null || c.dispose();
            return;
          }
          const u = l.values().next().value;
          u && this._commandService.executeCommand(K.id, {
            unitId: e,
            oKey: u.oKey,
            props: {
              width: u.width,
              height: u.height,
              left: u.left,
              top: u.top
            }
          });
        })
      );
    });
  }
  _getMenuItemsByObjectType(e, t, n) {
    return [{
      label: "slide.popup.edit",
      index: 0,
      commandId: He.id,
      commandParams: {
        visible: !0,
        objectType: e
      },
      disable: !1
    }, {
      label: "slide.popup.delete",
      index: 5,
      commandId: Lt.id,
      commandParams: {
        id: t,
        unitId: n
      },
      disable: !1
    }];
  }
};
Te = ur([
  Z(0, X(ye)),
  Z(1, Y),
  Z(2, j),
  Z(3, Ce),
  Z(4, X(M)),
  Z(5, Dt),
  Z(6, U)
], Te);
const Me = {
  id: "slide.operation.edit-arrow",
  type: D.OPERATION,
  handler: () => !0
}, ae = In("SLIDE_EDITOR");
var mr = Object.defineProperty, gr = Object.getOwnPropertyDescriptor, hr = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? gr(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && mr(e, t, r), r;
}, Le = (i, e) => (t, n) => e(t, n, i);
const Qe = Rt("univer.slide-editor-bridge.service");
let Fe = class extends Ue {
  constructor(e, t, n) {
    super();
    P(this, "_editorUnitId", ae);
    P(this, "_isForceKeepVisible", !1);
    P(this, "_editorIsDirty", !1);
    P(this, "_currentEditRectState", null);
    P(this, "_currentEditRectState$", new J(null));
    P(this, "currentEditRectState$", this._currentEditRectState$.asObservable());
    P(this, "_visibleParam", {
      visible: !1,
      eventType: q.Dblclick,
      unitId: ""
    });
    P(this, "_visible$", new J(this._visibleParam));
    P(this, "visible$", this._visible$.asObservable());
    P(this, "_afterVisible$", new J(this._visibleParam));
    P(this, "afterVisible$", this._afterVisible$.asObservable());
    P(this, "endEditing$", new Bt());
    P(this, "_currentEditRectInfo");
    this._editorService = e, this._contextService = t, this._renderManagerService = n;
  }
  dispose() {
    super.dispose();
  }
  getEditorRect() {
    return this._currentEditRectInfo;
  }
  /**
   * 1st part of startEditing.
   * @editorInfo editorInfo
   */
  setEditorRect(e) {
    this._currentEditRectInfo = e, this._editorService.getFocusEditor() || (this._editorService.focus(ae), this._contextService.setContextValue(Oe, !1), this._contextService.setContextValue(Ot, !1), this._contextService.setContextValue(jt, !1));
    const t = this.getEditRectState();
    this._currentEditRectState = t, this._currentEditRectState$.next(t);
  }
  changeVisible(e) {
    this._visibleParam = e, e.visible && (this._editorIsDirty = !1), this._visible$.next(this._visibleParam), this._afterVisible$.next(this._visibleParam);
  }
  /**
   * get info from _currentEditRectInfo
   *
   * invoked by slide-editing.render-controller.ts@_handleEditorVisible
   * && this@setEditorRect
   */
  getEditRectState() {
    const e = ae, t = this._currentEditRectInfo, n = t.unitId, r = t.richTextObj.documentData;
    r.id = e, r.documentStyle = {
      ...r.documentStyle,
      pageSize: { width: t.richTextObj.width, height: 1 / 0 }
    };
    const o = {
      documentModel: new Pn(r),
      fontString: "document",
      textRotation: { a: 0, v: 0 },
      wrapStrategy: 0,
      verticalAlign: je.TOP,
      horizontalAlign: yt.LEFT,
      paddingData: { t: 0, b: 1, l: 2, r: 2 }
    }, a = t.richTextObj.width, c = t.richTextObj.height, g = t.richTextObj.left, l = t.richTextObj.top, u = {
      left: 0,
      top: 0
    }, h = this._renderManagerService.getRenderById(n), v = h == null ? void 0 : h.scene, _ = v == null ? void 0 : v.getViewport(se.VIEW), C = v == null ? void 0 : v.getObject(se.COMPONENT), p = {
      x: (C == null ? void 0 : C.left) || 0,
      y: (C == null ? void 0 : C.top) || 0
    }, f = (_ == null ? void 0 : _.viewportScrollX) || 0, I = (_ == null ? void 0 : _.viewportScrollY) || 0;
    return u.left = p.x - f, u.top = p.y - I, {
      position: {
        startX: g,
        startY: l,
        endX: g + a,
        endY: l + c
      },
      scaleX: 1,
      scaleY: 1,
      slideCardOffset: u,
      unitId: n,
      editorUnitId: e,
      documentLayoutObject: o
    };
  }
  changeEditorDirty(e) {
    this._editorIsDirty = e;
  }
  isVisible() {
    return this._visibleParam.visible;
  }
  getEditorDirty() {
    return this._editorIsDirty;
  }
  getCurrentEditorId() {
    return this._editorUnitId;
  }
  /**
   * @deprecated
   */
  genDocData(e) {
    const t = this.getCurrentEditorId(), n = e.text, r = e.fs;
    return {
      id: t,
      body: {
        dataStream: `${n}\r
`,
        textRuns: [{ st: 0, ed: n.length }],
        paragraphs: [{
          paragraphStyle: {
            // no use
            // textStyle: { fs: 30 },
            // horizontalAlign: HorizontalAlign.CENTER,
            // verticalAlign: VerticalAlign.MIDDLE,
          },
          startIndex: n.length + 1
        }],
        sectionBreaks: [{ startIndex: n.length + 2 }]
      },
      documentStyle: {
        marginBottom: 0,
        marginLeft: 0,
        marginRight: 0,
        marginTop: 0,
        pageSize: { width: 1 / 0, height: 1 / 0 },
        textStyle: { fs: r },
        renderConfig: {
          // horizontalAlign: HorizontalAlign.CENTER,
          verticalAlign: je.MIDDLE,
          centerAngle: 0,
          vertexAngle: 0,
          wrapStrategy: 0
        }
      },
      drawings: {},
      drawingsOrder: [],
      settings: { zoomRatio: 1 }
    };
  }
};
Fe = hr([
  Le(0, Ze),
  Le(1, Ce),
  Le(2, Y)
], Fe);
class pr {
  constructor() {
    P(this, "_state", null);
    P(this, "_rect", null);
    P(this, "_state$", new J(null));
    P(this, "state$", this._state$.asObservable());
    P(this, "_rect$", new J(null));
    P(this, "rect$", this._rect$.asObservable());
    P(this, "_focus", !1);
    P(this, "_focus$", new J(this._focus));
    P(this, "focus$", this._focus$.asObservable());
  }
  dispose() {
    this._state$.complete(), this._state = null, this._rect$.complete(), this._rect = null;
  }
  setState(e) {
    this._state = e, this._refresh(e);
  }
  getRect() {
    return this._rect;
  }
  setRect(e) {
    this._rect = e, this._rect$.next(e);
  }
  getState() {
    return this._state;
  }
  setFocus(e = !1) {
    this._focus = e, this._focus$.next(e);
  }
  _refresh(e) {
    this._state$.next(e);
  }
}
const et = Rt(
  "univer.slide-editor-manager.service"
);
var k = /* @__PURE__ */ ((i) => (i[i.InitialState = 0] = "InitialState", i[i.StartEditor = 1] = "StartEditor", i[i.CursorChange = 2] = "CursorChange", i))(k || {}), fr = Object.defineProperty, Cr = Object.getOwnPropertyDescriptor, vr = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? Cr(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && fr(e, t, r), r;
}, L = (i, e) => (t, n) => e(t, n, i);
const St = -1e3, It = 5, _r = 2;
let ze = class extends Ue {
  constructor(e, t, n, r, s, o, a, c, g, l, u, h) {
    super();
    /**
     * It is used to distinguish whether the user has actively moved the cursor in the editor, mainly through mouse clicks.
     */
    P(this, "_cursorChange", k.InitialState);
    /** If the corresponding unit is active and prepared for editing. */
    P(this, "_isUnitEditing", !1);
    P(this, "_d");
    this._renderContext = e, this._layoutService = t, this._undoRedoService = n, this._contextService = r, this._instanceSrv = s, this._renderManagerService = o, this._editorBridgeService = a, this._cellEditorManagerService = c, this._textSelectionManagerService = g, this._commandService = l, this._localService = u, this._editorService = h, this.disposeWithMe(this._instanceSrv.getCurrentTypeOfUnit$(x.UNIVER_SLIDE).subscribe((v) => {
      v && v.getUnitId() === this._renderContext.unitId ? this._d = this._init() : (this._disposeCurrent(), this._isUnitEditing && (this._handleEditorInvisible({
        visible: !1,
        eventType: q.Keyboard,
        keycode: y.ESC,
        unitId: this._renderContext.unitId
      }), this._isUnitEditing = !1));
    })), this._initEditorVisibilityListener();
  }
  dispose() {
    super.dispose(), this._disposeCurrent();
  }
  _disposeCurrent() {
    var e;
    (e = this._d) == null || e.dispose(), this._d = null;
  }
  _init() {
    const e = new Ye();
    return this._subscribeToCurrentCell(e), this._initialKeyboardListener(e), this._initialCursorSync(e), this._listenEditorFocus(e), this._commandExecutedListener(e), setTimeout(() => {
      this._cursorStateListener(e);
    }, 1e3), e;
  }
  _initEditorVisibilityListener() {
    this.disposeWithMe(this._editorBridgeService.visible$.subscribe((e) => {
      e.visible ? (this._isUnitEditing = !0, this._handleEditorVisible(e)) : this._isUnitEditing && (this._handleEditorInvisible(e), this._isUnitEditing = !1);
    }));
  }
  _listenEditorFocus(e) {
    const t = this._getEditorObject();
    t && e.add(t.document.onPointerDown$.subscribeEvent(() => {
    }));
  }
  _getEditorSkeleton(e) {
    var t;
    return (t = this._renderManagerService.getRenderById(e)) == null ? void 0 : t.with(gt).getSkeleton();
  }
  _getEditorViewModel(e) {
    var t;
    return (t = this._renderManagerService.getRenderById(e)) == null ? void 0 : t.with(gt).getViewModel();
  }
  _initialCursorSync(e) {
    e.add(this._cellEditorManagerService.focus$.pipe(Gn((t) => !!t)).subscribe(() => {
      var n;
      const t = (n = this._renderManagerService.getCurrentTypeOfRenderer(x.UNIVER_DOC)) == null ? void 0 : n.with(ht);
      t && t.sync();
    }));
  }
  /**
   * Set editorUnitId to curr doc.
   * @param d DisposableCollection
   */
  _subscribeToCurrentCell(e) {
    e.add(this._editorBridgeService.currentEditRectState$.subscribe((t) => {
      var h;
      if (t == null || this._contextService.getContextValue(Ot) || this._contextService.getContextValue(jt))
        return;
      const {
        position: { startX: n, endX: r },
        documentLayoutObject: { textRotation: s, wrapStrategy: o, documentModel: a },
        scaleX: c,
        editorUnitId: g
      } = t, { vertexAngle: l } = at(s);
      a.updateDocumentId(g), o === it.WRAP && l === 0 && a.updateDocumentDataPageSize((r - n) / c), this._instanceSrv.changeDoc(g, a), this._contextService.setContextValue(bn, !0), this._textSelectionManagerService.replaceTextRanges([{
        startOffset: 0,
        endOffset: 0
      }]);
      const u = (h = this._renderManagerService.getCurrentTypeOfRenderer(x.UNIVER_DOC)) == null ? void 0 : h.with(ht);
      u && u.activate(St, St);
    }));
  }
  /**
   * Set size and pos of editing area.
   * @param positionFromEditRectState
   * @param canvasOffset
   * @param documentSkeleton
   * @param documentLayoutObject
   * @param scaleX
   * @param scaleY
   */
  _fitTextSize(e, t, n, r, s = 1, o = 1) {
    const { startX: a, startY: c, endX: g, endY: l } = e, u = r.documentModel;
    if (u == null)
      return;
    const { actualWidth: h, actualHeight: v } = this._predictingSize(
      e,
      t,
      n,
      r,
      s,
      o
    ), { verticalAlign: _, paddingData: C, fill: p } = r;
    let f = g - a, I = l - c;
    if (f < h && (f = h), I < v)
      I = v, u.updateDocumentDataMargin(C);
    else {
      let b = 0;
      _ === je.MIDDLE ? b = (I - v) / 2 / o : _ === je.TOP ? b = C.t || 0 : b = (I - v) / o - (C.b || 0), b = b < (C.t || 0) ? C.t || 0 : b, u.updateDocumentDataMargin({
        t: b
      });
    }
    n.calculate(), this._editAreaProcessing(f, I, e, t, p, s, o);
  }
  /**
   * Mainly used to pre-calculate the width of the editor,
   * to determine whether it needs to be automatically widened.
   */
  _predictingSize(e, t, n, r, s = 1, o = 1) {
    const { startX: a, endX: c } = e, { textRotation: g, wrapStrategy: l } = r, u = r.documentModel, { vertexAngle: h } = at(g), v = document.body.clientWidth;
    if (l === it.WRAP && h === 0) {
      const { actualWidth: p, actualHeight: f } = n.getActualSize();
      return {
        actualWidth: p * s,
        actualHeight: f * o
      };
    }
    u == null || u.updateDocumentDataPageSize((v - a - t.left) / s), n.calculate();
    const _ = n.getActualSize();
    let C = c - a;
    return C < _.actualWidth * s + It * s && (C = _.actualWidth * s + It * s), u == null || u.updateDocumentDataPageSize(C / s), u == null || u.updateDocumentRenderConfig({
      horizontalAlign: yt.UNSPECIFIED,
      cellValueType: void 0
    }), {
      actualWidth: C,
      actualHeight: _.actualHeight * o
    };
  }
  /**
   * control the size of editing area
   */
  // eslint-disable-next-line max-lines-per-function
  _editAreaProcessing(e, t, n, r, s, o = 1, a = 1) {
    var S;
    const c = this._getEditorObject();
    if (c == null)
      return;
    function g(le) {
      return Number.parseInt(le.replace("px", ""));
    }
    const u = this._renderContext.engine.getCanvasElement(), h = u.getBoundingClientRect(), v = g(u.style.width), { top: _, left: C, width: p } = h, f = p / v;
    let { startX: I, startY: b } = n;
    I += r.left, b += r.top;
    const { document: E, scene: w, engine: V } = c, R = w.getViewport(pt.VIEW_MAIN), G = document.body.clientHeight - b - r.top - _r * 2, ce = document.body.clientWidth - I - r.left;
    let F = t, z = R == null ? void 0 : R.getScrollBar();
    F > G ? (F = G, z == null ? R && new wt(R, { enableHorizontal: !1, barSize: 8 }) : R == null || R.resetCanvasSizeAndUpdateScroll()) : (z = null, (S = R == null ? void 0 : R.getScrollBar()) == null || S.dispose()), e += (z == null ? void 0 : z.barSize) || 0, e = Math.min(e, ce), I -= re, b -= re, this._addBackground(w, e / o, t / a, s);
    const { scaleX: ee, scaleY: te } = w.getPrecisionScale();
    w.transformByState({
      width: e * f / o,
      height: t * f / a,
      scaleX: o * f,
      scaleY: a * f
    }), E.resize(e / o, t / a), setTimeout(() => {
      V.resizeBySize(
        ct(e, ee),
        ct(F, te)
      );
    }, 0);
    const Se = this._layoutService.getContentElement().getBoundingClientRect(), Ie = u.getBoundingClientRect();
    I = I * f + (Ie.left - Se.left), b = b * f + (Ie.top - Se.top), this._cellEditorManagerService.setState({
      startX: I,
      startY: b,
      endX: e * f + I,
      endY: F * f + b,
      show: !0
    });
  }
  /**
   * Since the document does not support cell background color, an additional rect needs to be added.
   */
  _addBackground(e, t, n, r) {
    const s = "_backgroundRectHelperColor_", o = e.getObject(s);
    o == null && r == null || (o == null ? e.addObjects(
      [
        new Nt(s, {
          width: t,
          height: n,
          fill: r,
          evented: !1
        })
      ],
      kn
    ) : r == null ? o.dispose() : (o.setProps({
      fill: r
    }), o.transformByState({
      width: t,
      height: n
    })));
  }
  /**
   * Show input area, resize input area and then place input to right place.
   * @TODO why do resize in show input area?
   * @param param
   */
  // handleVisible is the 2nd part of editing.
  // first part: startEditing --> _updateEditor
  // 2nd part: startEditing --> changeVisible --> slide-editor-bridge.render-controller.ts@changeVisible --> _editorBridgeService.changeVisible
  _handleEditorVisible(e) {
    var p, f;
    const { eventType: t } = e;
    this._cursorChange = [q.PointerDown, q.Dblclick].includes(t) ? k.CursorChange : k.StartEditor;
    const n = this._editorBridgeService.getEditRectState();
    if (n == null)
      return;
    const {
      position: r,
      documentLayoutObject: s,
      slideCardOffset: o,
      scaleX: a,
      scaleY: c,
      editorUnitId: g,
      unitId: l
    } = n, u = this._getEditorObject();
    if (u == null)
      return;
    const { scene: h } = u;
    this._contextService.setContextValue(Oe, !0);
    const { documentModel: v } = s, _ = this._getEditorSkeleton(g);
    if (!_ || !v)
      return;
    this._fitTextSize(r, o, _, s, a, c);
    const C = v.getBody().dataStream.length - 2 || 0;
    (p = h.getViewport(pt.VIEW_MAIN)) == null || p.scrollToViewportPos({
      viewportScrollX: Number.POSITIVE_INFINITY
    }), this._textSelectionManagerService.replaceTextRanges([
      {
        startOffset: C,
        endOffset: C
      }
    ]), (f = this._renderManagerService.getRenderById(l)) == null || f.scene.resetCursor();
  }
  _resetBodyStyle(e, t = !1) {
    e.dataStream = Tt, e.textRuns != null && (e.textRuns.length === 1 && !t ? (e.textRuns[0].st = 0, e.textRuns[0].ed = 1) : e.textRuns = void 0), e.paragraphs != null && (e.paragraphs.length === 1 ? e.paragraphs[0].startIndex = 0 : e.paragraphs = [
      {
        startIndex: 0
      }
    ]), e.sectionBreaks != null && (e.sectionBreaks = void 0), e.tables != null && (e.tables = void 0), e.customRanges != null && (e.customRanges = void 0), e.customBlocks != null && (e.customBlocks = void 0);
  }
  /**
   * Should activate the editor when the user inputs text.
   * @param d DisposableCollection
   */
  _initialKeyboardListener(e) {
  }
  _showEditorByKeyboard(e) {
  }
  _commandExecutedListener(e) {
    const t = [Me.id], n = [Yn.id];
    e.add(this._commandService.onCommandExecuted((r) => {
      this._editorService.getFocusId() === ae && (t.includes(r.id) && this._moveCursorCmdHandler(r), n.includes(r.id) && this._editorBridgeService.isVisible() && this._editingChangedHandler());
    }));
  }
  _moveCursorCmdHandler(e) {
    const t = e.params, { keycode: n, isShift: r } = t;
    n != null && this._cursorChange === k.CursorChange ? this._moveInEditor(n, r) : this._editorBridgeService.changeVisible(t);
  }
  _editingChangedHandler() {
    const e = this._editorBridgeService.getEditorRect();
    if (!e)
      return;
    const t = e.richTextObj;
    t.refreshDocumentByDocData(), t.resizeToContentSize();
    const { unitId: n } = this._renderContext;
    this._handleEditorVisible({ visible: !0, eventType: 3, unitId: n });
  }
  _getEditorObject() {
    return Sr(this._editorBridgeService.getCurrentEditorId(), this._renderManagerService);
  }
  async _handleEditorInvisible(e) {
    const { keycode: t } = e;
    if (this._cursorChange = k.InitialState, this._exitInput(e), this._editorBridgeService.getEditRectState() == null)
      return;
    if (this._editorBridgeService.getEditorDirty() === !1) {
      this._moveCursor(t);
      return;
    }
    this._moveCursor(t);
  }
  _exitInput(e) {
    this._contextService.setContextValue(Oe, !1), this._cellEditorManagerService.setState({
      show: e.visible
    });
    const t = this._editorBridgeService.getCurrentEditorId();
    t != null && this._undoRedoService.clearUndoRedo(t);
  }
  _moveCursor(e) {
    if (e != null)
      switch (B.LEFT, e) {
        case y.ENTER:
          B.DOWN;
          break;
        case y.TAB:
          B.RIGHT;
          break;
        case y.ARROW_DOWN:
          B.DOWN;
          break;
        case y.ARROW_UP:
          B.UP;
          break;
        case y.ARROW_LEFT:
          B.LEFT;
          break;
        case y.ARROW_RIGHT:
          B.RIGHT;
          break;
      }
  }
  /**
   * The user's operations follow the sequence of opening the editor and then moving the cursor.
   * The logic here predicts the user's first cursor movement behavior based on this rule
   */
  _cursorStateListener(e) {
    const t = this._getEditorObject();
    if (!t)
      return;
    const { document: n } = t;
    e.add(Et(n.onPointerDown$.subscribeEvent(() => {
      this._cursorChange === k.StartEditor && (this._cursorChange = k.CursorChange);
    })));
  }
  // TODO: @JOCS, is it necessary to move these commands MoveSelectionOperation\MoveCursorOperation to shortcut? and use multi-commands?
  _moveInEditor(e, t) {
    let n = B.LEFT;
    e === y.ARROW_DOWN ? n = B.DOWN : e === y.ARROW_UP ? n = B.UP : e === y.ARROW_RIGHT && (n = B.RIGHT), t ? this._commandService.executeCommand(Zn.id, {
      direction: n
    }) : this._commandService.executeCommand(qn.id, {
      direction: n
    });
  }
};
ze = vr([
  L(1, Un),
  L(2, xn),
  L(3, Ce),
  L(4, j),
  L(5, Y),
  L(6, Qe),
  L(7, et),
  L(8, X(Xn)),
  L(9, U),
  L(10, X(Q)),
  L(11, Ze)
], ze);
function Sr(i, e) {
  if (i == null)
    return;
  const t = e.getRenderById(i);
  if (t == null)
    return;
  const { mainComponent: n, scene: r, engine: s, components: o } = t, a = n, c = o.get(Jn.BACKGROUND);
  return {
    document: a,
    docBackground: c,
    scene: r,
    engine: s
  };
}
var Ir = Object.defineProperty, Pr = Object.getOwnPropertyDescriptor, br = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? Pr(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && Ir(e, t, r), r;
}, Ae = (i, e) => (t, n) => e(t, n, i);
let We = class extends fe {
  constructor(e, t, n, r) {
    super();
    /**
     * It is used to distinguish whether the user has actively moved the cursor in the editor, mainly through mouse clicks.
     */
    // private _cursorChange: CursorChange = CursorChange.InitialState;
    /** If the corresponding unit is active and prepared for editing. */
    // private _isUnitEditing = false;
    P(this, "setSlideTextEditor$", new Bt());
    P(this, "_curRichText", null);
    P(this, "_d");
    this._renderContext = e, this._instanceSrv = t, this._commandService = n, this._editorBridgeService = r, this.disposeWithMe(this._instanceSrv.getCurrentTypeOfUnit$(x.UNIVER_SLIDE).subscribe((s) => {
      s && s.getUnitId() === this._renderContext.unitId ? this._d = this._init() : this._disposeCurrent();
    }));
  }
  _init() {
    const e = new Ye();
    return this._initEventListener(e), e;
  }
  _disposeCurrent() {
    var e;
    (e = this._d) == null || e.dispose(), this._d = null;
  }
  _setEditorRect(e, t) {
    this._curRichText = t;
    const { scene: n, engine: r } = this._renderContext, s = this._renderContext.unitId, o = {
      scene: n,
      engine: r,
      unitId: s,
      pageId: e,
      richTextObj: t
    };
    this._editorBridgeService.setEditorRect(o);
  }
  _initEventListener(e) {
    const t = (o) => {
      const a = o.getTransformer();
      a && (e.add(a.clearControl$.subscribe(() => {
        this.setEditorVisible(!1), this.pickOtherObjects();
      })), e.add(a.createControl$.subscribe(() => {
        this.setEditorVisible(!1);
      })), e.add(o.onDblclick$.subscribeEvent(() => {
        a.clearControls();
        const g = a.getSelectedObjectMap().values().next().value;
        g && (g.objectType !== he.RICH_TEXT ? this.pickOtherObjects() : this.startEditing(o.sceneKey, g));
      })), e.add(this._instanceSrv.focused$.subscribe((c) => {
        this.endEditing();
      })));
    }, { mainComponent: n } = this._renderContext;
    n.subSceneChanged$.subscribeEvent((o) => {
      t(o);
    });
    const s = Array.from(n.getSubScenes().values());
    for (let o = 0; o < s.length; o++) {
      const a = s[o];
      t(a);
    }
  }
  pickOtherObjects() {
    this.endEditing();
  }
  /**
   * invoked when picking other object.
   *
   * save editing state to curr richText.
   */
  endEditing() {
    var s;
    if (!this._curRichText) return;
    this.setEditorVisible(!1);
    const e = this._curRichText;
    if (!this._instanceSrv.getCurrentUnitForType(x.UNIVER_SLIDE)) return !1;
    e.refreshDocumentByDocData(), e.resizeToContentSize(), this._editorBridgeService.endEditing$.next(e);
    const n = {
      bl: 1,
      fs: e.fs,
      text: e.text
    }, r = (s = e.documentData.body) == null ? void 0 : s.textRuns;
    if (r && r.length) {
      const a = r[0].ts;
      n.cl = a == null ? void 0 : a.cl;
    }
    this._commandService.executeCommand(K.id, {
      unitId: this._renderContext.unitId,
      oKey: e == null ? void 0 : e.oKey,
      props: {
        richText: n
      }
    }), this._curRichText = null;
  }
  /**
   * TODO calling twice ？？？？
   * editingParam derives from RichText object.
   *
   * TODO @lumixraku need scale param
   * @param target
   */
  startEditing(e, t) {
    this._setEditorRect(e, t), this.setEditorVisible(!0);
  }
  setEditorVisible(e) {
    var n, r;
    e ? (n = this._curRichText) == null || n.hide() : (r = this._curRichText) == null || r.show();
    const { unitId: t } = this._renderContext;
    this._editorBridgeService.changeVisible({ visible: e, eventType: q.PointerDown, unitId: t });
  }
};
We = br([
  Ae(1, j),
  Ae(2, U),
  Ae(3, Qe)
], We);
const Jt = {
  id: "slide.operation.activate-slide",
  type: D.OPERATION,
  handler: (i, e) => {
    var g, l;
    const t = e.unitId, n = i.get(M), s = i.get(j).getUnit(t), o = (g = s == null ? void 0 : s.getActivePage()) == null ? void 0 : g.id;
    if (!o) return !1;
    const a = n.getRenderUnitByPageId(o, t);
    if (!a) return !1;
    const c = (l = a.scene) == null ? void 0 : l.getTransformer();
    return c && c.clearControls(), n.activePage(e.id, t), !0;
  }
}, Qt = {
  id: "slide.operation.append-slide",
  type: D.OPERATION,
  handler: (i, e) => {
    const t = e.unitId;
    return i.get(j).getUnit(t) ? (i.get(M).appendPage(t), !0) : !1;
  }
}, tt = {
  id: "slide.command.insert-float-image",
  type: D.COMMAND,
  handler: async (i, e) => {
    var V;
    const t = i.get(j), n = (V = t.getCurrentUnitForType(x.UNIVER_SLIDE)) == null ? void 0 : V.getUnitId();
    if (!n) return !1;
    const s = await i.get(Vn).openFile({
      multiple: !0,
      accept: ei.map((R) => `.${R.replace("image/", "")}`).join(",")
    });
    if (s.length !== 1) return !1;
    const a = await i.get(ti).saveImage(s[0]);
    if (!a) return !1;
    const { imageId: c, imageSourceType: g, source: l, base64Cache: u } = a, { width: h, height: v, image: _ } = await ni(u || ""), C = t.getUnit(n);
    if (!C) return !1;
    const p = C.getActivePage(), f = Object.values(p.pageElements), I = f != null && f.length ? Math.max(...f.map((R) => R.zIndex)) : 20, b = {
      id: c,
      zIndex: I + 1,
      left: 0,
      top: 0,
      width: h,
      height: v,
      title: "",
      description: "",
      type: Xe.IMAGE,
      image: {
        imageProperties: {
          contentUrl: u,
          imageSourceType: g,
          source: l,
          base64Cache: u,
          image: _
        }
      }
    };
    p.pageElements[c] = b, C.updatePage(p.id, p);
    const E = i.get(M), w = E.createObjectToPage(b, p.id, n);
    return w && E.setObjectActiveByPage(w, p.id, n), !0;
  }
}, nt = {
  id: "slide.command.add-text",
  type: D.COMMAND,
  handler: async (i) => {
    var r;
    const e = i.get(U), n = (r = i.get(j).getFocusedUnit()) == null ? void 0 : r.getUnitId();
    return await e.executeCommand(en.id, { unitId: n });
  }
}, en = {
  id: "slide.operation.add-text",
  type: D.OPERATION,
  handler: async (i, e) => {
    const t = e.unitId, n = xt(6), r = 220, s = 40, o = 230, a = 142, c = (e == null ? void 0 : e.text) || "A New Text", l = i.get(j).getUnit(t);
    if (!l) return !1;
    const u = l.getActivePage(), h = Object.values(u.pageElements), v = h != null && h.length ? Math.max(...h.map((f) => f.zIndex)) : 21, _ = {
      id: n,
      zIndex: v + 1,
      left: o,
      top: a,
      width: r,
      height: s,
      title: "text",
      description: "",
      type: Xe.TEXT,
      richText: {
        text: c,
        fs: 30,
        cl: {
          rgb: "rgb(51, 51, 51)"
        },
        bl: 1
      }
    };
    u.pageElements[n] = _, l.updatePage(u.id, u);
    const C = i.get(M), p = C.createObjectToPage(_, u.id, t);
    return p && C.setObjectActiveByPage(p, u.id, t), !0;
  }
}, tn = {
  id: "slide.operation.set-slide-page-thumb",
  type: D.OPERATION,
  handler: (i, e) => (i.get(M).createThumbs(e.unitId), !0)
}, xr = "univer-image-popup-menu", Er = "univer-image-popup-menu-item", Rr = "univer-image-popup-menu-item-icon", Or = "univer-image-popup-menu-item-title", jr = "univer-image-popup-menu-item-hide", yr = "univer-btn-container", Tr = "univer-btn-container-expand", ge = {
  imagePopupMenu: xr,
  imagePopupMenuItem: Er,
  imagePopupMenuItemIcon: Rr,
  imagePopupMenuItemTitle: Or,
  imagePopupMenuItemHide: jr,
  btnContainer: yr,
  btnContainerExpand: Tr
}, Mr = (i) => {
  var _, C;
  const e = (C = (_ = i.popup) == null ? void 0 : _.extraProps) == null ? void 0 : C.menuItems;
  if (!e)
    return null;
  const t = O(U), n = O(Q), [r, s] = W(!1), [o, a] = W(!1), c = () => {
    a(!0);
  }, g = () => {
    a(!1);
  }, l = (p) => {
    s(p);
  }, u = (p) => {
    t.executeCommand(p.commandId, p.commandParams), s(!1);
  }, h = r || o, v = e.filter((p) => !p.disable);
  return /* @__PURE__ */ d.jsx(
    "div",
    {
      onMouseEnter: c,
      onMouseLeave: g,
      children: /* @__PURE__ */ d.jsx(
        Vt,
        {
          placement: "bottomLeft",
          trigger: ["click"],
          overlay: /* @__PURE__ */ d.jsx("ul", { className: ge.imagePopupMenu, children: v.map((p) => /* @__PURE__ */ d.jsx(
            "li",
            {
              onClick: () => u(p),
              className: ge.imagePopupMenuItem,
              children: /* @__PURE__ */ d.jsx("span", { className: ge.imagePopupMenuItemTitle, children: n.t(p.label) })
            },
            p.index
          )) }),
          visible: r,
          onVisibleChange: l,
          children: /* @__PURE__ */ d.jsxs(
            "div",
            {
              className: T(ge.btnContainer, {
                [ge.btnContainerExpand]: r
              }),
              children: [
                /* @__PURE__ */ d.jsx(
                  Ft,
                  {
                    style: { color: "#35322B" },
                    extend: { colorChannel1: "rgb(var(--green-700, #409f11))" }
                  }
                ),
                h && /* @__PURE__ */ d.jsx(qe, { style: { color: "#CCCCCC", fontSize: "8px", marginLeft: "8px" } })
              ]
            }
          )
        }
      )
    }
  );
}, wr = "univer-slide-bar", Nr = "univer-slide-bar-content", Dr = "univer-slide-bar-content-header", Ur = "univer-slide-bar-item", Vr = "univer-slide-bar-box", Br = "univer-slide-bar-item-active", Lr = "univer-slide-add-button", ne = {
  slideBar: wr,
  slideBarContent: Nr,
  slideBarContentHeader: Dr,
  slideBarItem: Ur,
  slideBarBox: Vr,
  slideBarItemActive: Br,
  slideAddButton: Lr
};
function Ar() {
  var C, p;
  const i = O(j), e = O(U), t = O(Y), n = O(Q), r = Ut(null), s = i.getCurrentUnitForType(x.UNIVER_SLIDE), o = s == null ? void 0 : s.getPages(), a = s == null ? void 0 : s.getPageOrder();
  if (!o || !a)
    return null;
  const c = a.map((f) => o[f]), [g, l] = W([]), [u, h] = W((p = (C = s == null ? void 0 : s.getActivePage()) == null ? void 0 : C.id) != null ? p : null);
  oe(() => {
    l(c.map((f) => ke.createRef()));
  }, [c.length]), oe(() => {
    const f = s == null ? void 0 : s.activePage$.subscribe((I) => {
      var E;
      const b = (E = I == null ? void 0 : I.id) != null ? E : null;
      b && h(b);
    });
    return () => {
      f == null || f.unsubscribe();
    };
  }, []), oe(() => {
    g.forEach((f, I) => {
      var b;
      if (f.current) {
        const E = c[I];
        (b = t.getRenderById(E.id)) == null || b.engine.setContainer(f.current);
      }
    }), g.length > 0 && e.syncExecuteCommand(tn.id, { unitId: s == null ? void 0 : s.getUnitId() });
  }, [g, c, t, e, s]);
  const v = ut((f) => {
    e.syncExecuteCommand(Jt.id, { id: f, unitId: s == null ? void 0 : s.getUnitId() });
  }, [e, s]), _ = ut(() => {
    e.syncExecuteCommand(Qt.id, { unitId: s == null ? void 0 : s.getUnitId() });
  }, [e, s]);
  return /* @__PURE__ */ d.jsx("aside", { className: ne.slideBar, ref: r, children: /* @__PURE__ */ d.jsx(Kn, { children: /* @__PURE__ */ d.jsxs("div", { className: ne.slideBarContent, children: [
    /* @__PURE__ */ d.jsx("header", { className: ne.slideBarContentHeader, children: /* @__PURE__ */ d.jsx("a", { onClick: _, children: n.t("slide.append") }) }),
    c.map((f, I) => /* @__PURE__ */ d.jsxs(
      "div",
      {
        className: T(ne.slideBarItem, {
          [ne.slideBarItemActive]: f.id === u
        }),
        onClick: () => v(f.id),
        children: [
          /* @__PURE__ */ d.jsx("span", { children: I + 1 }),
          /* @__PURE__ */ d.jsx("div", { ref: g[I], className: ne.slideBarBox })
        ]
      },
      f.id
    ))
  ] }) }) });
}
const $r = "univer-slide-editor-container", Hr = "univer-editor-input", Fr = {
  slideEditorContainer: $r,
  editorInput: Hr
}, we = -1e3, Pt = {
  width: 0,
  height: 0,
  top: we,
  left: we
}, zr = () => {
  const [i, e] = W({
    ...Pt
  }), t = O(et), n = O(Ze), r = O(Ce), s = Bn(
    () => r.subscribeContextValue$(lt),
    !1,
    void 0,
    [r, lt]
  );
  return `${Tt}`, En.UNSPECIFIED, oe(() => {
    t.state$.subscribe((o) => {
      if (o == null)
        return;
      const {
        startX: a = we,
        startY: c = we,
        endX: g = 0,
        endY: l = 0,
        show: u = !1
      } = o;
      if (!u)
        e({
          ...Pt
        });
      else {
        e({
          width: g - a - re + 2,
          height: l - c - re + 2,
          left: a + re,
          top: c + re
        });
        const h = n.getEditor(ae);
        if (h == null)
          return;
        const { left: v, top: _, width: C, height: p } = h.getBoundingClientRect();
        t.setRect({ left: v, top: _, width: C, height: p });
      }
    });
  }, []), oe(() => {
    s || t.setFocus(!0);
  }, [s, i]), /* @__PURE__ */ d.jsx(
    "div",
    {
      className: Fr.slideEditorContainer,
      style: {
        left: i.left,
        top: i.top,
        width: i.width,
        height: i.height
      }
    }
  );
}, nn = "addition-and-subtraction-single", rn = "slide.menu.image";
function Wr(i) {
  return {
    id: rn,
    type: ve.SUBITEMS,
    icon: nn,
    tooltip: "slide.image.insert.title",
    hidden$: _e(i, x.UNIVER_SLIDE)
  };
}
function Kr(i) {
  return {
    id: tt.id,
    title: "slide.image.insert.float",
    type: ve.BUTTON,
    hidden$: _e(i, x.UNIVER_SLIDE)
  };
}
const sn = "graph-single", on = "slide.menu.shape";
function Gr(i) {
  return {
    id: on,
    type: ve.SUBITEMS,
    icon: sn,
    tooltip: "slide.shape.insert.title",
    hidden$: _e(i, x.UNIVER_SLIDE)
    // disabled$: getCurrentRangeDisable$(accessor, { workbookTypes: [WorkbookEditablePermission], worksheetTypes: [WorksheetEditPermission], rangeTypes: [RangeProtectionPermissionEditPoint] }),
  };
}
function Xr(i) {
  return {
    id: Je.id,
    title: "slide.shape.insert.rectangle",
    type: ve.BUTTON,
    hidden$: _e(i, x.UNIVER_SLIDE)
  };
}
const an = "text-single";
function Yr(i) {
  return {
    id: nt.id,
    type: ve.BUTTON,
    icon: an,
    tooltip: "slide.text.insert.title",
    hidden$: _e(i, x.UNIVER_SLIDE)
  };
}
const kr = {
  [Ln.FORMAT]: {
    [nt.id]: {
      order: 0,
      menuItemFactory: Yr
    },
    [rn]: {
      order: 0,
      menuItemFactory: Wr,
      [tt.id]: {
        order: 0,
        menuItemFactory: Kr
      }
    },
    [on]: {
      order: 0,
      menuItemFactory: Gr,
      [Je.id]: {
        order: 0,
        menuItemFactory: Xr
      }
    }
  }
};
function Ke(i) {
  return i.getContextValue(Mt) && i.getContextValue(Oe);
}
function Zr(i) {
  return i.getContextValue(Rn) && i.getContextValue(Mt);
}
const cn = [
  y.ARROW_DOWN,
  y.ARROW_UP,
  y.ARROW_LEFT,
  y.ARROW_RIGHT
];
[y.ENTER, y.TAB, ...cn];
function qr() {
  const i = [];
  for (const e of cn)
    i.push({
      id: Me.id,
      binding: e,
      preconditions: (t) => Ke(t),
      staticParameters: {
        visible: !1,
        eventType: q.Keyboard,
        keycode: e,
        isShift: !1
      }
    }), i.push({
      id: Me.id,
      binding: e | An.SHIFT,
      preconditions: (t) => Ke(t),
      staticParameters: {
        visible: !1,
        eventType: q.Keyboard,
        keycode: e,
        isShift: !0
      }
    });
  return i;
}
const Jr = {
  id: Qn.id,
  preconditions: (i) => Ke(i) || Zr(i),
  binding: y.BACKSPACE
};
var Qr = Object.defineProperty, es = Object.getOwnPropertyDescriptor, ts = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? es(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && Qr(e, t, r), r;
}, ie = (i, e) => (t, n) => e(t, n, i);
let Ne = class extends Ue {
  constructor(i, e, t, n, r, s) {
    super(), this._injector = i, this._menuManagerService = e, this._componentManager = t, this._uiPartsService = n, this._commandService = r, this._shortcutService = s, this._initCommands(), this._initCustomComponents(), this._initUIComponents(), this._initMenus(), this._initShortcuts();
  }
  _initMenus() {
    this._menuManagerService.mergeMenu(kr);
  }
  _initCustomComponents() {
    const i = this._componentManager;
    this.disposeWithMe(i.register(nn, Ht)), this.disposeWithMe(i.register(an, Yt)), this.disposeWithMe(i.register(sn, Wt)), this.disposeWithMe(i.register(qt, Mr)), this.disposeWithMe(i.register(Re, rr));
  }
  _initCommands() {
    [
      Qt,
      Jt,
      tn,
      tt,
      en,
      nt,
      Zt,
      Je,
      He,
      Lt,
      K,
      // commands for editor
      Me
    ].forEach((i) => this.disposeWithMe(this._commandService.registerCommand(i)));
  }
  _initUIComponents() {
    this.disposeWithMe(
      this._uiPartsService.registerComponent(dt.LEFT_SIDEBAR, () => rt(Ar, this._injector))
    ), this.disposeWithMe(
      this._uiPartsService.registerComponent(dt.CONTENT, () => rt(zr, this._injector))
    );
  }
  _initShortcuts() {
    [
      Jr,
      ...qr()
    ].forEach((i) => {
      this.disposeWithMe(this._shortcutService.registerShortcut(i));
    });
  }
};
Ne = ts([
  ie(0, X(Ge)),
  ie(1, Hn),
  ie(2, X($n)),
  ie(3, Fn),
  ie(4, U),
  ie(5, zn)
], Ne);
var ns = Object.defineProperty, is = Object.getOwnPropertyDescriptor, rs = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? is(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && ns(e, t, r), r;
}, $e = (i, e) => (t, n) => e(t, n, i);
let De = class extends fe {
  // private _skeletonChangeMutations = new Set<string>();
  constructor(i, e, t) {
    super(), this._contextService = i, this._instanceSrv = e, this._renderManagerService = t, Promise.resolve().then(() => this._init());
  }
  _init() {
    this._initSlideDataListener(), this._initContextListener();
  }
  _initSlideDataListener() {
    this._instanceSrv.getTypeOfUnitAdded$(x.UNIVER_SLIDE).pipe(mt(this.dispose$)).subscribe((i) => {
      this._createRenderer(i == null ? void 0 : i.getUnitId());
    }), this._instanceSrv.getAllUnitsForType(x.UNIVER_SLIDE).forEach((i) => {
      this._createRenderer(i.getUnitId());
    }), this._instanceSrv.getTypeOfUnitDisposed$(x.UNIVER_SLIDE).pipe(mt(this.dispose$)).subscribe((i) => this._disposeRenderer(i));
  }
  _createRenderer(i) {
    i == null || this._instanceSrv.getUnit(i, x.UNIVER_SLIDE) == null || (this._renderManagerService.createRender(i), this._renderManagerService.setCurrent(i));
  }
  _disposeRenderer(i) {
    const e = i.getUnitId();
    this._renderManagerService.removeRender(e);
  }
  _initContextListener() {
  }
};
De = rs([
  $e(0, Ce),
  $e(1, j),
  $e(2, Y)
], De);
var ss = Object.defineProperty, os = Object.getOwnPropertyDescriptor, as = (i, e, t, n) => {
  for (var r = n > 1 ? void 0 : n ? os(e, t) : e, s = i.length - 1, o; s >= 0; s--)
    (o = i[s]) && (r = (n ? o(e, t, r) : o(r)) || r);
  return n && r && ss(e, t, r), r;
}, xe = (i, e) => (t, n) => e(t, n, i);
const cs = "SLIDE_UI";
var Ee;
let bt = (Ee = class extends On {
  constructor(i = ft, e, t, n, r) {
    super(), this._config = i, this._injector = e, this._renderManagerService = t, this._univerInstanceService = n, this._configService = r;
    const { menu: s, ...o } = jn(
      {},
      ft,
      this._config
    );
    s && this._configService.setConfig("menu", s, { merge: !0 }), this._configService.setConfig(di, o);
  }
  onStarting() {
    st([
      [De],
      [Qe, { useClass: Fe }],
      // used by SlideUIController --> EditorContainer
      [et, { useClass: pr }],
      [ye]
    ], this._config.override).forEach((i) => this._injector.add(i));
  }
  onReady() {
    [
      // SlideRenderService will be init in ready stage, and then calling RenderManagerService@createRender --> init all deps in this rendering register block.
      [pe]
    ].forEach((i) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(x.UNIVER_SLIDE, i));
    }), st([
      [M],
      // cannot register in _renderManagerService now.
      // [ISlideEditorBridgeService, { useClass: SlideEditorBridgeService }],
      // // used by SlideUIController --> EditorContainer
      // [ISlideEditorManagerService, { useClass: SlideEditorManagerService }],
      // SlidesUIController controller should be registered in Ready stage.
      // SlidesUIController controller would add a new RenderUnit (__INTERNAL_EDITOR__DOCS_NORMAL)
      [Ne],
      // editor service was create in renderManagerService
      [pe],
      [Te]
    ], this._config.override).forEach((i) => {
      this._injector.add(i);
    }), this._injector.get(M), this._injector.get(De);
  }
  onRendered() {
    [
      // need slideEditorBridgeService
      // need TextSelectionRenderService which init by EditorContainer
      [We],
      [ze]
    ].forEach((i) => {
      this.disposeWithMe(this._renderManagerService.registerRenderModule(x.UNIVER_SLIDE, i));
    }), this._markSlideAsFocused(), this._injector.get(Ne);
  }
  onSteady() {
    this._injector.get(Te);
  }
  _markSlideAsFocused() {
    const i = this._univerInstanceService;
    try {
      const e = i.getCurrentUnitForType(x.UNIVER_SLIDE);
      i.focusUnit(e.getUnitId());
    } catch {
    }
  }
}, P(Ee, "pluginName", cs), P(Ee, "type", x.UNIVER_SLIDE), Ee);
bt = as([
  xe(1, X(Ge)),
  xe(2, Y),
  xe(3, j),
  xe(4, yn)
], bt);
export {
  Jt as ActivateSlidePageOperation,
  Qt as AppendSlideOperation,
  M as CanvasView,
  Lt as DeleteSlideElementOperation,
  Qe as ISlideEditorBridgeService,
  tt as InsertSlideFloatImageCommand,
  Je as InsertSlideShapeRectangleCommand,
  Zt as InsertSlideShapeRectangleOperation,
  on as SHAPE_MENU_ID,
  rn as SLIDES_IMAGE_MENU_ID,
  ae as SLIDE_EDITOR_ID,
  tn as SetSlidePageThumbOperation,
  nt as SlideAddTextCommand,
  en as SlideAddTextOperation,
  ye as SlideCanvasPopMangerService,
  zr as SlideEditorContainer,
  Ar as SlideSideBar,
  Ne as SlidesUIController,
  kr as SlidesUIMenuSchema,
  bt as UniverSlidesUIPlugin,
  K as UpdateSlideElementOperation
};
