// @univerjs/facade/lib/es/index.js
import { FPermission as I, FRange as j, FSelection as w, FSheetHooks as D, FWorkbook as P, FWorksheet as W } from "@univerjs/sheets/facade";
import "@univerjs/ui/facade";
import "@univerjs/docs-ui/facade";
import "@univerjs/sheets-ui/facade";
import "@univerjs/sheets-zen-editor/facade";
import { FDataValidation as A, FDataValidationBuilder as C } from "@univerjs/sheets-data-validation/facade";
import { FFormula as N } from "@univerjs/engine-formula/facade";
import { FFilter as R } from "@univerjs/sheets-filter/facade";
import "@univerjs/sheets-formula/facade";
import "@univerjs/sheets-numfmt/facade";
import "@univerjs/sheets-hyper-link/facade";
import "@univerjs/sheets-hyper-link-ui/facade";
import { FThreadComment as U } from "@univerjs/sheets-thread-comment/facade";
import { FConditionalFormattingBuilder as H } from "@univerjs/sheets-conditional-formatting/facade";
import { FUniver as i, Quantity as s, Univer as a } from "@univerjs/core";
import { FHooks as _, FUniver as y } from "@univerjs/core";
import { ISocketService as o, WebSocketService as d } from "@univerjs/network";
class n extends i {
  /**
   * Get dependencies for FUniver, you can override newAPI to add more dependencies.
   *
   * @static
   * @protected
   *
   * @param {Injector} injector - The injector instance used to retrieve dependencies.
   * @param {Dependency[]} [derivedDependencies] - Optional array of pre-derived dependencies.
   * @returns {Dependency[]} - An array of dependencies required by the service.
   */
  static getDependencies(e, t) {
    const r = t || [];
    return e.get(o, s.OPTIONAL) || r.push([o, { useClass: d }]), r;
  }
  /**
   * Create an FUniver instance, if the injector is not provided, it will create a new Univer instance.
   *
   * @static
   *
   * @param {Univer | Injector} wrapped - The Univer instance or injector instance.
   * @returns {FUniver} - The FUniver instance.
   */
  static newAPI(e) {
    const t = e instanceof a ? e.__getInjector() : e;
    return n.getDependencies(t).forEach((c) => t.add(c)), t.createInstance(i);
  }
  createSocket(e) {
    const r = this._injector.get(o).createSocket(e);
    if (!r)
      throw new Error("[WebSocketService]: failed to create socket!");
    return r;
  }
}
i.extend(n);
console.error(`[!DEPRECATION WARNING!] @univerjs/facade and @univerjs-pro/facade is deprecated and will be removed in v0.6.0.
Please refer to our official guide to migrate. https://univer.ai/guides/sheet/getting-started/facade`);
export {
  H as FConditionalFormattingBuilder,
  A as FDataValidation,
  C as FDataValidationBuilder,
  R as FFilter,
  N as FFormula,
  _ as FHooks,
  I as FPermission,
  j as FRange,
  w as FSelection,
  D as FSheetHooks,
  U as FThreadComment,
  y as FUniver,
  P as FWorkbook,
  W as FWorksheet
};
