// @univerjs/sheets-filter/lib/es/facade.js
import { Inject as S, Injector as C, ICommandService as g, FEnum as F, FEventName as d, FUniver as u } from "@univerjs/core";
import { SetSheetsFilterCriteriaCommand as o, ClearSheetsFilterCriteriaCommand as l, RemoveSheetFilterCommand as f, SetSheetFilterRangeCommand as v, SheetsFilterService as _, CustomFilterOperator as w } from "@univerjs/sheets-filter";
import { FRange as c, FWorksheet as k, FSheetEventName as b } from "@univerjs/sheets/facade";
var I = Object.defineProperty, R = Object.getOwnPropertyDescriptor, E = (i, e, t, r) => {
  for (var n = r > 1 ? void 0 : r ? R(e, t) : e, a = i.length - 1, h; a >= 0; a--)
    (h = i[a]) && (n = (r ? h(e, t, n) : h(n)) || n);
  return r && n && I(e, t, n), n;
}, m = (i, e) => (t, r) => e(t, r, i);
let s = class {
  constructor(i, e, t, r, n) {
    this._workbook = i, this._worksheet = e, this._filterModel = t, this._injector = r, this._commandSrv = n;
  }
  /**
   * Get the filtered out rows by this filter.
   * @returns {number[]} Filtered out rows by this filter.
   */
  getFilteredOutRows() {
    return Array.from(this._filterModel.filteredOutRows).sort();
  }
  /**
   * Get the filter criteria of a column.
   * @param {number} col The column number.
   * @returns {Nullable<IFilterColumn>} The filter criteria of the column.
   */
  getColumnFilterCriteria(i) {
    var e;
    return (e = this._filterModel.getFilterColumn(i)) == null ? void 0 : e.getColumnData();
  }
  /**
   * Clear the filter criteria of a column.
   * @param {number} col The column number.
   * @returns {FFilter} The interface class to handle the filter.
   */
  removeColumnFilterCriteria(i) {
    return this._commandSrv.syncExecuteCommand(o.id, {
      unitId: this._workbook.getUnitId(),
      subUnitId: this._worksheet.getSheetId(),
      col: i,
      criteria: null
    }), this;
  }
  /**
   * Set the filter criteria of a column.
   * @param {number} col  The column number.
   * @param {ISetSheetsFilterCriteriaCommandParams['criteria']} criteria The new filter criteria.
   * @returns {FFilter} The interface class to handle the filter.
   */
  setColumnFilterCriteria(i, e) {
    return this._commandSrv.syncExecuteCommand(o.id, {
      unitId: this._workbook.getUnitId(),
      subUnitId: this._worksheet.getSheetId(),
      col: i,
      criteria: e
    }), this;
  }
  /**
   * Get the range of the filter.
   * @returns {FRange} The range of the filter.
   */
  getRange() {
    const i = this._filterModel.getRange();
    return this._injector.createInstance(c, this._workbook, this._worksheet, i);
  }
  /**
   * Remove the filter criteria of all columns.
   * @returns {Promise<boolean>} If the filter criteria is removed.
   */
  removeFilterCriteria() {
    return this._commandSrv.syncExecuteCommand(l.id), this;
  }
  /**
   * Remove the filter from the worksheet.
   * @returns {boolean} If the filter is removed.
   */
  remove() {
    return this._commandSrv.syncExecuteCommand(f.id, {
      unitId: this._workbook.getUnitId(),
      subUnitId: this._worksheet.getSheetId()
    });
  }
};
s = E([
  m(3, S(C)),
  m(4, g)
], s);
class x extends c {
  createFilter() {
    return this._getFilterModel() || !this._commandService.syncExecuteCommand(v.id, {
      unitId: this._workbook.getUnitId(),
      subUnitId: this._worksheet.getSheetId(),
      range: this._range
    }) ? null : this.getFilter();
  }
  /**
   * Get the filter for the current range's worksheet.
   * @returns {FFilter | null} The interface class to handle the filter. If the worksheet does not have a filter,
   * this method would return `null`.
   */
  getFilter() {
    const e = this._getFilterModel();
    return e ? this._injector.createInstance(s, this._workbook, this._worksheet, e) : null;
  }
  _getFilterModel() {
    return this._injector.get(_).getFilterModel(
      this._workbook.getUnitId(),
      this._worksheet.getSheetId()
    );
  }
}
c.extend(x);
class p extends k {
  getFilter() {
    const e = this._getFilterModel();
    return e ? this._injector.createInstance(s, this._workbook, this._worksheet, e) : null;
  }
  _getFilterModel() {
    return this._injector.get(_).getFilterModel(
      this._workbook.getUnitId(),
      this._worksheet.getSheetId()
    );
  }
}
k.extend(p);
class M extends F {
  get CustomFilterOperation() {
    return w;
  }
}
F.extend(M);
class U extends d {
  get SheetBeforeRangeFilter() {
    return "SheetBeforeRangeFilter";
  }
  get SheetRangeFiltered() {
    return "SheetRangeFiltered";
  }
  get SheetRangeFilterCleared() {
    return "SheetRangeFilterCleared";
  }
  get SheetBeforeRangeFilterClear() {
    return "SheetBeforeRangeFilterClear";
  }
}
d.extend(U);
d.extend(b);
class j extends u {
  _initialize(e) {
    const t = e.get(g);
    this.disposeWithMe(t.beforeCommandExecuted((r) => {
      switch (r.id) {
        case o.id:
          this._beforeRangeFilter(r);
          break;
        case l.id:
          this._beforeRangeFilterClear();
          break;
      }
    })), this.disposeWithMe(t.onCommandExecuted((r) => {
      switch (r.id) {
        case o.id:
          this._onRangeFiltered(r);
          break;
        case l.id:
          this._onRangeFilterCleared();
          break;
      }
    }));
  }
  _beforeRangeFilter(e) {
    const t = e.params, r = this.getUniverSheet(t.unitId), n = {
      workbook: r,
      worksheet: r.getSheetBySheetId(t.subUnitId),
      col: t.col,
      criteria: t.criteria
    };
    if (this.fireEvent(this.Event.SheetBeforeRangeFilter, n), n.cancel)
      throw new Error("SetSheetsFilterCriteriaCommand canceled.");
  }
  _onRangeFiltered(e) {
    const t = e.params, r = this.getUniverSheet(t.unitId), n = {
      workbook: r,
      worksheet: r.getSheetBySheetId(t.subUnitId),
      col: t.col,
      criteria: t.criteria
    };
    if (this.fireEvent(this.Event.SheetRangeFiltered, n), n.cancel)
      throw new Error("SetSheetsFilterCriteriaCommand canceled.");
  }
  _beforeRangeFilterClear() {
    const e = this.getActiveWorkbook();
    if (!e) return;
    const t = {
      workbook: e,
      worksheet: e.getActiveSheet()
    };
    if (this.fireEvent(this.Event.SheetBeforeRangeFilterClear, t), t.cancel)
      throw new Error("SetSheetsFilterCriteriaCommand canceled.");
  }
  _onRangeFilterCleared() {
    const e = this.getActiveWorkbook();
    if (!e) return;
    const t = {
      workbook: e,
      worksheet: e.getActiveSheet()
    };
    if (this.fireEvent(this.Event.SheetRangeFilterCleared, t), t.cancel)
      throw new Error("SetSheetsFilterCriteriaCommand canceled.");
  }
}
u.extend(j);
export {
  s as FFilter
};
