// @univerjs/preset-docs-collaboration/lib/es/index.js
import { UniverCollaborationPlugin as i } from "@univerjs-pro/collaboration";
export * from "@univerjs-pro/collaboration";
import { UniverCollaborationClientPlugin as n } from "@univerjs-pro/collaboration-client";
export * from "@univerjs-pro/collaboration-client";
import { UniverCollaborationClientUIPlugin as t } from "@univerjs-pro/collaboration-client-ui";
export * from "@univerjs-pro/collaboration-client-ui";
import "@univerjs-pro/collaboration-client/facade";
function l(e) {
  const r = new URL(e, window.location.origin);
  switch (r.protocol) {
    case "https:":
      r.protocol = "wss:";
      break;
    case "http:":
      r.protocol = "ws:";
      break;
  }
  return r.toString();
}
function u(e = {}) {
  const { universerEndpoint: r } = e, o = r != null ? r : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      i,
      [n, {
        enableOfflineEditing: !0,
        enableSingleActiveInstanceLock: !0,
        enableAuthServer: !0,
        authzUrl: `${o}/universer-api/authz`,
        snapshotServerUrl: `${o}/universer-api/snapshot`,
        collabSubmitChangesetUrl: `${o}/universer-api/comb`,
        collabWebSocketUrl: l(`${o}/universer-api/comb/connect`),
        loginUrlKey: `${o}/universer-api/oidc/authpage`,
        uploadFileServerUrl: `${o}/universer-api/stream/file/upload`,
        signUrlServerUrl: `${o}/universer-api/file/{fileID}/sign-url`,
        downloadEndpointUrl: `${o}/`,
        wsSessionTicketUrl: `${o}/universer-api/user/session-ticket`,
        sendChangesetTimeout: 200
      }],
      t
    ]
  };
}
export {
  u as UniverDocsCollaborationPreset
};
