// @univerjs/docs-hyper-link/lib/es/index.js
var m = Object.defineProperty;
var O = (r, e, n) => e in r ? m(r, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : r[e] = n;
var _ = (r, e, n) => O(r, typeof e != "symbol" ? e + "" : e, n);
import { CommandType as l, Inject as y, IResourceManagerService as M, Disposable as N, UniverInstanceType as d, IUniverInstanceService as E, CustomRangeType as f, Plugin as C, merge as D, Injector as S, IConfigService as U, ICommandService as L } from "@univerjs/core";
const R = {
  id: "docs.mutation.add-hyper-link",
  type: l.MUTATION,
  handler: () => !0
}, T = {
  id: "docs.mutation.update-hyper-link",
  type: l.MUTATION,
  handler: () => !0
}, H = {
  id: "docs.mutation.delete-hyper-link",
  type: l.MUTATION,
  handler: () => !0
}, j = "docs-hyper-link.config", g = {};
var b = Object.defineProperty, K = Object.getOwnPropertyDescriptor, Y = (r, e, n, o) => {
  for (var t = o > 1 ? void 0 : o ? K(e, n) : e, i = r.length - 1, s; i >= 0; i--)
    (s = r[i]) && (t = (o ? s(e, n, t) : s(t)) || t);
  return o && t && b(e, n, t), t;
}, v = (r, e) => (n, o) => e(n, o, r);
const P = "DOC_HYPER_LINK_PLUGIN";
let u = class extends N {
  constructor(r, e) {
    super(), this._resourceManagerService = r, this._univerInstanceService = e, this._init();
  }
  _init() {
    this._resourceManagerService.registerPluginResource({
      pluginName: P,
      businesses: [d.UNIVER_DOC],
      onLoad: (r, e) => {
        const n = this._univerInstanceService.getUnit(r, d.UNIVER_DOC);
        if (!n)
          return;
        const o = /* @__PURE__ */ new Map(), t = (i) => {
          var s, a;
          return (a = (s = i.getBody()) == null ? void 0 : s.customRanges) == null || a.forEach((c) => {
            c.rangeType === f.HYPERLINK && o.set(c.rangeId, c);
          }), o;
        };
        n.headerModelMap.forEach((i) => {
          t(i);
        }), n.footerModelMap.forEach((i) => {
          t(i);
        }), t(n), e.links.forEach((i) => {
          const s = o.get(i.id);
          s && (s.properties = {
            ...s.properties,
            url: i.payload
          });
        });
      },
      onUnLoad: (r) => {
      },
      toJson: (r) => {
        const e = this._univerInstanceService.getUnit(r, d.UNIVER_DOC), n = [];
        if (e) {
          const o = (t) => {
            var i, s;
            (s = (i = t.getBody()) == null ? void 0 : i.customRanges) == null || s.forEach((a) => {
              var c;
              a.rangeType === f.HYPERLINK && n.push({
                id: a.rangeId,
                payload: ((c = a.properties) == null ? void 0 : c.url) || ""
              });
            });
          };
          e.headerModelMap.forEach((t) => {
            o(t);
          }), e.footerModelMap.forEach((t) => {
            o(t);
          }), o(e);
        }
        return JSON.stringify({ links: n });
      },
      parseJson(r) {
        return JSON.parse(r);
      }
    });
  }
};
u = Y([
  v(0, y(M)),
  v(1, E)
], u);
var w = Object.defineProperty, A = Object.getOwnPropertyDescriptor, G = (r, e, n, o) => {
  for (var t = o > 1 ? void 0 : o ? A(e, n) : e, i = r.length - 1, s; i >= 0; i--)
    (s = r[i]) && (t = (o ? s(e, n, t) : s(t)) || t);
  return o && t && w(e, n, t), t;
}, h = (r, e) => (n, o) => e(n, o, r), p;
let I = (p = class extends C {
  constructor(r = g, e, n, o) {
    super(), this._config = r, this._injector = e, this._configService = n, this._commandService = o;
    const { ...t } = D(
      {},
      g,
      this._config
    );
    this._configService.setConfig(j, t);
  }
  onStarting() {
    [[u]].forEach((e) => this._injector.add(e)), [R, H, T].forEach((e) => {
      this.disposeWithMe(this._commandService.registerCommand(e));
    }), this._injector.get(u);
  }
}, _(p, "pluginName", P), _(p, "type", d.UNIVER_DOC), p);
I = G([
  h(1, y(S)),
  h(2, U),
  h(3, L)
], I);
export {
  I as UniverDocsHyperLinkPlugin
};
