// @univerjs/sheets-graphics/lib/es/index.js
var N = Object.defineProperty;
var S = (n, r, e) => r in n ? N(n, r, { enumerable: !0, configurable: !0, writable: !0, value: e }) : n[r] = e;
var i = (n, r, e) => S(n, typeof r != "symbol" ? r + "" : r, e);
import { Range as v, Disposable as x, Inject as P, Injector as M, Plugin as m, merge as G, UniverInstanceType as C, IConfigService as U } from "@univerjs/core";
import { SheetExtension as y, IRenderManagerService as D } from "@univerjs/engine-render";
const O = "graphics.config", g = {}, f = "SheetGraphicsExtension", j = 35;
class w extends y {
  constructor() {
    super(...arguments);
    i(this, "uKey", f);
    i(this, "Z_INDEX", j);
    i(this, "_graphicsRenderMap", /* @__PURE__ */ new Map());
  }
  registerRenderer(e, t) {
    this._graphicsRenderMap.set(e, t);
  }
  draw(e, t, s, a, { viewRanges: c }) {
    const E = Array.from(this._graphicsRenderMap.keys());
    c.forEach((l) => {
      v.foreach(l, (I, R) => {
        const p = s.getCellByIndexWithNoHeader(I, R);
        p && E.forEach((u) => {
          const o = this._graphicsRenderMap.get(u);
          o == null || o(e, s, p);
        });
      });
    });
  }
  dispose() {
    this._graphicsRenderMap.clear();
  }
}
class T extends x {
  constructor(e) {
    super();
    i(this, "_graphicsExtensionInstance", null);
    this._context = e, this._initRender();
  }
  _initRender() {
    const e = this._context.mainComponent;
    e && !e.getExtensionByKey(f) && (this._graphicsExtensionInstance = new w(), e.register(this._graphicsExtensionInstance));
  }
  registerRenderer(e, t) {
    var s;
    (s = this._graphicsExtensionInstance) == null || s.registerRenderer(e, t);
  }
}
var A = Object.defineProperty, H = Object.getOwnPropertyDescriptor, K = (n, r, e, t) => {
  for (var s = t > 1 ? void 0 : t ? H(r, e) : r, a = n.length - 1, c; a >= 0; a--)
    (c = n[a]) && (s = (t ? c(r, e, s) : c(s)) || s);
  return t && s && A(r, e, s), s;
}, h = (n, r) => (e, t) => r(e, t, n);
const b = "UNIVER_SHEET_DRAWING_PLUGIN";
var _;
let d = (_ = class extends m {
  constructor(n = g, r, e, t) {
    super(), this._config = n, this._injector = r, this._configService = e, this._renderManagerService = t;
    const { ...s } = G(
      {},
      g,
      this._config
    );
    this._configService.setConfig(O, s);
  }
  onRendered() {
    [
      [T]
    ].forEach((n) => {
      this._renderManagerService.registerRenderModule(C.UNIVER_SHEET, n);
    });
  }
}, i(_, "pluginName", b), _);
d = K([
  h(1, P(M)),
  h(2, U),
  h(3, D)
], d);
export {
  T as SheetGraphicsRenderController,
  d as UniverSheetGraphicsPlugin
};
