// @univerjs/presets/lib/es/index.js
import { IUndoRedoService as A, IAuthzIoService as h, IMentionIOService as m, Univer as I, LogLevel as y, FUniver as P } from "@univerjs/core";
export * from "@univerjs/core";
export * from "@univerjs/design";
function d(c) {
  const { presets: t, plugins: a, collaboration: g, override: o = [], ...p } = c;
  g && (o.push([A, null]), o.push([h, null]), o.push([m, null]));
  const u = new I({
    logLevel: y.WARN,
    ...p,
    override: o
  }), e = /* @__PURE__ */ new Map();
  t == null || t.forEach((n) => {
    const r = Array.isArray(n) ? n[0] : n, { plugins: s } = r;
    s.forEach((i) => {
      const [l, v] = Array.isArray(i) ? [i[0], i[1]] : [i];
      e.has(l.pluginName) && e.delete(l.pluginName), e.set(l.pluginName, { plugin: l, options: v });
    });
  }), a == null || a.forEach((n) => {
    const [r, s] = Array.isArray(n) ? [n[0], n[1]] : [n];
    e.set(r.pluginName, { plugin: r, options: s });
  }), e.forEach(({ plugin: n, options: r }) => {
    u.registerPlugin(n, r);
  });
  const f = P.newAPI(u);
  return {
    univer: u,
    univerAPI: f
  };
}
export {
  d as createUniver
};
