// @univerjs/preset-sheets-collaboration/lib/es/index.js
import { UniverCollaborationPlugin as t } from "@univerjs-pro/collaboration";
export * from "@univerjs-pro/collaboration";
import { UniverCollaborationClientPlugin as n } from "@univerjs-pro/collaboration-client";
export * from "@univerjs-pro/collaboration-client";
import { UniverCollaborationClientUIPlugin as l } from "@univerjs-pro/collaboration-client-ui";
export * from "@univerjs-pro/collaboration-client-ui";
import { UniverEditHistoryLoaderPlugin as s } from "@univerjs-pro/edit-history-loader";
export * from "@univerjs-pro/edit-history-loader";
function a(o) {
  const e = new URL(o, window.location.origin);
  switch (e.protocol) {
    case "https:":
      e.protocol = "wss:";
      break;
    case "http:":
      e.protocol = "ws:";
      break;
  }
  return e.toString();
}
function U(o = {}) {
  const { universerEndpoint: e, univerContainerId: i = "app" } = o, r = e != null ? e : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      t,
      [n, {
        enableOfflineEditing: !0,
        enableSingleActiveInstanceLock: !0,
        enableAuthServer: !0,
        authzUrl: `${r}/universer-api/authz`,
        snapshotServerUrl: `${r}/universer-api/snapshot`,
        collabSubmitChangesetUrl: `${r}/universer-api/comb`,
        collabWebSocketUrl: a(`${r}/universer-api/comb/connect`),
        loginUrlKey: `${r}/universer-api/oidc/authpage`,
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        downloadEndpointUrl: `${r}/`,
        wsSessionTicketUrl: `${r}/universer-api/user/session-ticket`,
        sendChangesetTimeout: 200
      }],
      [l, {}],
      [s, {
        univerContainerId: i,
        historyListServerUrl: `${r}/universer-api/history`
      }]
    ]
  };
}
export {
  U as UniverSheetsCollaborationPreset
};
