// @univerjs/rpc/lib/es/index.js
var X = Object.defineProperty;
var Z = (r, t, e) => t in r ? X(r, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : r[t] = e;
var a = (r, t, e) => Z(r, typeof t != "symbol" ? t + "" : t, e);
import { createIdentifier as O, UniverInstanceType as u, ICommandService as b, IUniverInstanceService as H, ILogService as ee, RxDisposable as j, Inject as C, Injector as R, toDisposable as te, CommandType as k, Disposable as ne, Plugin as A, merge as V, IConfigService as F } from "@univerjs/core";
import { takeUntil as v, filter as se, take as re } from "rxjs/operators";
import { isObservable as ie, of as ce, BehaviorSubject as oe, firstValueFrom as ae, Observable as $, shareReplay as z } from "rxjs";
const he = "rpc.main-thread.config", D = {}, _e = "rpc.worker-thread.config", N = {};
var le = Object.defineProperty, ue = Object.getOwnPropertyDescriptor, G = (r, t, e, n) => {
  for (var s = n > 1 ? void 0 : n ? ue(t, e) : t, i = r.length - 1, c; i >= 0; i--)
    (c = r[i]) && (s = (n ? c(t, e, s) : c(s)) || s);
  return n && s && le(t, e, s), s;
}, p = (r, t) => (e, n) => t(e, n, r);
const T = "rpc.remote-sync.service", f = O(T);
let M = class {
  constructor(r) {
    this._commandService = r;
  }
  async syncMutation(r) {
    return this._commandService.syncExecuteCommand(r.mutationInfo.id, r.mutationInfo.params, {
      onlyLocal: !0,
      fromSync: !0
    });
  }
};
M = G([
  p(0, b)
], M);
const W = "univer.remote-instance-service", m = O(W);
let q = class {
  constructor(r, t, e) {
    this._univerInstanceService = r, this._commandService = t, this._logService = e;
  }
  whenReady() {
    return Promise.resolve(!0);
  }
  async syncMutation(r) {
    return this._applyMutation(r.mutationInfo);
  }
  async createInstance(r) {
    const { type: t, snapshot: e } = r;
    try {
      switch (t) {
        case u.UNIVER_SHEET:
          return this._univerInstanceService.createUnit(u.UNIVER_SHEET, e), !0;
        default:
          throw new Error(
            `[WebWorkerRemoteInstanceService]: cannot create replica for document type: ${t}.`
          );
      }
    } catch (n) {
      throw n instanceof Error ? n : new TypeError(`${n}`);
    }
  }
  async disposeInstance(r) {
    return this._univerInstanceService.disposeUnit(r.unitID);
  }
  _applyMutation(r) {
    const { id: t, params: e } = r;
    return this._commandService.syncExecuteCommand(t, e, {
      onlyLocal: !0,
      fromSync: !0
    });
  }
};
q = G([
  p(0, H),
  p(1, b),
  p(2, ee)
], q);
function Y(r) {
  const t = r;
  return new class {
    call(e, n) {
      const s = t[e];
      if (typeof s == "function") {
        let i = n ? s.apply(t, n) : s.call(t);
        return i instanceof Promise || (i = Promise.resolve(i)), i;
      }
      throw new Error(`[RPC]: method not found for ${e}!`);
    }
    subscribe(e, n) {
      const s = t[e];
      if (typeof s == "function") {
        const i = n ? s.apply(t, n) : s.call(t);
        return ie(i) ? i : ce(i);
      }
      throw new Error(`[RPC]: observable method not found for ${e}!`);
    }
  }();
}
function B(r) {
  return new Proxy({}, {
    get(t, e) {
      if (e !== "dispose")
        return function(...n) {
          return de(e) ? r.subscribe(e, n) : r.call(e, n);
        };
    }
  });
}
function de(r) {
  return r.endsWith("$");
}
class pe extends j {
  constructor(e) {
    super();
    a(this, "_initialized", new oe(!1));
    a(this, "_lastRequestCounter", 0);
    a(this, "_pendingRequests", /* @__PURE__ */ new Map());
    this._protocol = e, this._protocol.send({
      type: 50
      /* REQUEST_INITIALIZATION */
    }), this._protocol.onMessage.pipe(v(this.dispose$)).subscribe((n) => this._onMessage(n));
  }
  dispose() {
    this._pendingRequests.clear();
  }
  getChannel(e) {
    const n = this;
    return {
      call(s, i) {
        return n._disposed ? Promise.reject() : n._remoteCall(e, s, i);
      },
      subscribe(s, i) {
        if (n._disposed)
          throw new Error("[ChannelClient]: client is disposed!");
        return n._remoteSubscribe(e, s, i);
      }
    };
  }
  _whenReady() {
    return ae(
      this._initialized.pipe(
        se((e) => e),
        re(1)
      )
    );
  }
  async _remoteCall(e, n, s) {
    await this._whenReady();
    const i = ++this._lastRequestCounter, h = { seq: i, type: 100, channelName: e, method: n, args: s }, o = this;
    return new Promise((w, _) => {
      const Q = {
        handle(P) {
          switch (P.type) {
            case 201:
              o._pendingRequests.delete(i), w(P.data);
              break;
            case 202:
              o._pendingRequests.delete(i), _(P.data);
              break;
            default:
              throw new Error("[ChannelClient]: unknown response type!");
          }
        }
      };
      this._pendingRequests.set(i, Q), this._sendRequest(h);
    });
  }
  _remoteSubscribe(e, n, s) {
    return new $((i) => {
      let c = -1;
      return this._whenReady().then(() => {
        c = ++this._lastRequestCounter;
        const o = { seq: c, type: 101, channelName: e, method: n, args: s }, w = {
          handle(_) {
            switch (_.type) {
              case 300:
                i.next(_.data);
                break;
              case 301:
                i.error(_.data);
                break;
              case 302:
                i.complete();
                break;
              default:
                throw new Error("[ChannelClient]: unknown response type!");
            }
          }
        };
        this._pendingRequests.set(c, w), this._sendRequest(o);
      }), () => {
        if (c === -1)
          return;
        const h = {
          type: 102,
          seq: c,
          channelName: e,
          method: n
        };
        this._sendRequest(h);
      };
    });
  }
  _sendRequest(e) {
    this._protocol.send(e);
  }
  _onMessage(e) {
    var n;
    switch (e.type) {
      case 0:
        this._initialized.next(!0);
        break;
      case 201:
      case 202:
      case 300:
      case 302:
      case 301:
        (n = this._pendingRequests.get(e.seq)) == null || n.handle(e);
        break;
    }
  }
}
class ve extends j {
  constructor(e) {
    super();
    a(this, "_channels", /* @__PURE__ */ new Map());
    a(this, "_subscriptions", /* @__PURE__ */ new Map());
    this._protocol = e, this._protocol.onMessage.pipe(v(this.dispose$)).subscribe((n) => this._onRequest(n)), this._sendInitialize();
  }
  dispose() {
    super.dispose(), this._subscriptions.clear(), this._channels.clear();
  }
  registerChannel(e, n) {
    this._channels.set(e, n);
  }
  _onRequest(e) {
    switch (e.type) {
      case 50:
        this._sendInitialize();
        break;
      case 100:
        this._onMethodCall(e);
        break;
      case 101:
        this._onSubscribe(e);
        break;
      case 102:
        this._onUnsubscribe(e);
        break;
    }
  }
  _sendInitialize() {
    this._sendResponse({
      seq: -1,
      type: 0
      /* INITIALIZE */
    });
  }
  _onMethodCall(e) {
    const { channelName: n, method: s, args: i } = e, c = this._channels.get(n);
    let h;
    try {
      if (!c)
        throw new Error(`[ChannelServer]: Channel ${n} not found!`);
      h = i ? c.call(s, i) : c.call(s);
    } catch (o) {
      h = Promise.reject(o);
    }
    h.then((o) => {
      this._sendResponse({ seq: e.seq, type: 201, data: o });
    }).catch((o) => {
      o instanceof Error ? this._sendResponse({ seq: e.seq, type: 202, data: o.message }) : this._sendResponse({ seq: e.seq, type: 202, data: String(o) });
    });
  }
  _onSubscribe(e) {
    const { channelName: n, seq: s } = e, i = this._channels.get(n);
    try {
      if (!i)
        throw new Error(`[ChannelServer]: Channel ${n} not found!`);
      const h = i.subscribe(e.method, e.args).subscribe({
        next: (o) => {
          this._sendResponse({ seq: s, type: 300, data: o });
        },
        error: (o) => {
          this._sendResponse({ seq: s, type: 301, data: o.message }), this._sendResponse({
            seq: s,
            type: 302
            /* SUBSCRIBE_COMPLETE */
          });
        },
        complete: () => {
          this._sendResponse({
            seq: s,
            type: 302
            /* SUBSCRIBE_COMPLETE */
          });
        }
      });
      this._subscriptions.set(e.seq, h);
    } catch (c) {
      c instanceof Error ? this._sendResponse({ seq: e.seq, type: 301, data: c.message }) : this._sendResponse({ seq: e.seq, type: 301, data: String(c) });
    }
  }
  _onUnsubscribe(e) {
    const n = this._subscriptions.get(e.seq);
    n && (n.unsubscribe(), this._subscriptions.delete(e.seq));
  }
  _sendResponse(e) {
    this._protocol.send(e);
  }
}
const I = O("IRPCChannelService");
class K {
  constructor(t) {
    a(this, "_client");
    a(this, "_server");
    this._client = new pe(t), this._server = new ve(t);
  }
  dispose() {
    this._client.dispose(), this._server.dispose();
  }
  requestChannel(t) {
    return this._client.getChannel(t);
  }
  registerChannel(t, e) {
    this._server.registerChannel(t, e);
  }
}
var fe = Object.defineProperty, me = Object.getOwnPropertyDescriptor, ge = (r, t, e, n) => {
  for (var s = n > 1 ? void 0 : n ? me(t, e) : t, i = r.length - 1, c; i >= 0; i--)
    (c = r[i]) && (s = (n ? c(t, e, s) : c(s)) || s);
  return n && s && fe(t, e, s), s;
}, l = (r, t) => (e, n) => t(e, n, r);
let g = class extends j {
  constructor(t, e, n, s, i) {
    super();
    a(this, "_remoteInstanceService");
    a(this, "_syncingUnits", /* @__PURE__ */ new Set());
    a(this, "_syncingMutations", /* @__PURE__ */ new Set());
    this._injector = t, this._commandService = e, this._univerInstanceService = n, this._rpcChannelService = s, this._remoteSyncService = i, this._initRPCChannels(), this._init();
  }
  registerSyncingMutations(t) {
    this._syncingMutations.add(t.id);
  }
  /**
   * Only spreadsheets would be synced to the web worker in normal situations. If you would like to
   * sync other types of documents, you should manually call this method with that document's id.
   */
  syncUnit(t) {
    return this._syncingUnits.add(t), te(() => this._syncingUnits.delete(t));
  }
  _initRPCChannels() {
    this._rpcChannelService.registerChannel(T, Y(this._remoteSyncService)), this._injector.add([
      m,
      { useFactory: () => B(this._rpcChannelService.requestChannel(W)) }
    ]), this._remoteInstanceService = this._injector.get(m);
  }
  _init() {
    this._univerInstanceService.getTypeOfUnitAdded$(u.UNIVER_SHEET).pipe(v(this.dispose$)).subscribe((t) => {
      this._syncingUnits.add(t.getUnitId()), this._remoteInstanceService.createInstance({
        unitID: t.getUnitId(),
        type: u.UNIVER_SHEET,
        snapshot: t.getSnapshot()
      });
    }), this._univerInstanceService.getTypeOfUnitDisposed$(u.UNIVER_SHEET).pipe(v(this.dispose$)).subscribe((t) => {
      this._syncingUnits.delete(t.getUnitId()), this._remoteInstanceService.disposeInstance({
        unitID: t.getUnitId()
      });
    }), this.disposeWithMe(this._commandService.onCommandExecuted((t, e) => {
      const { type: n, params: s, id: i } = t, c = (s == null ? void 0 : s.unitId) || "";
      n === k.MUTATION && // only sync mutations to the worker thread
      (!c || this._syncingUnits.has(c)) && // do not sync mutations from the web worker back to the web worker
      !(e != null && e.fromSync) && // do not sync mutations those are not meant to be synced
      this._syncingMutations.has(i) && this._remoteInstanceService.syncMutation({ mutationInfo: t });
    }));
  }
};
g = ge([
  l(0, C(R)),
  l(1, b),
  l(2, H),
  l(3, I),
  l(4, f)
], g);
var ye = Object.defineProperty, Se = Object.getOwnPropertyDescriptor, be = (r, t, e, n) => {
  for (var s = n > 1 ? void 0 : n ? Se(t, e) : t, i = r.length - 1, c; i >= 0; i--)
    (c = r[i]) && (s = (n ? c(t, e, s) : c(s)) || s);
  return n && s && ye(t, e, s), s;
}, d = (r, t) => (e, n) => t(e, n, r);
let y = class extends ne {
  constructor(t, e, n, s) {
    super();
    a(this, "_remoteSyncService");
    this._injector = t, this._remoteInstanceService = e, this._commandService = n, this._rpcChannelService = s, this._initRPCChannels(), this._init();
  }
  _initRPCChannels() {
    this._rpcChannelService.registerChannel(W, Y(this._remoteInstanceService)), this._injector.add([
      f,
      { useFactory: () => B(this._rpcChannelService.requestChannel(T)) }
    ]), this._remoteSyncService = this._injector.get(f);
  }
  _init() {
    this.disposeWithMe(
      // Mutations executed on the main thread should be synced to the worker thread.
      this._commandService.onCommandExecuted((t, e) => {
        t.type === k.MUTATION && !(e != null && e.fromSync) && this._remoteSyncService.syncMutation({
          mutationInfo: t
        });
      })
    );
  }
};
y = be([
  d(0, C(R)),
  d(1, m),
  d(2, b),
  d(3, I)
], y);
function Ce() {
  return {
    send(r) {
      postMessage(r);
    },
    onMessage: new $((r) => {
      const t = (e) => {
        r.next(e.data);
      };
      return addEventListener("message", t), () => removeEventListener("message", t);
    }).pipe(z(1))
  };
}
function Re(r) {
  return {
    send(t) {
      r.postMessage(t);
    },
    onMessage: new $((t) => {
      const e = (n) => {
        t.next(n.data);
      };
      return r.addEventListener("message", e), () => r.removeEventListener("message", e);
    }).pipe(z(1))
  };
}
var Ie = Object.defineProperty, we = Object.getOwnPropertyDescriptor, J = (r, t, e, n) => {
  for (var s = n > 1 ? void 0 : n ? we(t, e) : t, i = r.length - 1, c; i >= 0; i--)
    (c = r[i]) && (s = (n ? c(t, e, s) : c(s)) || s);
  return n && s && Ie(t, e, s), s;
}, S = (r, t) => (e, n) => t(e, n, r), E;
let x = (E = class extends A {
  constructor(t = D, e, n) {
    super();
    a(this, "_internalWorker", null);
    this._config = t, this._injector = e, this._configService = n;
    const { ...s } = V(
      {},
      D,
      this._config
    );
    this._configService.setConfig(he, s);
  }
  dispose() {
    super.dispose(), this._internalWorker && (this._internalWorker.terminate(), this._internalWorker = null);
  }
  onStarting() {
    const { workerURL: t } = this._config;
    if (!t)
      throw new Error("[UniverRPCMainThreadPlugin]: The workerURL is required for the RPC main thread plugin.");
    const e = t instanceof Worker ? t : new Worker(t);
    this._internalWorker = t instanceof Worker ? null : e;
    const n = Re(e);
    [
      [
        I,
        {
          useFactory: () => new K(n)
        }
      ],
      [g],
      [f, { useClass: M }]
    ].forEach((i) => this._injector.add(i)), this._injector.get(g);
  }
}, a(E, "pluginName", "UNIVER_RPC_MAIN_THREAD_PLUGIN"), E);
x = J([
  S(1, C(R)),
  S(2, F)
], x);
var U;
let L = (U = class extends A {
  constructor(r = N, t, e) {
    super(), this._config = r, this._injector = t, this._configService = e;
    const { ...n } = V(
      {},
      N,
      this._config
    );
    this._configService.setConfig(_e, n);
  }
  onStarting() {
    [
      [y],
      [
        I,
        {
          useFactory: () => new K(Ce())
        }
      ],
      [m, { useClass: q }]
    ].forEach((r) => this._injector.add(r)), this._injector.get(y);
  }
}, a(U, "pluginName", "UNIVER_RPC_WORKER_THREAD_PLUGIN"), U);
L = J([
  S(1, C(R)),
  S(2, F)
], L);
export {
  pe as ChannelClient,
  ve as ChannelServer,
  K as ChannelService,
  g as DataSyncPrimaryController,
  y as DataSyncReplicaController,
  I as IRPCChannelService,
  m as IRemoteInstanceService,
  f as IRemoteSyncService,
  W as RemoteInstanceServiceName,
  M as RemoteSyncPrimaryService,
  T as RemoteSyncServiceName,
  x as UniverRPCMainThreadPlugin,
  L as UniverRPCWorkerThreadPlugin,
  q as WebWorkerRemoteInstanceService,
  Y as fromModule,
  B as toModule
};
