// @univerjs/sheets-filter-ui/lib/es/index.js
var Kt = Object.defineProperty;
var Xt = (t, e, r) => e in t ? Kt(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r;
var _ = (t, e, r) => Xt(t, typeof e != "symbol" ? e + "" : e, r);
import { BooleanNumber as nt, createIdentifier as gt, Inject as E, LocaleService as K, Disposable as pe, IUniverInstanceService as Ye, ILogService as Zt, extractPureTextFromCell as Jt, numfmt as it, Injector as X, ICommandService as W, Quantity as vt, Tools as Fe, CommandType as $e, IContextService as we, useDependency as te, UniverInstanceType as Z, ThemeService as Et, RxDisposable as Tt, fromCallback as zt, InterceptorEffectEnum as er, DependentOn as Ft, Optional as tr, Plugin as qe, merge as Ct, IConfigService as Nt } from "@univerjs/core";
import { IRPCChannelService as Pt, toModule as rr, fromModule as sr } from "@univerjs/rpc";
import { CustomFilterOperator as u, SetSheetsFilterCriteriaCommand as re, SheetsFilterService as k, SmartToggleSheetsFilterCommand as me, ClearSheetsFilterCriteriaCommand as Ke, ReCalcSheetsFilterCommand as Xe, FILTER_MUTATIONS as nr, SetSheetsFilterRangeMutation as ir, SetSheetsFilterCriteriaMutation as or, RemoveSheetsFilterMutation as ar, ReCalcSheetsFilterMutation as lr, RemoveSheetFilterCommand as cr, SetSheetFilterRangeCommand as ur, UniverSheetsFilterPlugin as Ot } from "@univerjs/sheets-filter";
import { RefRangeService as hr, SheetPermissionCheckController as dr, SheetsSelectionsService as fr, getSheetCommandTarget as pr, RangeProtectionPermissionViewPoint as le, WorksheetFilterPermission as ce, WorksheetViewPermission as ue, expandToContinuousRange as mr, SheetInterceptorService as _r, SetRangeValuesMutation as Sr, INTERCEPTOR_POINT as gr } from "@univerjs/sheets";
import { ILayoutService as vr, useObservable as V, useComponentsOfPart as Er, ComponentContainer as Tr, MenuItemType as Ze, getMenuHiddenObservable as Je, RibbonStartGroup as Fr, KeyCode as Cr, MetaKeys as ot, ComponentManager as Nr, IShortcutService as Pr, IMenuManagerService as Or, IMessageService as Rr } from "@univerjs/ui";
import { SetCellEditVisibleOperation as Ir, SheetsUIPart as yr, getObservableWithExclusiveRange$ as Ar, getCurrentRangeDisable$ as br, whenSheetEditorFocused as Lr, SheetSkeletonManagerService as Mr, ISheetSelectionRenderService as $r, SelectionControl as wr, attachSelectionWithCoord as Ur, getCoordByCell as Hr, SheetsRenderService as Rt, SheetCanvasPopManagerService as Dr } from "@univerjs/sheets-ui";
import { BehaviorSubject as q, ReplaySubject as xr, of as _e, Subject as Br, merge as Vr, combineLatest as jr, throttleTime as It, startWith as yt, map as se, shareReplay as Wr, switchMap as ze, filter as kr, takeUntil as Qr, distinctUntilChanged as Gr } from "rxjs";
import { Select as at, RadioGroup as Yr, Radio as lt, Input as At, Checkbox as qr, Tree as Kr, Button as Ce, Segmented as Xr, MessageType as Zr } from "@univerjs/design";
import { Rect as ct, Shape as Jr, IRenderManagerService as bt } from "@univerjs/engine-render";
import zr, { forwardRef as Lt, useRef as es, createElement as Mt, useCallback as D, useMemo as et } from "react";
const $t = "sheets-filter-ui.config", Ne = {};
var N = /* @__PURE__ */ ((t) => (t[t.FIRST = 0] = "FIRST", t[t.SECOND = 1] = "SECOND", t))(N || {}), S = /* @__PURE__ */ ((t) => (t.NONE = "none", t.STARTS_WITH = "startsWith", t.DOES_NOT_START_WITH = "doesNotStartWith", t.ENDS_WITH = "endsWith", t.DOES_NOT_END_WITH = "doesNotEndWith", t.CONTAINS = "contains", t.DOES_NOT_CONTAIN = "doesNotContain", t.EQUALS = "equals", t.NOT_EQUALS = "notEquals", t.EMPTY = "empty", t.NOT_EMPTY = "notEmpty", t.BETWEEN = "between", t.NOT_BETWEEN = "notBetween", t.CUSTOM = "custom", t))(S || {}), f;
((t) => {
  t.NONE = {
    label: "sheets-filter.conditions.none",
    operator: S.NONE,
    order: N.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.NONE]: should not have initial form params!");
    },
    testMappingParams: (s) => s.operator1 === S.NONE,
    mapToFilterColumn: () => null,
    testMappingFilterColumn: (s) => !s.customFilters && !s.filters ? {} : !1
  }, t.EMPTY = {
    label: "sheets-filter.conditions.empty",
    operator: S.EMPTY,
    order: N.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.EMPTY]: should not have initial form params!");
    },
    testMappingParams: ({ operator1: s }) => s === S.EMPTY,
    mapToFilterColumn: () => ({ customFilters: { customFilters: [{ val: "" }] } }),
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.val === "" && o.operator === void 0 ? { operator1: S.EMPTY } : !1;
    }
  }, t.NOT_EMPTY = {
    label: "sheets-filter.conditions.not-empty",
    operator: S.NOT_EMPTY,
    order: N.SECOND,
    numOfParameters: 0,
    getDefaultFormParams: () => {
      throw new Error("[FilterConditionItems.NOT_EMPTY]: should not have initial form params!");
    },
    testMappingParams: ({ operator1: s }) => s === S.NOT_EMPTY,
    mapToFilterColumn: () => ({ customFilters: { customFilters: [{ val: "", operator: u.NOT_EQUALS }] } }),
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.val === " " && o.operator === u.NOT_EQUALS ? { operator1: S.NOT_EMPTY } : !1;
    }
  }, t.TEXT_CONTAINS = {
    label: "sheets-filter.conditions.text-contains",
    operator: S.CONTAINS,
    order: N.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: S.CONTAINS, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === S.CONTAINS;
    },
    mapToFilterColumn: (s) => {
      const { val1: o } = s;
      return o === "" ? null : {
        customFilters: { customFilters: [{ val: `*${o}*` }] }
      };
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0], a = o.val.toString();
      return !o.operator && a.startsWith("*") && a.endsWith("*") ? { operator1: S.CONTAINS, val1: a.slice(1, -1) } : !1;
    }
  }, t.DOES_NOT_CONTAIN = {
    label: "sheets-filter.conditions.does-not-contain",
    operator: S.DOES_NOT_CONTAIN,
    order: N.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: S.DOES_NOT_CONTAIN, val1: "" }),
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: `*${s.val1}*`, operator: u.NOT_EQUALS }] }
    }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === S.DOES_NOT_CONTAIN;
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0], a = o.val.toString();
      return o.operator === u.NOT_EQUALS && a.startsWith("*") && a.endsWith("*") ? { operator1: S.DOES_NOT_CONTAIN, val1: a.slice(1, -1) } : !1;
    }
  }, t.STARTS_WITH = {
    label: "sheets-filter.conditions.starts-with",
    operator: S.STARTS_WITH,
    order: N.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: S.STARTS_WITH, val1: "" }),
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: `${s.val1}*` }] }
    }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === S.STARTS_WITH;
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0], a = o.val.toString();
      return !o.operator && a.endsWith("*") && !a.startsWith("*") ? { operator1: S.STARTS_WITH, val1: a.slice(0, -1) } : !1;
    }
  }, t.ENDS_WITH = {
    label: "sheets-filter.conditions.ends-with",
    operator: S.ENDS_WITH,
    order: N.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: S.ENDS_WITH, val1: "" }),
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: `*${s.val1}` }] }
    }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === S.ENDS_WITH;
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0], a = o.val.toString();
      return !o.operator && a.startsWith("*") && !a.endsWith("*") ? { operator1: S.ENDS_WITH, val1: a.slice(1) } : !1;
    }
  }, t.EQUALS = {
    label: "sheets-filter.conditions.equals",
    operator: S.EQUALS,
    order: N.FIRST,
    numOfParameters: 1,
    getDefaultFormParams: () => ({ operator1: S.EQUALS, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === S.EQUALS;
    },
    mapToFilterColumn: (s) => {
      const { val1: o } = s;
      return o === "" ? null : {
        customFilters: { customFilters: [{ val: o }] }
      };
    },
    testMappingFilterColumn: (s) => {
      var o, a, c;
      return ((a = (o = s.filters) == null ? void 0 : o.filters) == null ? void 0 : a.length) === 1 ? { operator1: S.EQUALS, val1: "" } : ((c = s.customFilters) == null ? void 0 : c.customFilters.length) === 1 && !s.customFilters.customFilters[0].operator ? { operator1: S.EQUALS, val1: s.customFilters.customFilters[0].val.toString() } : !1;
    }
  }, t.GREATER_THAN = {
    label: "sheets-filter.conditions.greater-than",
    operator: u.GREATER_THAN,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.GREATER_THAN, val1: "" }),
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.GREATER_THAN }] }
    }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.GREATER_THAN;
    },
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.GREATER_THAN ? !1 : { operator1: u.GREATER_THAN, val1: o.val.toString() };
    }
  }, t.GREATER_THAN_OR_EQUAL = {
    label: "sheets-filter.conditions.greater-than-or-equal",
    operator: u.GREATER_THAN_OR_EQUAL,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.GREATER_THAN_OR_EQUAL, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.GREATER_THAN_OR_EQUAL;
    },
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.GREATER_THAN_OR_EQUAL }] }
    }),
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.GREATER_THAN_OR_EQUAL ? !1 : { operator1: u.GREATER_THAN_OR_EQUAL, val1: o.val.toString() };
    }
  }, t.LESS_THAN = {
    label: "sheets-filter.conditions.less-than",
    operator: u.LESS_THAN,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.LESS_THAN, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.LESS_THAN;
    },
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.LESS_THAN }] }
    }),
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.LESS_THAN ? !1 : { operator1: u.LESS_THAN, val1: o.val.toString() };
    }
  }, t.LESS_THAN_OR_EQUAL = {
    label: "sheets-filter.conditions.less-than-or-equal",
    operator: u.LESS_THAN_OR_EQUAL,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.LESS_THAN_OR_EQUAL, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.LESS_THAN_OR_EQUAL;
    },
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.LESS_THAN_OR_EQUAL }] }
    }),
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.LESS_THAN_OR_EQUAL ? !1 : { operator1: u.LESS_THAN_OR_EQUAL, val1: o.val.toString() };
    }
  }, t.EQUAL = {
    label: "sheets-filter.conditions.equal",
    operator: u.EQUAL,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.EQUAL, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.EQUAL;
    },
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.EQUAL }] }
    }),
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.EQUAL ? !1 : { operator1: u.EQUAL, val1: o.val.toString() };
    }
  }, t.NOT_EQUAL = {
    label: "sheets-filter.conditions.not-equal",
    operator: u.NOT_EQUALS,
    numOfParameters: 1,
    order: N.FIRST,
    getDefaultFormParams: () => ({ operator1: u.NOT_EQUALS, val1: "" }),
    testMappingParams: (s) => {
      const [o] = L(s);
      return o === u.NOT_EQUALS;
    },
    mapToFilterColumn: (s) => ({
      customFilters: { customFilters: [{ val: s.val1, operator: u.NOT_EQUALS }] }
    }),
    testMappingFilterColumn: (s) => {
      var a;
      if (((a = s.customFilters) == null ? void 0 : a.customFilters.length) !== 1)
        return !1;
      const o = s.customFilters.customFilters[0];
      return o.operator !== u.NOT_EQUALS ? !1 : { operator1: u.NOT_EQUALS, val1: o.val.toString() };
    }
  }, t.BETWEEN = {
    label: "sheets-filter.conditions.between",
    operator: S.BETWEEN,
    order: N.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      and: !0,
      operator1: u.GREATER_THAN_OR_EQUAL,
      val1: "",
      operator2: u.LESS_THAN_OR_EQUAL,
      val2: ""
    }),
    testMappingParams: (s) => {
      const { and: o, operator1: a, operator2: c } = s;
      if (!o) return !1;
      const h = [a, c];
      return h.includes(u.GREATER_THAN_OR_EQUAL) && h.includes(u.LESS_THAN_OR_EQUAL);
    },
    mapToFilterColumn: (s) => {
      const { val1: o, val2: a, operator1: c } = s, h = c === u.GREATER_THAN_OR_EQUAL;
      return {
        customFilters: {
          and: nt.TRUE,
          customFilters: [
            { val: h ? o : a, operator: u.GREATER_THAN_OR_EQUAL },
            { val: h ? a : o, operator: u.LESS_THAN_OR_EQUAL }
          ]
        }
      };
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const [o, a] = s.customFilters.customFilters;
      return o.operator === u.GREATER_THAN_OR_EQUAL && a.operator === u.LESS_THAN_OR_EQUAL && s.customFilters.and ? {
        and: !0,
        operator1: u.GREATER_THAN_OR_EQUAL,
        val1: o.val.toString(),
        operator2: u.LESS_THAN_OR_EQUAL,
        val2: a.val.toString()
      } : a.operator === u.GREATER_THAN_OR_EQUAL && o.operator === u.LESS_THAN_OR_EQUAL && s.customFilters.and ? {
        and: !0,
        operator1: u.GREATER_THAN_OR_EQUAL,
        val1: a.val.toString(),
        operator2: u.LESS_THAN_OR_EQUAL,
        val2: o.val.toLocaleString()
      } : !1;
    }
  }, t.NOT_BETWEEN = {
    label: "sheets-filter.conditions.not-between",
    operator: S.NOT_BETWEEN,
    order: N.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      operator1: u.LESS_THAN,
      val1: "",
      operator2: u.GREATER_THAN,
      val2: ""
    }),
    testMappingParams: (s) => {
      const { and: o, operator1: a, operator2: c } = s;
      if (o) return !1;
      const h = [a, c];
      return h.includes(u.GREATER_THAN) && h.includes(u.LESS_THAN);
    },
    mapToFilterColumn: (s) => {
      const { val1: o, val2: a, operator1: c } = s, h = c === u.GREATER_THAN;
      return {
        customFilters: {
          customFilters: [
            { val: h ? o : a, operator: u.GREATER_THAN },
            { val: h ? a : o, operator: u.LESS_THAN }
          ]
        }
      };
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const [o, a] = s.customFilters.customFilters;
      return o.operator === u.LESS_THAN && a.operator === u.GREATER_THAN && !s.customFilters.and ? {
        operator1: u.LESS_THAN,
        val1: o.val.toString(),
        operator2: u.GREATER_THAN,
        val2: a.val.toString()
      } : a.operator === u.LESS_THAN && o.operator === u.GREATER_THAN && !s.customFilters.and ? {
        operator1: u.GREATER_THAN,
        val1: a.val.toString(),
        operator2: u.LESS_THAN,
        val2: o.val.toLocaleString()
      } : !1;
    }
  }, t.CUSTOM = {
    label: "sheets-filter.conditions.custom",
    operator: S.CUSTOM,
    order: N.SECOND,
    numOfParameters: 2,
    getDefaultFormParams: () => ({
      operator1: S.NONE,
      val1: "",
      operator2: S.NONE,
      val2: ""
    }),
    testMappingParams: () => !0,
    mapToFilterColumn: (s) => {
      const { and: o, val1: a, val2: c, operator1: h, operator2: d } = s;
      function g(I, R) {
        for (const M of t.ALL_CONDITIONS)
          if (M.operator === I)
            return M.mapToFilterColumn({ val1: R, operator1: I });
      }
      const m = !h || h === t.NONE.operator, T = !d || d === t.NONE.operator;
      if (m && T)
        return t.NONE.mapToFilterColumn({});
      if (m)
        return g(d, c);
      if (T)
        return g(h, a);
      const v = g(h, a), P = g(d, c), F = {
        customFilters: [
          v.customFilters.customFilters[0],
          P.customFilters.customFilters[0]
        ]
      };
      return o && (F.and = nt.TRUE), { customFilters: F };
    },
    testMappingFilterColumn: (s) => {
      var c;
      if (((c = s.customFilters) == null ? void 0 : c.customFilters.length) !== 2)
        return !1;
      const o = s.customFilters.customFilters.map((h) => l({ customFilters: { customFilters: [h] } })), a = {
        operator1: o[0][0].operator,
        val1: o[0][1].val1,
        operator2: o[1][0].operator,
        val2: o[1][1].val1
      };
      return s.customFilters.and && (a.and = !0), a;
    }
  }, t.ALL_CONDITIONS = [
    // ------------------------------
    t.NONE,
    // ------------------------------
    t.EMPTY,
    t.NOT_EMPTY,
    // ------------------------------
    t.TEXT_CONTAINS,
    t.DOES_NOT_CONTAIN,
    t.STARTS_WITH,
    t.ENDS_WITH,
    t.EQUALS,
    // ------------------------------
    t.GREATER_THAN,
    t.GREATER_THAN_OR_EQUAL,
    t.LESS_THAN,
    t.LESS_THAN_OR_EQUAL,
    t.EQUAL,
    t.NOT_EQUAL,
    t.BETWEEN,
    t.NOT_BETWEEN,
    // ------------------------------
    t.CUSTOM
  ];
  function e(s) {
    const o = t.ALL_CONDITIONS.find((a) => a.operator === s);
    if (!o)
      throw new Error(`[SheetsFilter]: no condition item found for operator: ${s}`);
    return o;
  }
  t.getItemByOperator = e;
  function r(s, o) {
    for (const a of t.ALL_CONDITIONS.filter((c) => c.numOfParameters === o))
      if (a.numOfParameters !== 0 && a.testMappingParams(s))
        return a;
    for (const a of t.ALL_CONDITIONS)
      if (a.testMappingParams(s))
        return a;
    throw new Error("[SheetsFilter]: no condition item can be mapped from the filter map params!");
  }
  t.testMappingParams = r;
  function n(s) {
    const o = t.ALL_CONDITIONS.find((a) => a.operator === s);
    return (o == null ? void 0 : o.numOfParameters) === 0 ? { operator1: o.operator } : o.getDefaultFormParams();
  }
  t.getInitialFormParams = n;
  function i(s, o) {
    return s.mapToFilterColumn(o);
  }
  t.mapToFilterColumn = i;
  function l(s) {
    if (!s)
      return [t.NONE, {}];
    for (const o of t.ALL_CONDITIONS) {
      const a = o.testMappingFilterColumn(s);
      if (a)
        return [o, a];
    }
    return [t.NONE, {}];
  }
  t.testMappingFilterColumn = l;
})(f || (f = {}));
function L(t) {
  const { operator1: e, operator2: r, val1: n, val2: i } = t;
  if (e && r)
    throw new Error("Both operator1 and operator2 are set!");
  if (!e && !r)
    throw new Error("Neither operator1 and operator2 and both not set!");
  return e ? [e, n] : [r, i];
}
function je(t) {
  const e = [], r = [];
  let n = 0, i = 0;
  function l(s) {
    s.leaf && (s.checked ? (e.push(s), n += s.count) : (r.push(s), i += s.count)), s.children && s.children.forEach(l);
  }
  return t.forEach(l), {
    checkedItems: e,
    uncheckedItems: r,
    checked: n,
    unchecked: i
  };
}
var ts = Object.defineProperty, rs = Object.getOwnPropertyDescriptor, ss = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? rs(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && ts(e, r, i), i;
}, De = (t, e) => (r, n) => e(r, n, t);
const tt = "sheets-filter.generate-filter-values.service", Pe = gt(tt), ns = ["yyyy-mm-dd", "yyyy-mm-dd;@", "yyyy/mm/dd;@", "yyyy/mm/dd hh:mm", "yyyy-m-d am/pm h:mm", "yyyy-MM-dd", "yyyy/MM/dd", "yyyy/mm/dd", 'yyyy"年"MM"月"dd"日"', "MM-dd", 'M"月"d"日"', "MM-dd A/P hh:mm"];
let We = class extends pe {
  constructor(t, e, r) {
    super(), this._localeService = t, this._univerInstanceService = e, this._logService = r;
  }
  async getFilterValues(t) {
    var h;
    const { unitId: e, subUnitId: r, filteredOutRowsByOtherColumns: n, filters: i, blankChecked: l, iterateRange: s, alreadyChecked: o } = t, a = this._univerInstanceService.getUnit(e), c = (h = this._univerInstanceService.getUnit(e)) == null ? void 0 : h.getSheetBySheetId(r);
    return !a || !c ? [] : (this._logService.debug("[SheetsGenerateFilterValuesService]", "getFilterValues for", { unitId: e, subUnitId: r }), wt(i, this._localeService, s, c, new Set(n), new Set(o.map(String)), l, a.getStyles()));
  }
};
We = ss([
  De(0, E(K)),
  De(1, Ye),
  De(2, Zt)
], We);
function wt(t, e, r, n, i, l, s, o) {
  var v, P, F, I, R, M, $, Se, ge;
  const a = /* @__PURE__ */ new Map(), c = /* @__PURE__ */ new Map(), h = "yyyy-mm-dd", d = new Set(ns), g = "empty";
  let m = 0;
  for (const y of n.iterateByColumn(r, !1, !1)) {
    const { row: Gt, rowSpan: st = 1 } = y;
    let J = 0;
    for (; J < st; ) {
      const Yt = Gt + J;
      if (i.has(Yt)) {
        J++;
        continue;
      }
      const Q = y != null && y.value ? Jt(y.value) : "";
      if (!Q) {
        m += 1, J += st;
        continue;
      }
      const ve = (v = y.value) != null && v.v && !y.value.p ? (I = (F = o.get((P = y.value) == null ? void 0 : P.s)) == null ? void 0 : F.n) == null ? void 0 : I.pattern : "", qt = ve && it.isDate(ve);
      if (ve && qt && d.has(ve)) {
        const x = (R = n.getCellRaw(y.row, y.col)) == null ? void 0 : R.v;
        if (!x) {
          J++;
          continue;
        }
        const z = it.format(h, x), [O, w, oe] = z.split("-").map(Number);
        let G = a.get(`${O}`);
        G || (G = {
          title: `${O}`,
          key: `${O}`,
          children: [],
          count: 0,
          leaf: !1,
          checked: !1
        }, a.set(`${O}`, G), c.set(`${O}`, [`${O}`]));
        let U = (M = G.children) == null ? void 0 : M.find((He) => He.key === `${O}-${w}`);
        U || (U = {
          title: e.t(`sheets-filter.date.${w}`),
          key: `${O}-${w}`,
          children: [],
          count: 0,
          leaf: !1,
          checked: !1
        }, ($ = G.children) == null || $.push(U), c.set(`${O}-${w}`, [`${O}`, `${O}-${w}`]));
        const Ue = (Se = U == null ? void 0 : U.children) == null ? void 0 : Se.find((He) => He.key === `${O}-${w}-${oe}`);
        Ue ? (Ue.originValues.add(Q), Ue.count++, U.count++, G.count++) : ((ge = U.children) == null || ge.push({
          title: `${oe}`,
          key: `${O}-${w}-${oe}`,
          count: 1,
          originValues: /* @__PURE__ */ new Set([Q]),
          leaf: !0,
          checked: l.size ? l.has(Q) : !s
        }), U.count++, G.count++, c.set(`${O}-${w}-${oe}`, [`${O}`, `${O}-${w}`, `${O}-${w}-${oe}`]));
      } else {
        const x = Q;
        let z = a.get(x);
        z ? z.count++ : (z = {
          title: Q,
          leaf: !0,
          checked: l.size ? l.has(Q) : !s,
          key: x,
          count: 1
        }, a.set(x, z), c.set(x, [x]));
      }
      J++;
    }
  }
  const T = t ? s : !0;
  if (m > 0) {
    const y = {
      title: e.t("sheets-filter.panel.empty"),
      count: m,
      leaf: !0,
      checked: T,
      key: g
    };
    a.set("empty", y), c.set("empty", [g]);
  }
  return {
    filterTreeItems: is(Array.from(a.values())),
    filterTreeMapCache: c
  };
}
function is(t) {
  return Array.from(t).sort((e, r) => e.children && !r.children ? -1 : !e.children && r.children ? 1 : r.title.localeCompare(e.title)).map((e) => (e.children && e.children.sort((r, n) => {
    const i = Number.parseInt(r.key.split("-")[1], 10), l = Number.parseInt(n.key.split("-")[1], 10);
    return i - l;
  }).forEach((r) => {
    r.children && r.children.sort((n, i) => {
      const l = Number.parseInt(n.key.split("-")[2], 10), s = Number.parseInt(i.key.split("-")[2], 10);
      return l - s;
    });
  }), e));
}
function ke(t, e) {
  for (const r of t) {
    if (r.key === e)
      return r;
    if (r.children) {
      const n = ke(r.children, e);
      if (n)
        return n;
    }
  }
  return null;
}
function Ut(t) {
  return t.leaf ? t.checked : t.children ? t.children.every((e) => Ut(e)) : !0;
}
function he(t, e) {
  t.leaf && (e !== void 0 ? t.checked = e : t.checked = !t.checked), t.children && t.children.forEach((r) => he(r, e));
}
function Ht(t, e) {
  const r = [];
  return t.forEach((n) => {
    const i = n.originValues ? e.some(
      (o) => Array.from(n.originValues).some(
        (a) => a.toLowerCase().includes(o.toLowerCase())
      )
    ) : !1, l = !i && e.some(
      (o) => n.title.toLowerCase().includes(o.toLowerCase())
    );
    if (i || l)
      r.push({ ...n });
    else if (n.children) {
      const o = Ht(n.children, e);
      if (o.length > 0) {
        const a = o.reduce((c, h) => c + h.count, 0);
        r.push({ ...n, count: a, children: o });
      }
    }
  }), r;
}
var os = Object.defineProperty, as = Object.getOwnPropertyDescriptor, rt = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? as(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && os(e, r, i), i;
}, Oe = (t, e) => (r, n) => e(r, n, t), Re = /* @__PURE__ */ ((t) => (t[t.VALUES = 0] = "VALUES", t[t.CONDITIONS = 1] = "CONDITIONS", t))(Re || {});
gt("sheets-filter-ui.sheets-filter-panel.service");
let j = class extends pe {
  constructor(e, r) {
    super();
    _(this, "_filterBy$", new q(
      0
      /* VALUES */
    ));
    _(this, "filterBy$", this._filterBy$.asObservable());
    _(this, "_filterByModel$", new xr(1));
    _(this, "filterByModel$", this._filterByModel$.asObservable());
    _(this, "_filterByModel", null);
    _(this, "_hasCriteria$", new q(!1));
    _(this, "hasCriteria$", this._hasCriteria$.asObservable());
    _(this, "_filterModel", null);
    _(this, "_col$", new q(-1));
    _(this, "col$", this._col$.asObservable());
    _(this, "_filterHeaderListener", null);
    this._injector = e, this._refRangeService = r;
  }
  get filterBy() {
    return this._filterBy$.getValue();
  }
  get filterByModel() {
    return this._filterByModel;
  }
  set filterByModel(e) {
    this._filterByModel = e, this._filterByModel$.next(e);
  }
  get filterModel() {
    return this._filterModel;
  }
  get col() {
    return this._col$.getValue();
  }
  dispose() {
    this._filterBy$.complete(), this._filterByModel$.complete(), this._hasCriteria$.complete();
  }
  setupCol(e, r) {
    this.terminate(), this._filterModel = e, this._col$.next(r);
    const n = e.getFilterColumn(r);
    if (n) {
      const i = n.getColumnData();
      if (i.customFilters) {
        this._hasCriteria$.next(!0), this._setupByConditions(e, r);
        return;
      }
      if (i.filters) {
        this._hasCriteria$.next(!0), this._setupByValues(e, r);
        return;
      }
      this._hasCriteria$.next(!1), this._setupByValues(e, r);
      return;
    }
    this._hasCriteria$.next(!1), this._setupByValues(e, r);
  }
  changeFilterBy(e) {
    return !this._filterModel || this.col === -1 ? !1 : (e === 0 ? this._setupByValues(this._filterModel, this.col) : this._setupByConditions(this._filterModel, this.col), !0);
  }
  terminate() {
    return this._filterModel = null, this._col$.next(-1), this._disposeFilterHeaderChangeListener(), !0;
  }
  _disposeFilterHeaderChangeListener() {
    var e;
    (e = this._filterHeaderListener) == null || e.dispose(), this._filterHeaderListener = null;
  }
  _listenToFilterHeaderChange(e, r) {
    this._disposeFilterHeaderChangeListener();
    const n = e.unitId, i = e.subUnitId, l = e.getRange(), s = {
      startColumn: r,
      startRow: l.startRow,
      endRow: l.startRow,
      endColumn: r
    };
    this._filterHeaderListener = this._refRangeService.watchRange(n, i, s, (o, a) => {
      if (!a)
        this.terminate();
      else {
        const c = a.startColumn - o.startColumn;
        c !== 0 && this._filterByModel.deltaCol(c);
      }
    });
  }
  async _setupByValues(e, r) {
    this._disposePreviousModel();
    const n = e.getRange();
    if (n.startRow === n.endRow) return !1;
    const i = await ye.fromFilterColumn(
      this._injector,
      e,
      r
    );
    return this.filterByModel = i, this._filterBy$.next(
      0
      /* VALUES */
    ), this._listenToFilterHeaderChange(e, r), !0;
  }
  _setupByConditions(e, r) {
    this._disposePreviousModel();
    const n = e.getRange();
    if (n.startRow === n.endRow) return !1;
    const i = Ie.fromFilterColumn(
      this._injector,
      e,
      r,
      e.getFilterColumn(r)
    );
    return this.filterByModel = i, this._filterBy$.next(
      1
      /* CONDITIONS */
    ), this._listenToFilterHeaderChange(e, r), !0;
  }
  _disposePreviousModel() {
    var e;
    (e = this._filterByModel) == null || e.dispose(), this.filterByModel = null;
  }
};
j = rt([
  Oe(0, E(X)),
  Oe(1, E(hr))
], j);
let Ie = class extends pe {
  constructor(e, r, n, i, l) {
    super();
    _(this, "canApply$", _e(!0));
    _(this, "_conditionItem$");
    _(this, "conditionItem$");
    _(this, "_filterConditionFormParams$");
    _(this, "filterConditionFormParams$");
    this._filterModel = e, this.col = r, this._commandService = l, this._conditionItem$ = new q(n), this.conditionItem$ = this._conditionItem$.asObservable(), this._filterConditionFormParams$ = new q(i), this.filterConditionFormParams$ = this._filterConditionFormParams$.asObservable();
  }
  /**
   * Create a model with targeting filter column. If there is not a filter column, the model would be created with
   * default values.
   *
   * @param injector
   * @param filterModel
   * @param col
   * @param filterColumn
   *
   * @returns the model to control the panel's state
   */
  static fromFilterColumn(e, r, n, i) {
    const [l, s] = f.testMappingFilterColumn(i == null ? void 0 : i.getColumnData());
    return e.createInstance(Ie, r, n, l, s);
  }
  get conditionItem() {
    return this._conditionItem$.getValue();
  }
  get filterConditionFormParams() {
    return this._filterConditionFormParams$.getValue();
  }
  dispose() {
    super.dispose(), this._conditionItem$.complete(), this._filterConditionFormParams$.complete();
  }
  deltaCol(e) {
    this.col += e;
  }
  clear() {
    return this._disposed ? Promise.resolve(!1) : this._commandService.executeCommand(re.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: null
    });
  }
  /**
   * Apply the filter condition to the target filter column.
   */
  async apply() {
    if (this._disposed) return !1;
    const e = f.mapToFilterColumn(this.conditionItem, this.filterConditionFormParams);
    return this._commandService.executeCommand(re.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: e
    });
  }
  /**
   * This method would be called when user changes the primary condition. The model would load the corresponding
   * `IFilterConditionFormParams` and load default condition form params.
   */
  onPrimaryConditionChange(e) {
    const r = f.ALL_CONDITIONS.find((n) => n.operator === e);
    if (!r)
      throw new Error(`[ByConditionsModel]: condition item not found for operator: ${e}!`);
    this._conditionItem$.next(r), this._filterConditionFormParams$.next(f.getInitialFormParams(e));
  }
  /**
   * This method would be called when user changes the primary conditions, the input values or "AND" "OR" ratio.
   * If the primary conditions or the ratio is changed, the method would load the corresponding `IFilterCondition`.
   *
   * When the panel call this method, it only has to pass the changed keys.
   *
   * @param params
   */
  onConditionFormChange(e) {
    const r = { ...this.filterConditionFormParams, ...e };
    if (r.and !== !0 && delete r.and, typeof e.and < "u" || typeof e.operator1 < "u" || typeof e.operator2 < "u") {
      const n = f.testMappingParams(r, this.conditionItem.numOfParameters);
      this._conditionItem$.next(n);
    }
    this._filterConditionFormParams$.next(r);
  }
};
Ie = rt([
  Oe(4, W)
], Ie);
let ye = class extends pe {
  constructor(e, r, n, i, l) {
    super();
    _(this, "_rawFilterItems$");
    _(this, "rawFilterItems$");
    _(this, "filterItems$");
    _(this, "_filterItems", []);
    _(this, "_treeMapCache");
    _(this, "canApply$");
    _(this, "_manuallyUpdateFilterItems$");
    _(this, "_searchString$");
    _(this, "searchString$");
    this._filterModel = e, this.col = r, this._commandService = l, this._treeMapCache = i, this._searchString$ = new q(""), this.searchString$ = this._searchString$.asObservable(), this._rawFilterItems$ = new q(n), this.rawFilterItems$ = this._rawFilterItems$.asObservable(), this._manuallyUpdateFilterItems$ = new Br(), this.filterItems$ = Vr(
      jr([
        this._searchString$.pipe(
          It(500, void 0, { leading: !0, trailing: !0 }),
          yt(void 0)
        ),
        this._rawFilterItems$
      ]).pipe(
        se(([s, o]) => {
          if (!s) return o;
          const c = s.toLowerCase().split(/\s+/).filter((h) => !!h);
          return Ht(o, c);
        })
      ),
      this._manuallyUpdateFilterItems$
    ).pipe(Wr(1)), this.canApply$ = this.filterItems$.pipe(se((s) => je(s).checked > 0)), this.disposeWithMe(this.filterItems$.subscribe((s) => this._filterItems = s));
  }
  /**
   * Create a model with targeting filter column. If there is not a filter column, the model would be created with
   * default values.
   *
   * @param injector
   * @param filterModel
   * @param col
   *
   * @returns the model to control the panel's state
   */
  static async fromFilterColumn(e, r, n) {
    var M;
    const i = e.get(Ye), l = e.get(K), s = e.get(Pe, vt.OPTIONAL), { unitId: o, subUnitId: a } = r, c = i.getUniverSheetInstance(o);
    if (!c) throw new Error(`[ByValuesModel]: Workbook not found for filter model with unitId: ${o}!`);
    const h = c == null ? void 0 : c.getSheetBySheetId(a);
    if (!h) throw new Error(`[ByValuesModel]: Worksheet not found for filter model with unitId: ${o} and subUnitId: ${a}!`);
    const d = r.getRange(), g = n, m = (M = r.getFilterColumn(n)) == null ? void 0 : M.getColumnData().filters, T = new Set(m == null ? void 0 : m.filters), v = !!(m && m.blank), P = r.getFilteredOutRowsExceptCol(n), F = { ...d, startRow: d.startRow + 1, startColumn: g, endColumn: g };
    let I, R;
    if (s) {
      const $ = await s.getFilterValues({
        unitId: o,
        subUnitId: a,
        filteredOutRowsByOtherColumns: Array.from(P),
        filters: !!m,
        blankChecked: v,
        iterateRange: F,
        alreadyChecked: Array.from(T)
      });
      I = $.filterTreeItems, R = $.filterTreeMapCache;
    } else {
      const $ = wt(!!m, l, F, h, P, T, v, c.getStyles());
      I = $.filterTreeItems, R = $.filterTreeMapCache;
    }
    return e.createInstance(ye, r, n, I, R);
  }
  get rawFilterItems() {
    return this._rawFilterItems$.getValue();
  }
  get filterItems() {
    return this._filterItems;
  }
  get treeMapCache() {
    return this._treeMapCache;
  }
  dispose() {
    this._rawFilterItems$.complete(), this._searchString$.complete();
  }
  deltaCol(e) {
    this.col += e;
  }
  setSearchString(e) {
    this._searchString$.next(e);
  }
  onCheckAllToggled(e) {
    const r = Fe.deepClone(this._filterItems);
    r.forEach((n) => he(n, e)), this._manuallyUpdateFilterItems(r);
  }
  /**
   * Toggle a filter item.
   */
  onFilterCheckToggled(e) {
    const r = Fe.deepClone(this._filterItems), n = ke(r, e.key);
    if (!n)
      return;
    const i = Ut(n);
    he(n, !i), this._manuallyUpdateFilterItems(r);
  }
  onFilterOnly(e) {
    const r = Fe.deepClone(this._filterItems);
    r.forEach((n) => he(n, !1)), e.forEach((n) => {
      const i = ke(r, n);
      i && he(i, !0);
    }), this._manuallyUpdateFilterItems(r);
  }
  _manuallyUpdateFilterItems(e) {
    this._manuallyUpdateFilterItems$.next(e);
  }
  // expose method here to let the panel change filter items
  // #region ByValuesModel apply methods
  clear() {
    return this._disposed ? Promise.resolve(!1) : this._commandService.executeCommand(re.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: null
    });
  }
  /**
   * Apply the filter condition to the target filter column.
   */
  async apply() {
    if (this._disposed)
      return !1;
    const e = je(this._filterItems), { checked: r, checkedItems: n } = e, i = this.rawFilterItems;
    let l = 0;
    for (const c of i)
      l += c.count;
    const s = r === 0, o = e.checked === l, a = { colId: this.col };
    if (s)
      throw new Error("[ByValuesModel]: no checked items!");
    if (o)
      return this._commandService.executeCommand(re.id, {
        unitId: this._filterModel.unitId,
        subUnitId: this._filterModel.subUnitId,
        col: this.col,
        criteria: null
      });
    {
      a.filters = {};
      const c = n.filter((d) => d.key !== "empty");
      c.length > 0 && (a.filters = {
        filters: c.flatMap((d) => d.originValues ? Array.from(d.originValues) : [d.title])
      }), c.length !== n.length && (a.filters.blank = !0);
    }
    return this._commandService.executeCommand(re.id, {
      unitId: this._filterModel.unitId,
      subUnitId: this._filterModel.subUnitId,
      col: this.col,
      criteria: a
    });
  }
  // #endregion
};
ye = rt([
  Oe(4, W)
], ye);
const ne = "FILTER_PANEL_OPENED", Ae = {
  id: "sheet.operation.open-filter-panel",
  type: $e.OPERATION,
  handler: (t, e) => {
    const r = t.get(we), n = t.get(k), i = t.get(j);
    t.get(W).syncExecuteCommand(Ir.id, { visible: !1 });
    const { unitId: s, subUnitId: o, col: a } = e, c = n.getFilterModel(s, o);
    return c ? (i.setupCol(c, a), r.getContextValue(ne) || r.setContextValue(ne, !0), !0) : !1;
  }
}, de = {
  id: "sheet.operation.close-filter-panel",
  type: $e.OPERATION,
  handler: (t) => {
    const e = t.get(we), r = t.get(j), n = t.get(vr, vt.OPTIONAL);
    return e.getContextValue(ne) ? (e.setContextValue(ne, !1), n == null || n.focus(), r.terminate()) : !1;
  }
}, Dt = {
  id: "sheet.operation.apply-filter",
  type: $e.OPERATION,
  handler: (t, e) => {
    const { filterBy: r } = e;
    return t.get(j).changeFilterBy(r);
  }
};
var ls = Object.defineProperty, cs = Object.getOwnPropertyDescriptor, us = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? cs(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && ls(e, r, i), i;
}, ee = (t, e) => (r, n) => e(r, n, t);
let ie = class extends pe {
  constructor(t, e, r, n, i, l) {
    super(), this._sheetsFilterService = t, this._localeService = e, this._commandService = r, this._sheetPermissionCheckPermission = n, this._injector = i, this._sheetsSelectionService = l, this._commandExecutedListener();
  }
  _commandExecutedListener() {
    this.disposeWithMe(
      this._commandService.beforeCommandExecuted((t) => {
        var e, r, n;
        if (t.id === me.id) {
          const i = this._injector.get(Ye), l = pr(i);
          if (!l) return;
          const { unitId: s, subUnitId: o, worksheet: a } = l, c = (e = this._sheetsFilterService.getFilterModel(s, o)) == null ? void 0 : e.getRange();
          let h;
          if (c)
            h = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
              rangeTypes: [le],
              worksheetTypes: [ce, ue]
            }, [c]);
          else {
            const d = (r = this._sheetsSelectionService.getCurrentLastSelection()) == null ? void 0 : r.range;
            if (d) {
              let g = { ...d };
              g = d.startColumn === d.endColumn && d.startRow === d.endRow ? mr(g, { left: !0, right: !0, up: !0, down: !0 }, a) : g, h = this._sheetPermissionCheckPermission.permissionCheckWithRanges({
                rangeTypes: [le],
                worksheetTypes: [ue, ce]
              }, [g], s, o);
            } else
              h = this._sheetPermissionCheckPermission.permissionCheckWithoutRange({
                rangeTypes: [le],
                worksheetTypes: [ue, ce]
              });
          }
          h || this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("permission.dialog.filterErr"));
        }
        if (t.id === Ae.id) {
          const i = t.params, { unitId: l, subUnitId: s } = i, o = (n = this._sheetsFilterService.getFilterModel(l, s)) == null ? void 0 : n.getRange(), a = Fe.deepClone(o);
          a && (a.startColumn = i.col, a.endColumn = i.col, this._sheetPermissionCheckPermission.permissionCheckWithRanges({
            rangeTypes: [le],
            worksheetTypes: [ce, ue]
          }, [a]) || this._sheetPermissionCheckPermission.blockExecuteWithoutPermission(this._localeService.t("permission.dialog.filterErr")));
        }
      })
    );
  }
};
ie = us([
  ee(0, E(k)),
  ee(1, E(K)),
  ee(2, W),
  ee(3, E(dr)),
  ee(4, E(X)),
  ee(5, E(fr))
], ie);
var b = function() {
  return b = Object.assign || function(t) {
    for (var e, r = 1, n = arguments.length; r < n; r++) {
      e = arguments[r];
      for (var i in e) Object.prototype.hasOwnProperty.call(e, i) && (t[i] = e[i]);
    }
    return t;
  }, b.apply(this, arguments);
}, hs = function(t, e) {
  var r = {};
  for (var n in t) Object.prototype.hasOwnProperty.call(t, n) && e.indexOf(n) < 0 && (r[n] = t[n]);
  if (t != null && typeof Object.getOwnPropertySymbols == "function")
    for (var i = 0, n = Object.getOwnPropertySymbols(t); i < n.length; i++)
      e.indexOf(n[i]) < 0 && Object.prototype.propertyIsEnumerable.call(t, n[i]) && (r[n[i]] = t[n[i]]);
  return r;
}, xt = Lt(function(t, e) {
  var r = t.icon, n = t.id, i = t.className, l = t.extend, s = hs(t, ["icon", "id", "className", "extend"]), o = "univerjs-icon univerjs-icon-".concat(n, " ").concat(i || "").trim(), a = es("_".concat(ps()));
  return Bt(r, "".concat(n), { defIds: r.defIds, idSuffix: a.current }, b({ ref: e, className: o }, s), l);
});
function Bt(t, e, r, n, i) {
  return Mt(t.tag, b(b({ key: e }, ds(t, r, i)), n), (fs(t, r).children || []).map(function(l, s) {
    return Bt(l, "".concat(e, "-").concat(t.tag, "-").concat(s), r, void 0, i);
  }));
}
function ds(t, e, r) {
  var n = b({}, t.attrs);
  r != null && r.colorChannel1 && n.fill === "colorChannel1" && (n.fill = r.colorChannel1);
  var i = e.defIds;
  return !i || i.length === 0 || (t.tag === "use" && n["xlink:href"] && (n["xlink:href"] = n["xlink:href"] + e.idSuffix), Object.entries(n).forEach(function(l) {
    var s = l[0], o = l[1];
    typeof o == "string" && (n[s] = o.replace(/url\(#(.*)\)/, "url(#$1".concat(e.idSuffix, ")")));
  })), n;
}
function fs(t, e) {
  var r, n = e.defIds;
  return !n || n.length === 0 ? t : t.tag === "defs" && (!((r = t.children) === null || r === void 0) && r.length) ? b(b({}, t), { children: t.children.map(function(i) {
    return typeof i.attrs.id == "string" && n && n.indexOf(i.attrs.id) > -1 ? b(b({}, i), { attrs: b(b({}, i.attrs), { id: i.attrs.id + e.idSuffix }) }) : i;
  }) }) : t;
}
function ps() {
  return Math.random().toString(36).substring(2, 8);
}
xt.displayName = "UniverIcon";
var ms = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M3.32182 2.60967C2.98161 2.60967 2.79671 3.0074 3.01601 3.2675L6.85819 7.8246C6.94943 7.93282 6.99947 8.06981 6.99947 8.21136V12.7338C6.99947 12.898 7.0998 13.0455 7.2525 13.1058L8.73833 13.6928C9.00085 13.7965 9.28531 13.6031 9.28531 13.3208V8.21136C9.28531 8.06981 9.33535 7.93282 9.42659 7.8246L13.2688 3.2675C13.4881 3.0074 13.3032 2.60967 12.963 2.60967H3.32182ZM2.09858 4.04101C1.22139 3.0006 1.96097 1.40967 3.32182 1.40967H12.963C14.3238 1.40967 15.0634 3.0006 14.1862 4.04101L10.4853 8.43054V13.3208C10.4853 14.4498 9.34747 15.2237 8.29742 14.8089L6.81158 14.2219C6.20078 13.9806 5.79947 13.3905 5.79947 12.7338V8.43054L2.09858 4.04101Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, Vt = Lt(function(t, e) {
  return Mt(xt, Object.assign({}, t, {
    id: "filter-single",
    ref: e,
    icon: ms
  }));
});
Vt.displayName = "FilterSingle";
var xe = { exports: {} }, ae = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ut;
function _s() {
  if (ut) return ae;
  ut = 1;
  var t = zr, e = Symbol.for("react.element"), r = Symbol.for("react.fragment"), n = Object.prototype.hasOwnProperty, i = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, l = { key: !0, ref: !0, __self: !0, __source: !0 };
  function s(o, a, c) {
    var h, d = {}, g = null, m = null;
    c !== void 0 && (g = "" + c), a.key !== void 0 && (g = "" + a.key), a.ref !== void 0 && (m = a.ref);
    for (h in a) n.call(a, h) && !l.hasOwnProperty(h) && (d[h] = a[h]);
    if (o && o.defaultProps) for (h in a = o.defaultProps, a) d[h] === void 0 && (d[h] = a[h]);
    return { $$typeof: e, type: o, key: g, ref: m, props: d, _owner: i.current };
  }
  return ae.Fragment = r, ae.jsx = s, ae.jsxs = s, ae;
}
var ht;
function Ss() {
  return ht || (ht = 1, xe.exports = _s()), xe.exports;
}
var p = Ss();
const gs = "univer-sheets-filter-panel", vs = "univer-sheets-filter-panel-header", Es = "univer-sheets-filter-panel-content", Ts = "univer-sheets-filter-panel-select-all", Fs = "univer-sheets-filter-panel-select-all-count", Cs = "univer-sheets-filter-panel-values-container", Ns = "univer-input-affix-wrapper", Ps = "univer-select", Os = "univer-radio-group", Rs = "univer-sheets-filter-panel-values-find", Is = "univer-sheets-filter-panel-values-list", ys = "univer-sheets-filter-panel-values-list-inner-container", As = "univer-sheets-filter-panel-values-virtual", bs = "univer-sheets-filter-panel-values-item", Ls = "univer-sheets-filter-panel-values-item-inner", Ms = "univer-sheets-filter-panel-values-item-count", $s = "univer-sheets-filter-panel-values-item-exclude-button", ws = "univer-sheets-filter-panel-values-item-text", Us = "univer-sheets-filter-panel-conditions-container", Hs = "univer-sheets-filter-panel-conditions-container-inner", Ds = "univer-sheets-filter-panel-conditions-desc", xs = "univer-sheets-filter-panel-footer", Bs = "univer-sheets-filter-panel-footer-primary-buttons", Vs = "univer-button", js = "univer-input", Ws = "univer-form-dual-column-layout", ks = "univer-sheets-filter-tree-node", Qs = "univer-sheets-filter-tree-node-filter-only", Gs = "univer-sheets-filter-tree-node-attach", C = {
  sheetsFilterPanel: gs,
  sheetsFilterPanelHeader: vs,
  sheetsFilterPanelContent: Es,
  sheetsFilterPanelSelectAll: Ts,
  sheetsFilterPanelSelectAllCount: Fs,
  sheetsFilterPanelValuesContainer: Cs,
  inputAffixWrapper: Ns,
  select: Ps,
  radioGroup: Os,
  sheetsFilterPanelValuesFind: Rs,
  sheetsFilterPanelValuesList: Is,
  sheetsFilterPanelValuesListInnerContainer: ys,
  sheetsFilterPanelValuesVirtual: As,
  sheetsFilterPanelValuesItem: bs,
  sheetsFilterPanelValuesItemInner: Ls,
  sheetsFilterPanelValuesItemCount: Ms,
  sheetsFilterPanelValuesItemExcludeButton: $s,
  sheetsFilterPanelValuesItemText: ws,
  sheetsFilterPanelConditionsContainer: Us,
  sheetsFilterPanelConditionsContainerInner: Hs,
  sheetsFilterPanelConditionsDesc: Ds,
  sheetsFilterPanelFooter: xs,
  sheetsFilterPanelFooterPrimaryButtons: Bs,
  button: Vs,
  input: js,
  formDualColumnLayout: Ws,
  sheetsFilterTreeNode: ks,
  sheetsFilterTreeNodeFilterOnly: Qs,
  sheetsFilterTreeNodeAttach: Gs
};
function Ys(t) {
  var m, T;
  const { model: e } = t, r = te(K), n = V(e.conditionItem$, void 0), i = V(e.filterConditionFormParams$, void 0), l = i != null && i.and ? "AND" : "OR", s = D((v) => {
    e.onConditionFormChange({ and: v === "AND" });
  }, [e]), o = qs(r), a = D((v) => {
    e.onPrimaryConditionChange(v);
  }, [e]), c = Ks(r), h = D((v) => {
    e.onConditionFormChange(v);
  }, [e]), d = r.t("sheets-filter.panel.input-values-placeholder");
  function g(v, P, F) {
    const I = f.getItemByOperator(v).numOfParameters === 1;
    return /* @__PURE__ */ p.jsxs(p.Fragment, { children: [
      F === "operator2" && /* @__PURE__ */ p.jsxs(Yr, { value: l, onChange: s, children: [
        /* @__PURE__ */ p.jsx(lt, { value: "AND", children: r.t("sheets-filter.panel.and") }),
        /* @__PURE__ */ p.jsx(lt, { value: "OR", children: r.t("sheets-filter.panel.or") })
      ] }),
      /* @__PURE__ */ p.jsx(at, { value: v, options: c, onChange: (R) => h({ [F]: R }) }),
      I && /* @__PURE__ */ p.jsx("div", { children: /* @__PURE__ */ p.jsx(At, { value: P, placeholder: d, onChange: (R) => h({ [F === "operator1" ? "val1" : "val2"]: R }) }) })
    ] });
  }
  return /* @__PURE__ */ p.jsx("div", { className: C.sheetsFilterPanelConditionsContainer, children: n && i && /* @__PURE__ */ p.jsxs(p.Fragment, { children: [
    /* @__PURE__ */ p.jsx(at, { value: n.operator, options: o, onChange: a }),
    f.getItemByOperator(n.operator).numOfParameters !== 0 ? /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelConditionsContainerInner, children: [
      n.numOfParameters >= 1 && g(i.operator1, (m = i.val1) != null ? m : "", "operator1"),
      n.numOfParameters >= 2 && g(i.operator2, (T = i.val2) != null ? T : "", "operator2"),
      /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelConditionsDesc, children: [
        r.t("sheets-filter.panel.?"),
        /* @__PURE__ */ p.jsx("br", {}),
        r.t("sheets-filter.panel.*")
      ] })
    ] }) : null
  ] }) });
}
function qs(t) {
  const e = t.getCurrentLocale();
  return et(
    () => [
      {
        options: [
          { label: t.t(f.NONE.label), value: f.NONE.operator }
        ]
      },
      {
        options: [
          { label: t.t(f.EMPTY.label), value: f.EMPTY.operator },
          { label: t.t(f.NOT_EMPTY.label), value: f.NOT_EMPTY.operator }
        ]
      },
      {
        options: [
          { label: t.t(f.TEXT_CONTAINS.label), value: f.TEXT_CONTAINS.operator },
          { label: t.t(f.DOES_NOT_CONTAIN.label), value: f.DOES_NOT_CONTAIN.operator },
          { label: t.t(f.STARTS_WITH.label), value: f.STARTS_WITH.operator },
          { label: t.t(f.ENDS_WITH.label), value: f.ENDS_WITH.operator },
          { label: t.t(f.EQUALS.label), value: f.EQUALS.operator }
        ]
      },
      {
        options: [
          { label: t.t(f.GREATER_THAN.label), value: f.GREATER_THAN.operator },
          { label: t.t(f.GREATER_THAN_OR_EQUAL.label), value: f.GREATER_THAN_OR_EQUAL.operator },
          { label: t.t(f.LESS_THAN.label), value: f.LESS_THAN.operator },
          { label: t.t(f.LESS_THAN_OR_EQUAL.label), value: f.LESS_THAN_OR_EQUAL.operator },
          { label: t.t(f.EQUAL.label), value: f.EQUAL.operator },
          { label: t.t(f.NOT_EQUAL.label), value: f.NOT_EQUAL.operator },
          { label: t.t(f.BETWEEN.label), value: f.BETWEEN.operator },
          { label: t.t(f.NOT_BETWEEN.label), value: f.NOT_BETWEEN.operator }
        ]
      },
      {
        options: [
          { label: t.t(f.CUSTOM.label), value: f.CUSTOM.operator }
        ]
      }
    ],
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [e, t]
  );
}
function Ks(t) {
  const e = t.getCurrentLocale();
  return et(
    () => f.ALL_CONDITIONS.filter((r) => r.numOfParameters !== 2).map((r) => ({ label: t.t(r.label), value: r.operator })),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [e, t]
  );
}
function Xs(t) {
  const { model: e } = t, r = te(K), n = V(e.searchString$, "", !0), i = V(e.filterItems$, void 0, !0), l = r.t("sheets-filter.panel.filter-only"), s = je(i), o = s.checked > 0 && s.unchecked === 0, a = s.checked > 0 && s.unchecked > 0, c = e.treeMapCache, h = D(() => {
    e.onCheckAllToggled(!o);
  }, [e, o]), d = D((m) => {
    e.setSearchString(m);
  }, [e]);
  function g(m) {
    let T = [];
    return m.forEach((v) => {
      v.checked && T.push(v.key), v.children && (T = T.concat(g(v.children)));
    }), T;
  }
  return /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelValuesContainer, children: [
    /* @__PURE__ */ p.jsx(At, { autoFocus: !0, value: n, placeholder: r.t("sheets-filter.panel.search-placeholder"), onChange: d }),
    /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelValuesList, children: [
      /* @__PURE__ */ p.jsx("div", { className: C.sheetsFilterPanelValuesItem, children: /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelValuesItemInner, children: [
        /* @__PURE__ */ p.jsx(
          qr,
          {
            indeterminate: a,
            disabled: i.length === 0,
            checked: o,
            onChange: h
          }
        ),
        /* @__PURE__ */ p.jsx("span", { className: C.sheetsFilterPanelValuesItemText, children: `${r.t("sheets-filter.panel.select-all")}` }),
        /* @__PURE__ */ p.jsx("span", { className: C.sheetsFilterPanelValuesItemCount, children: `(${s.checked}/${s.checked + s.unchecked})` })
      ] }) }),
      /* @__PURE__ */ p.jsx("div", { className: C.sheetsFilterPanelValuesVirtual, children: /* @__PURE__ */ p.jsx(
        Kr,
        {
          data: i,
          defaultExpandAll: !1,
          valueGroup: g(i),
          height: 180,
          onChange: (m) => {
            e.onFilterCheckToggled(m);
          },
          defaultCache: c,
          itemHeight: 28,
          treeNodeClassName: C.sheetsFilterTreeNode,
          attachRender: (m) => /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterTreeNodeAttach, children: [
            /* @__PURE__ */ p.jsx("span", { className: C.sheetsFilterPanelValuesItemCount, children: `(${m.count})` }),
            /* @__PURE__ */ p.jsx(
              Ce,
              {
                className: C.sheetsFilterTreeNodeFilterOnly,
                size: "small",
                type: "link",
                onClick: () => {
                  const T = [];
                  m.children ? m.children.forEach((v) => {
                    v.children ? v.children.forEach((P) => {
                      T.push(P.key);
                    }) : T.push(v.key);
                  }) : T.push(m.key), e.onFilterOnly(T);
                },
                children: l
              }
            )
          ] })
        }
      ) })
    ] })
  ] });
}
function Zs() {
  var P;
  const t = te(j), e = te(K), r = te(W), n = V(t.filterBy$, void 0, !0), i = V(t.filterByModel$, void 0, !1), l = V(() => (i == null ? void 0 : i.canApply$) || _e(!1), void 0, !1, [i]), s = Js(e), o = !V(t.hasCriteria$), a = D((F) => {
    r.executeCommand(Dt.id, { filterBy: F });
  }, [r]), c = D(async () => {
    await (i == null ? void 0 : i.clear()), r.executeCommand(de.id);
  }, [i, r]), h = D(() => {
    r.executeCommand(de.id);
  }, [r]), d = D(async () => {
    await (i == null ? void 0 : i.apply()), r.executeCommand(de.id);
  }, [i, r]), m = (P = te(k).activeFilterModel) == null ? void 0 : P.getRange(), T = t.col, v = Er(yr.FILTER_PANEL_EMBED_POINT);
  return /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanel, children: [
    /* @__PURE__ */ p.jsx(
      Tr,
      {
        components: v,
        sharedProps: { range: m, colIndex: T, onClose: h }
      }
    ),
    /* @__PURE__ */ p.jsx("div", { className: C.sheetsFilterPanelHeader, children: /* @__PURE__ */ p.jsx(Xr, { value: n, options: s, onChange: (F) => a(F) }) }),
    i ? /* @__PURE__ */ p.jsx("div", { className: C.sheetsFilterPanelContent, children: n === Re.VALUES ? /* @__PURE__ */ p.jsx(Xs, { model: i }) : /* @__PURE__ */ p.jsx(Ys, { model: i }) }) : /* @__PURE__ */ p.jsx("div", { style: { flex: 1 } }),
    /* @__PURE__ */ p.jsxs("div", { className: C.sheetsFilterPanelFooter, children: [
      /* @__PURE__ */ p.jsx(Ce, { type: "link", onClick: c, disabled: o, children: e.t("sheets-filter.panel.clear-filter") }),
      /* @__PURE__ */ p.jsxs("span", { className: C.sheetsFilterPanelFooterPrimaryButtons, children: [
        /* @__PURE__ */ p.jsx(Ce, { type: "default", onClick: h, children: e.t("sheets-filter.panel.cancel") }),
        /* @__PURE__ */ p.jsx(Ce, { disabled: !l, type: "primary", onClick: d, children: e.t("sheets-filter.panel.confirm") })
      ] })
    ] })
  ] });
}
function Js(t) {
  const e = t.getCurrentLocale();
  return et(
    () => [
      { label: t.t("sheets-filter.panel.by-values"), value: Re.VALUES },
      { label: t.t("sheets-filter.panel.by-conditions"), value: Re.CONDITIONS }
    ],
    [e, t]
  );
}
function zs(t) {
  const e = t.get(k);
  return {
    id: me.id,
    type: Ze.BUTTON_SELECTOR,
    icon: "FilterSingle",
    tooltip: "sheets-filter.toolbar.smart-toggle-filter-tooltip",
    hidden$: Je(t, Z.UNIVER_SHEET),
    activated$: e.activeFilterModel$.pipe(se((r) => !!r)),
    disabled$: Ar(t, br(t, { worksheetTypes: [ce, ue], rangeTypes: [le] }))
  };
}
function en(t) {
  const e = t.get(k);
  return {
    id: Ke.id,
    type: Ze.BUTTON,
    title: "sheets-filter.toolbar.clear-filter-criteria",
    hidden$: Je(t, Z.UNIVER_SHEET),
    disabled$: e.activeFilterModel$.pipe(ze((r) => {
      var n;
      return (n = r == null ? void 0 : r.hasCriteria$.pipe(se((i) => !i))) != null ? n : _e(!0);
    }))
  };
}
function tn(t) {
  const e = t.get(k);
  return {
    id: Xe.id,
    type: Ze.BUTTON,
    title: "sheets-filter.toolbar.re-calc-filter-conditions",
    hidden$: Je(t, Z.UNIVER_SHEET),
    disabled$: e.activeFilterModel$.pipe(ze((r) => {
      var n;
      return (n = r == null ? void 0 : r.hasCriteria$.pipe(se((i) => !i))) != null ? n : _e(!0);
    }))
  };
}
const rn = {
  [Fr.FORMULAS_INSERT]: {
    [me.id]: {
      order: 10,
      menuItemFactory: zs,
      [Ke.id]: {
        order: 0,
        menuItemFactory: en
      },
      [Xe.id]: {
        order: 1,
        menuItemFactory: tn
      }
    }
  }
}, sn = {
  id: me.id,
  binding: Cr.L | ot.CTRL_COMMAND | ot.SHIFT,
  description: "sheets-filter.shortcut.smart-toggle-filter",
  preconditions: Lr,
  group: "4_sheet-edit"
}, B = 16, nn = new Path2D("M3.30363 3C2.79117 3 2.51457 3.60097 2.84788 3.99024L6.8 8.60593V12.5662C6.8 12.7184 6.8864 12.8575 7.02289 12.9249L8.76717 13.7863C8.96655 13.8847 9.2 13.7396 9.2 13.5173V8.60593L13.1521 3.99024C13.4854 3.60097 13.2088 3 12.6964 3H3.30363Z");
class dt {
  static drawNoCriteria(e, r, n, i) {
    e.save(), ct.drawWith(e, {
      radius: 2,
      width: B,
      height: B,
      fill: i
    }), e.lineCap = "square", e.strokeStyle = n, e.scale(r / B, r / B), e.beginPath(), e.lineWidth = 1, e.lineCap = "round", e.moveTo(3, 4), e.lineTo(13, 4), e.moveTo(4.5, 8), e.lineTo(11.5, 8), e.moveTo(6, 12), e.lineTo(10, 12), e.stroke(), e.restore();
  }
  static drawHasCriteria(e, r, n, i) {
    e.save(), ct.drawWith(e, {
      radius: 2,
      width: B,
      height: B,
      fill: i
    }), e.scale(r / B, r / B), e.fillStyle = n, e.fill(nn), e.restore();
  }
}
var on = Object.defineProperty, an = Object.getOwnPropertyDescriptor, ln = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? an(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && on(e, r, i), i;
}, Be = (t, e) => (r, n) => e(r, n, t);
const H = 16, Ee = 1;
let Qe = class extends Jr {
  constructor(e, r, n, i, l) {
    super(e, r);
    _(this, "_cellWidth", 0);
    _(this, "_cellHeight", 0);
    _(this, "_filterParams");
    _(this, "_hovered", !1);
    this._contextService = n, this._commandService = i, this._themeService = l, this.setShapeProps(r), this.onPointerDown$.subscribeEvent((s) => this.onPointerDown(s)), this.onPointerEnter$.subscribeEvent(() => this.onPointerEnter()), this.onPointerLeave$.subscribeEvent(() => this.onPointerLeave());
  }
  setShapeProps(e) {
    typeof e.cellHeight < "u" && (this._cellHeight = e.cellHeight), typeof e.cellWidth < "u" && (this._cellWidth = e.cellWidth), typeof e.filterParams < "u" && (this._filterParams = e.filterParams), this.transformByState({
      width: e.width,
      height: e.height
    });
  }
  _draw(e) {
    const r = this._cellHeight, n = this._cellWidth, i = H - n, l = H - r;
    e.save();
    const s = new Path2D();
    s.rect(i, l, n, r), e.clip(s);
    const { hasCriteria: o } = this._filterParams, a = this._themeService.getCurrentTheme().primaryColor, c = this._hovered ? this._themeService.getCurrentTheme().grey50 : "rgba(255, 255, 255, 1.0)";
    o ? dt.drawHasCriteria(e, H, a, c) : dt.drawNoCriteria(e, H, a, c), e.restore();
  }
  onPointerDown(e) {
    if (e.button === 2)
      return;
    const { col: r, unitId: n, subUnitId: i } = this._filterParams;
    this._contextService.getContextValue(ne) || !this._commandService.hasCommand(Ae.id) || setTimeout(() => {
      this._commandService.executeCommand(Ae.id, {
        unitId: n,
        subUnitId: i,
        col: r
      });
    }, 200);
  }
  onPointerEnter() {
    this._hovered = !0, this.makeDirty(!0);
  }
  onPointerLeave() {
    this._hovered = !1, this.makeDirty(!0);
  }
};
Qe = ln([
  Be(2, we),
  Be(3, W),
  Be(4, E(Et))
], Qe);
var cn = Object.defineProperty, un = Object.getOwnPropertyDescriptor, hn = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? un(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && cn(e, r, i), i;
}, Y = (t, e) => (r, n) => e(r, n, t);
const dn = 1e3, fn = 5e3;
let Ge = class extends Tt {
  constructor(e, r, n, i, l, s, o, a) {
    super();
    _(this, "_filterRangeShape", null);
    _(this, "_buttonRenderDisposable", null);
    _(this, "_filterButtonShapes", []);
    this._context = e, this._injector = r, this._sheetSkeletonManagerService = n, this._sheetsFilterService = i, this._themeService = l, this._sheetInterceptorService = s, this._commandService = o, this._selectionRenderService = a, this._initRenderer();
  }
  dispose() {
    super.dispose(), this._disposeRendering();
  }
  _initRenderer() {
    this._sheetSkeletonManagerService.currentSkeleton$.pipe(
      ze((e) => {
        var o, a;
        if (!e) return _e(null);
        const { unit: r, unitId: n } = this._context, i = ((o = r.getActiveSheet()) == null ? void 0 : o.getSheetId()) || "", l = (a = this._sheetsFilterService.getFilterModel(n, i)) != null ? a : void 0, s = () => ({
          unitId: n,
          worksheetId: i,
          filterModel: l,
          range: l == null ? void 0 : l.getRange(),
          skeleton: e.skeleton
        });
        return zt(this._commandService.onCommandExecuted.bind(this._commandService)).pipe(
          kr(
            ([c]) => {
              var h;
              return c.type === $e.MUTATION && ((h = c.params) == null ? void 0 : h.unitId) === r.getUnitId() && (nr.has(c.id) || c.id === Sr.id);
            }
          ),
          It(20, void 0, { leading: !1, trailing: !0 }),
          se(s),
          yt(s())
          // must trigger once
        );
      }),
      Qr(this.dispose$)
    ).subscribe((e) => {
      this._disposeRendering(), !(!e || !e.range) && (this._renderRange(e.range, e.skeleton), this._renderButtons(e));
    });
  }
  _renderRange(e, r) {
    const { scene: n } = this._context, { rowHeaderWidth: i, columnHeaderHeight: l } = r, s = this._filterRangeShape = new wr(
      n,
      dn,
      this._themeService,
      {
        rowHeaderWidth: i,
        columnHeaderHeight: l,
        enableAutoFill: !1,
        highlightHeader: !1
      }
    ), a = Ur({
      range: e,
      primary: null,
      style: { fill: "rgba(0, 0, 0, 0.0)" }
    }, r);
    s.updateRangeBySelectionWithCoord(a), s.setEvent(!1), n.makeDirty(!0);
  }
  _renderButtons(e) {
    const { range: r, filterModel: n, unitId: i, skeleton: l, worksheetId: s } = e, { scene: o } = this._context;
    this._interceptCellContent(i, s, e.range);
    const { startColumn: a, endColumn: c, startRow: h } = r;
    for (let d = a; d <= c; d++) {
      const g = `sheets-filter-button-${d}`, m = Hr(h, d, o, l), { startX: T, startY: v, endX: P, endY: F } = m, I = P - T, R = F - v;
      if (R <= Ee || I <= Ee)
        continue;
      const M = !!n.getFilterColumn(d), $ = P - H - Ee, Se = F - H - Ee, ge = {
        left: $,
        top: Se,
        height: H,
        width: H,
        zIndex: fn,
        cellHeight: R,
        cellWidth: I,
        filterParams: { unitId: i, subUnitId: s, col: d, hasCriteria: M }
      }, y = this._injector.createInstance(Qe, g, ge);
      this._filterButtonShapes.push(y);
    }
    o.addObjects(this._filterButtonShapes), o.makeDirty();
  }
  _interceptCellContent(e, r, n) {
    const { startRow: i, startColumn: l, endColumn: s } = n;
    this._buttonRenderDisposable = this._sheetInterceptorService.intercept(gr.CELL_CONTENT, {
      effect: er.Style,
      handler: (o, a, c) => {
        const { row: h, col: d, unitId: g, subUnitId: m } = a;
        return g !== e || m !== r || h !== i || d < l || d > s ? c(o) : c({
          ...o,
          fontRenderExtension: {
            ...o == null ? void 0 : o.fontRenderExtension,
            rightOffset: H
          }
        });
      },
      priority: 10
    });
  }
  _disposeRendering() {
    var e, r;
    (e = this._filterRangeShape) == null || e.dispose(), this._filterButtonShapes.forEach((n) => n.dispose()), (r = this._buttonRenderDisposable) == null || r.dispose(), this._filterRangeShape = null, this._buttonRenderDisposable = null, this._filterButtonShapes = [];
  }
};
Ge = hn([
  Y(1, E(X)),
  Y(2, E(Mr)),
  Y(3, E(k)),
  Y(4, E(Et)),
  Y(5, E(_r)),
  Y(6, W),
  Y(7, $r)
], Ge);
var pn = Object.defineProperty, mn = Object.getOwnPropertyDescriptor, _n = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? mn(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && pn(e, r, i), i;
}, ft = (t, e) => (r, n) => e(r, n, t);
let fe = class extends Tt {
  constructor(t, e) {
    super(), this._renderManagerService = t, this._sheetsRenderService = e, [
      ir,
      or,
      ar,
      lr
    ].forEach((r) => this.disposeWithMe(this._sheetsRenderService.registerSkeletonChangingMutations(r.id))), this.disposeWithMe(this._renderManagerService.registerRenderModule(
      Z.UNIVER_SHEET,
      [Ge]
    ));
  }
};
fe = _n([
  ft(0, bt),
  ft(1, E(Rt))
], fe);
var Sn = Object.defineProperty, gn = Object.getOwnPropertyDescriptor, vn = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? gn(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && Sn(e, r, i), i;
}, A = (t, e) => (r, n) => e(r, n, t);
const pt = "FILTER_PANEL_POPUP";
let be = class extends fe {
  constructor(e, r, n, i, l, s, o, a, c, h, d, g, m) {
    super(m, g);
    _(this, "_popupDisposable");
    this._injector = e, this._componentManager = r, this._sheetsFilterPanelService = n, this._sheetCanvasPopupService = i, this._sheetsFilterService = l, this._localeService = s, this._shortcutService = o, this._commandService = a, this._menuManagerService = c, this._contextService = h, this._messageService = d, this._initCommands(), this._initShortcuts(), this._initMenuItems(), this._initUI();
  }
  dispose() {
    super.dispose(), this._closeFilterPopup();
  }
  _initShortcuts() {
    [
      sn
    ].forEach((e) => {
      this.disposeWithMe(this._shortcutService.registerShortcut(e));
    });
  }
  _initCommands() {
    [
      me,
      cr,
      ur,
      re,
      Ke,
      Xe,
      Dt,
      Ae,
      de
    ].forEach((e) => {
      this.disposeWithMe(this._commandService.registerCommand(e));
    });
  }
  _initMenuItems() {
    this._menuManagerService.mergeMenu(rn);
  }
  _initUI() {
    this.disposeWithMe(this._componentManager.register(pt, Zs)), this.disposeWithMe(this._componentManager.register("FilterSingle", Vt)), this.disposeWithMe(this._contextService.subscribeContextValue$(ne).pipe(Gr()).subscribe((e) => {
      e ? this._openFilterPopup() : this._closeFilterPopup();
    })), this.disposeWithMe(this._sheetsFilterService.errorMsg$.subscribe((e) => {
      e && this._messageService.show({
        type: Zr.Error,
        content: this._localeService.t(e)
      });
    }));
  }
  _openFilterPopup() {
    const e = this._sheetsFilterPanelService.filterModel;
    if (!e)
      throw new Error("[SheetsFilterUIController]: no filter model when opening filter popup!");
    const r = e.getRange(), n = this._sheetsFilterPanelService.col, { startRow: i } = r;
    this._popupDisposable = this._sheetCanvasPopupService.attachPopupToCell(i, n, {
      componentKey: pt,
      direction: "horizontal",
      onClickOutside: () => this._commandService.syncExecuteCommand(de.id),
      offset: [5, 0]
    });
  }
  _closeFilterPopup() {
    var e;
    (e = this._popupDisposable) == null || e.dispose(), this._popupDisposable = null;
  }
};
be = vn([
  A(0, E(X)),
  A(1, E(Nr)),
  A(2, E(j)),
  A(3, E(Dr)),
  A(4, E(k)),
  A(5, E(K)),
  A(6, Pr),
  A(7, W),
  A(8, Or),
  A(9, we),
  A(10, Rr),
  A(11, E(Rt)),
  A(12, bt)
], be);
var jt = Object.defineProperty, En = Object.getOwnPropertyDescriptor, Tn = (t, e, r) => e in t ? jt(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r, Fn = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? En(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && jt(e, r, i), i;
}, Ve = (t, e) => (r, n) => e(r, n, t), Wt = (t, e, r) => Tn(t, typeof e != "symbol" ? e + "" : e, r);
const Cn = "SHEET_FILTER_UI_PLUGIN";
let Le = class extends qe {
  constructor(t = Ne, e, r, n) {
    super(), this._config = t, this._injector = e, this._configService = r, this._rpcChannelService = n;
    const { menu: i, ...l } = Ct(
      {},
      Ne,
      this._config
    );
    i && this._configService.setConfig("menu", i, { merge: !0 }), this._configService.setConfig($t, l);
  }
  onStarting() {
    [
      [j],
      [ie],
      [be]
    ].forEach((t) => this._injector.add(t)), this._config.useRemoteFilterValuesGenerator && this._rpcChannelService && this._injector.add([Pe, {
      useFactory: () => rr(
        this._rpcChannelService.requestChannel(tt)
      )
    }]);
  }
  onReady() {
    this._injector.get(ie);
  }
  onRendered() {
    this._injector.get(be);
  }
};
Wt(Le, "type", Z.UNIVER_SHEET);
Wt(Le, "pluginName", Cn);
Le = Fn([
  Ft(Ot),
  Ve(1, E(X)),
  Ve(2, Nt),
  Ve(3, tr(Pt))
], Le);
var kt = Object.defineProperty, Nn = Object.getOwnPropertyDescriptor, Pn = (t, e, r) => e in t ? kt(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r, On = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? Nn(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && kt(e, r, i), i;
}, mt = (t, e) => (r, n) => e(r, n, t), Qt = (t, e, r) => Pn(t, typeof e != "symbol" ? e + "" : e, r);
const Rn = "SHEET_FILTER_UI_PLUGIN";
let Me = class extends qe {
  constructor(t = Ne, e, r) {
    super(), this._config = t, this._injector = e, this._configService = r;
    const { menu: n, ...i } = Ct(
      {},
      Ne,
      this._config
    );
    n && this._configService.setConfig("menu", n, { merge: !0 }), this._configService.setConfig($t, i);
  }
  onStarting() {
    [
      [ie],
      [fe]
    ].forEach((t) => this._injector.add(t));
  }
  onReady() {
    this._injector.get(ie);
  }
  onRendered() {
    this._injector.get(fe);
  }
};
Qt(Me, "type", Z.UNIVER_SHEET);
Qt(Me, "pluginName", Rn);
Me = On([
  Ft(Ot),
  mt(1, E(X)),
  mt(2, Nt)
], Me);
var In = Object.defineProperty, yn = Object.getOwnPropertyDescriptor, An = (t, e, r, n) => {
  for (var i = n > 1 ? void 0 : n ? yn(e, r) : e, l = t.length - 1, s; l >= 0; l--)
    (s = t[l]) && (i = (n ? s(e, r, i) : s(i)) || i);
  return n && i && In(e, r, i), i;
}, _t = (t, e) => (r, n) => e(r, n, t), Te;
let St = (Te = class extends qe {
  constructor(t, e, r) {
    super(), this._config = t, this._injector = e, this._rpcChannelService = r;
  }
  onStarting() {
    [
      [Pe, { useClass: We }]
    ].forEach((t) => this._injector.add(t));
  }
  onReady() {
    this._rpcChannelService.registerChannel(
      tt,
      sr(this._injector.get(Pe))
    );
  }
}, _(Te, "type", Z.UNIVER_SHEET), _(Te, "pluginName", "SHEET_FILTER_UI_WORKER_PLUGIN"), Te);
St = An([
  _t(1, E(X)),
  _t(2, Pt)
], St);
export {
  Dt as ChangeFilterByOperation,
  de as CloseFilterPanelOperation,
  Ae as OpenFilterPanelOperation,
  Me as UniverSheetsFilterMobileUIPlugin,
  Le as UniverSheetsFilterUIPlugin,
  St as UniverSheetsFilterUIWorkerPlugin
};
