// @univerjs/docs-thread-comment-ui/lib/es/index.js
var Re = Object.defineProperty;
var ye = (e, t, r) => t in e ? Re(e, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : e[t] = r;
var K = (e, t, r) => ye(e, typeof t != "symbol" ? t + "" : t, r);
import { Inject as v, Disposable as B, CommandType as w, ICommandService as M, CustomDecorationType as X, sequenceExecute as se, UniverInstanceType as S, SHEET_EDITOR_UNITS as Te, useDependency as P, IUniverInstanceService as F, Injector as de, isInternalEditorID as ce, useObservable as G, UserManagerService as Ue, BuildTextUtils as Pe, DependentOn as Ee, Plugin as we, merge as Ne, IConfigService as Ae } from "@univerjs/core";
import { IRenderManagerService as L, DocumentEditArea as te } from "@univerjs/engine-render";
import { ThreadCommentPanelService as T, SetActiveCommentOperation as ae, ThreadCommentPanel as $e, UniverThreadCommentUIPlugin as Ve } from "@univerjs/thread-comment-ui";
import { DocSelectionManagerService as H, DocSkeletonManagerService as je, RichTextEditingMutation as me, SetTextSelectionsOperation as Be, DocInterceptorService as Fe, DOC_INTERCEPTOR_POINT as Le } from "@univerjs/docs";
import { addCustomDecorationBySelectionFactory as He, deleteCustomDecorationFactory as We, DocSelectionRenderService as qe, DocBackScrollRenderController as Je, DocRenderController as Ke } from "@univerjs/docs-ui";
import { IThreadCommentDataSourceService as Ge, AddCommentMutation as Ze, getDT as ze, ThreadCommentModel as le } from "@univerjs/thread-comment";
import { ISidebarService as W, MenuItemType as ue, getMenuHiddenObservable as fe, RibbonStartGroup as Ye, ContextMenuPosition as Qe, ContextMenuGroup as Xe, ComponentManager as ke, IMenuManagerService as et } from "@univerjs/ui";
import { BehaviorSubject as tt, Observable as he, debounceTime as Ce, filter as nt } from "rxjs";
import rt, { useMemo as Z, useState as it, useEffect as ot, forwardRef as _e, useRef as st, createElement as pe } from "react";
const dt = "DOC_THREAD_COMMENT_UI_PLUGIN", I = "default_doc", ct = "docs-thread-comment-ui.config", ne = {};
var at = Object.defineProperty, mt = Object.getOwnPropertyDescriptor, lt = (e, t, r, n) => {
  for (var i = n > 1 ? void 0 : n ? mt(t, r) : t, o = e.length - 1, s; o >= 0; o--)
    (s = e[o]) && (i = (n ? s(t, r, i) : s(i)) || i);
  return n && i && at(t, r, i), i;
}, re = (e, t) => (r, n) => t(r, n, e);
let R = class extends B {
  constructor(t, r) {
    super();
    K(this, "_addingComment$", new tt(void 0));
    K(this, "addingComment$", this._addingComment$.asObservable());
    this._sidebarService = t, this._threadCommentPanelService = r, this.disposeWithMe(() => {
      this._addingComment$.complete();
    });
  }
  get addingComment() {
    return this._addingComment$.getValue();
  }
  startAdd(t) {
    this._addingComment$.next(t);
  }
  endAdd() {
    this._addingComment$.next(void 0);
  }
};
R = lt([
  re(0, W),
  re(1, v(T))
], R);
var z = { exports: {} }, E = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ie;
function ut() {
  if (ie) return E;
  ie = 1;
  var e = rt, t = Symbol.for("react.element"), r = Symbol.for("react.fragment"), n = Object.prototype.hasOwnProperty, i = e.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, o = { key: !0, ref: !0, __self: !0, __source: !0 };
  function s(d, a, u) {
    var c, f = {}, C = null, _ = null;
    u !== void 0 && (C = "" + u), a.key !== void 0 && (C = "" + a.key), a.ref !== void 0 && (_ = a.ref);
    for (c in a) n.call(a, c) && !o.hasOwnProperty(c) && (f[c] = a[c]);
    if (d && d.defaultProps) for (c in a = d.defaultProps, a) f[c] === void 0 && (f[c] = a[c]);
    return { $$typeof: t, type: d, key: C, ref: _, props: f, _owner: i.current };
  }
  return E.Fragment = r, E.jsx = s, E.jsxs = s, E;
}
var oe;
function ft() {
  return oe || (oe = 1, z.exports = ut()), z.exports;
}
var ht = ft();
const ve = {
  id: "docs.command.add-comment",
  type: w.COMMAND,
  async handler(e, t) {
    if (!t)
      return !1;
    const { comment: r, unitId: n } = t, o = await e.get(Ge).addComment(r), s = e.get(M), d = He(
      e,
      {
        id: o.threadId,
        type: X.COMMENT,
        unitId: n
      }
    );
    if (d) {
      const a = {
        id: Ze.id,
        params: {
          unitId: n,
          subUnitId: I,
          comment: o
        }
      }, u = {
        id: ae.id,
        params: {
          unitId: n,
          subUnitId: I,
          commentId: o.id
        }
      };
      return (await se([a, d, u], s)).result;
    }
    return !1;
  }
}, ge = {
  id: "docs.command.delete-comment",
  type: w.COMMAND,
  async handler(e, t) {
    if (!t)
      return !1;
    const { commentId: r, unitId: n } = t, i = e.get(M), o = We(e, {
      id: r,
      unitId: n
    });
    return o ? (await se([o], i)).result : !1;
  }
}, Ie = (e) => {
  const t = e.get(L), r = e.get(H), n = t.getCurrent(), i = n == null ? void 0 : n.with(je).getSkeleton(), o = i == null ? void 0 : i.getViewModel().getEditArea();
  if (o === te.FOOTER || o === te.HEADER)
    return !0;
  const s = r.getActiveTextRange();
  return !!(s == null || s.collapsed);
};
function Ct(e) {
  return {
    id: q.id,
    type: ue.BUTTON,
    icon: "CommentSingle",
    title: "threadCommentUI.panel.addComment",
    tooltip: "threadCommentUI.panel.addComment",
    hidden$: fe(e, S.UNIVER_DOC, void 0, Te),
    disabled$: new he(function(t) {
      const n = e.get(H).textSelection$.pipe(Ce(16)).subscribe(() => {
        t.next(Ie(e));
      });
      return () => {
        n.unsubscribe();
      };
    })
  };
}
function _t(e) {
  return {
    id: ee.id,
    type: ue.BUTTON,
    icon: "CommentSingle",
    title: "threadCommentUI.panel.addComment",
    tooltip: "threadCommentUI.panel.addComment",
    hidden$: fe(e, S.UNIVER_DOC)
  };
}
const y = () => {
  const e = P(F), t = P(de), r = Z(() => e.getCurrentTypeOfUnit$(S.UNIVER_DOC).pipe(nt((m) => !!m && !ce(m.getUnitId()))), [e]), n = G(r), i = Z(() => new he((m) => m.next(I)), []), o = P(H), s = Z(
    () => o.textSelection$.pipe(Ce(16)),
    [o.textSelection$]
  );
  G(s);
  const d = P(M), a = P(R), u = G(a.addingComment$), [c, f] = it([]);
  if (ot(() => {
    var b;
    const m = /* @__PURE__ */ new Set(), l = n == null ? void 0 : n.getCustomDecorations();
    f((b = l == null ? void 0 : l.map((h) => h.id).filter((h) => {
      const O = m.has(h);
      return m.add(h), !O;
    })) != null ? b : []);
    const g = d.onCommandExecuted((h) => {
      var O;
      if (h.id === me.id) {
        const U = /* @__PURE__ */ new Set(), J = n == null ? void 0 : n.getCustomDecorations();
        f((O = J == null ? void 0 : J.map((N) => N.id).filter((N) => {
          const De = U.has(N);
          return U.add(N), !De;
        })) != null ? O : []);
      }
    });
    return () => {
      g.dispose();
    };
  }, [d, n]), !n)
    return null;
  const C = Ie(t), _ = n.getUnitId();
  return /* @__PURE__ */ ht.jsx(
    $e,
    {
      unitId: _,
      subUnitId$: i,
      type: S.UNIVER_DOC,
      onAdd: () => {
        d.executeCommand(q.id);
      },
      getSubUnitName: () => "",
      disableAdd: C,
      tempComment: u,
      onAddComment: (m) => {
        if (!m.parentId) {
          const l = {
            unitId: _,
            range: u,
            comment: m
          };
          return d.executeCommand(ve.id, l), a.endAdd(), !1;
        }
        return !0;
      },
      onDeleteComment: (m) => {
        if (!m.parentId) {
          const l = {
            unitId: _,
            commentId: m.id
          };
          return d.executeCommand(ge.id, l), !1;
        }
        return !0;
      },
      showComments: c
    }
  );
};
y.componentKey = "univer.doc.thread-comment-panel";
const A = {
  id: "docs.operation.show-comment-panel",
  type: w.OPERATION,
  handler(e, t) {
    var i;
    const r = e.get(T), n = e.get(W);
    return (!r.panelVisible || ((i = n.options.children) == null ? void 0 : i.label) !== y.componentKey) && (n.open({
      header: { title: "threadCommentUI.panel.title" },
      children: { label: y.componentKey },
      width: 320,
      onClose: () => r.setPanelVisible(!1)
    }), r.setPanelVisible(!0)), t && r.setActiveComment(t == null ? void 0 : t.activeComment), !0;
  }
}, ee = {
  id: "docs.operation.toggle-comment-panel",
  type: w.OPERATION,
  handler(e) {
    var n;
    const t = e.get(T), r = e.get(W);
    return !t.panelVisible || ((n = r.options.children) == null ? void 0 : n.label) !== y.componentKey ? (r.open({
      header: { title: "threadCommentUI.panel.title" },
      children: { label: y.componentKey },
      width: 320,
      onClose: () => t.setPanelVisible(!1)
    }), t.setPanelVisible(!0)) : (r.close(), t.setPanelVisible(!1), t.setActiveComment(null)), !0;
  }
}, q = {
  id: "docs.operation.start-add-comment",
  type: w.OPERATION,
  handler(e) {
    var h, O, U;
    const t = e.get(T), n = e.get(F).getCurrentUnitForType(S.UNIVER_DOC), i = e.get(H), o = e.get(L), s = e.get(Ue), d = e.get(R), a = e.get(M), u = e.get(W), c = i.getActiveTextRange();
    if (!n || !c)
      return !1;
    const f = (h = o.getRenderById(n.getUnitId())) == null ? void 0 : h.with(qe);
    if (f == null || f.setReserveRangesStatus(!0), c.collapsed)
      return t.panelVisible ? (t.setPanelVisible(!1), u.close()) : a.executeCommand(A.id), !0;
    a.executeCommand(A.id);
    const C = n.getUnitId(), _ = ((U = (O = n.getBody()) == null ? void 0 : O.dataStream) != null ? U : "").slice(c.startOffset, c.endOffset), m = Pe.transform.getPlainText(_), l = I, g = "", b = {
      unitId: C,
      subUnitId: l,
      id: g,
      ref: m,
      dT: ze(),
      personId: s.getCurrentUser().userID,
      text: {
        dataStream: `\r
`
      },
      startOffset: c.startOffset,
      endOffset: c.endOffset,
      collapsed: !0,
      threadId: g
    };
    return f == null || f.blur(), d.startAdd(b), t.setActiveComment({
      unitId: C,
      subUnitId: l,
      commentId: g
    }), !0;
  }
};
var pt = Object.defineProperty, vt = Object.getOwnPropertyDescriptor, gt = (e, t, r, n) => {
  for (var i = n > 1 ? void 0 : n ? vt(t, r) : t, o = e.length - 1, s; o >= 0; o--)
    (s = e[o]) && (i = (n ? s(t, r, i) : s(i)) || i);
  return n && i && pt(t, r, i), i;
}, x = (e, t) => (r, n) => t(r, n, e);
let $ = class extends B {
  constructor(e, t, r, n, i, o) {
    super(), this._threadCommentPanelService = e, this._univerInstanceService = t, this._commandService = r, this._docThreadCommentService = n, this._renderManagerService = i, this._threadCommentModel = o, this._initSelectionChange(), this._initActiveCommandChange();
  }
  _initSelectionChange() {
    let e;
    this.disposeWithMe(
      this._commandService.onCommandExecuted((t) => {
        var r, n, i, o;
        if (t.id === Be.id) {
          const s = t.params, { unitId: d, ranges: a } = s;
          if (ce(d)) return;
          const u = this._univerInstanceService.getUnit(d, S.UNIVER_DOC), c = a[0];
          if ((e == null ? void 0 : e.startOffset) === (c == null ? void 0 : c.startOffset) && (e == null ? void 0 : e.endOffset) === (c == null ? void 0 : c.endOffset))
            return;
          if (e = c, c && u) {
            const { startOffset: f, endOffset: C, collapsed: _ } = c;
            let m;
            if (_ ? m = (n = (r = u.getBody()) == null ? void 0 : r.customDecorations) == null ? void 0 : n.find((l) => l.startIndex <= f && l.endIndex >= C - 1) : m = (o = (i = u.getBody()) == null ? void 0 : i.customDecorations) == null ? void 0 : o.find((l) => l.startIndex <= f && l.endIndex >= C - 1), m) {
              const l = this._threadCommentModel.getComment(d, I, m.id);
              l && !l.resolved && this._commandService.executeCommand(A.id, {
                activeComment: {
                  unitId: d,
                  subUnitId: I,
                  commentId: m.id
                }
              });
              return;
            }
          }
          if (!this._threadCommentPanelService.activeCommentId)
            return;
          this._commandService.executeCommand(ae.id);
        }
      })
    );
  }
  _initActiveCommandChange() {
    this.disposeWithMe(this._threadCommentPanelService.activeCommentId$.subscribe((e) => {
      var t, r, n, i;
      if (e) {
        const o = this._univerInstanceService.getUnit(e.unitId);
        if (o) {
          const s = (t = this._renderManagerService.getRenderById(e.unitId)) == null ? void 0 : t.with(Je), d = (n = (r = o.getBody()) == null ? void 0 : r.customDecorations) == null ? void 0 : n.find((a) => a.id === e.commentId);
          d && s && s.scrollToRange({
            startOffset: d.startIndex,
            endOffset: d.endIndex,
            collapsed: !1
          });
        }
      }
      (!e || e.commentId !== ((i = this._docThreadCommentService.addingComment) == null ? void 0 : i.id)) && this._docThreadCommentService.endAdd();
    }));
  }
};
$ = gt([
  x(0, v(T)),
  x(1, F),
  x(2, M),
  x(3, v(R)),
  x(4, L),
  x(5, v(le))
], $);
var p = function() {
  return p = Object.assign || function(e) {
    for (var t, r = 1, n = arguments.length; r < n; r++) {
      t = arguments[r];
      for (var i in t) Object.prototype.hasOwnProperty.call(t, i) && (e[i] = t[i]);
    }
    return e;
  }, p.apply(this, arguments);
}, It = function(e, t) {
  var r = {};
  for (var n in e) Object.prototype.hasOwnProperty.call(e, n) && t.indexOf(n) < 0 && (r[n] = e[n]);
  if (e != null && typeof Object.getOwnPropertySymbols == "function")
    for (var i = 0, n = Object.getOwnPropertySymbols(e); i < n.length; i++)
      t.indexOf(n[i]) < 0 && Object.prototype.propertyIsEnumerable.call(e, n[i]) && (r[n[i]] = e[n[i]]);
  return r;
}, Se = _e(function(e, t) {
  var r = e.icon, n = e.id, i = e.className, o = e.extend, s = It(e, ["icon", "id", "className", "extend"]), d = "univerjs-icon univerjs-icon-".concat(n, " ").concat(i || "").trim(), a = st("_".concat(Mt()));
  return Oe(r, "".concat(n), { defIds: r.defIds, idSuffix: a.current }, p({ ref: t, className: d }, s), o);
});
function Oe(e, t, r, n, i) {
  return pe(e.tag, p(p({ key: t }, St(e, r, i)), n), (Ot(e, r).children || []).map(function(o, s) {
    return Oe(o, "".concat(t, "-").concat(e.tag, "-").concat(s), r, void 0, i);
  }));
}
function St(e, t, r) {
  var n = p({}, e.attrs);
  r != null && r.colorChannel1 && n.fill === "colorChannel1" && (n.fill = r.colorChannel1);
  var i = t.defIds;
  return !i || i.length === 0 || (e.tag === "use" && n["xlink:href"] && (n["xlink:href"] = n["xlink:href"] + t.idSuffix), Object.entries(n).forEach(function(o) {
    var s = o[0], d = o[1];
    typeof d == "string" && (n[s] = d.replace(/url\(#(.*)\)/, "url(#$1".concat(t.idSuffix, ")")));
  })), n;
}
function Ot(e, t) {
  var r, n = t.defIds;
  return !n || n.length === 0 ? e : e.tag === "defs" && (!((r = e.children) === null || r === void 0) && r.length) ? p(p({}, e), { children: e.children.map(function(i) {
    return typeof i.attrs.id == "string" && n && n.indexOf(i.attrs.id) > -1 ? p(p({}, i), { attrs: p(p({}, i.attrs), { id: i.attrs.id + t.idSuffix }) }) : i;
  }) }) : e;
}
function Mt() {
  return Math.random().toString(36).substring(2, 8);
}
Se.displayName = "UniverIcon";
var bt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 17", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M5.83725 6.78345C6.22188 6.78345 6.53368 7.10742 6.53368 7.50706V8.41159C6.53368 8.81123 6.22188 9.13521 5.83725 9.13521C5.45263 9.13521 5.14082 8.81123 5.14082 8.41159V7.50706C5.14082 7.10742 5.45263 6.78345 5.83725 6.78345ZM8.73904 6.78345C9.12366 6.78345 9.43546 7.10742 9.43546 7.50706V8.41159C9.43546 8.81123 9.12366 9.13521 8.73904 9.13521C8.35441 9.13521 8.04261 8.81123 8.04261 8.41159V7.50706C8.04261 7.10742 8.35441 6.78345 8.73904 6.78345ZM11.6408 6.78345C12.0254 6.78345 12.3372 7.10742 12.3372 7.50706V8.41159C12.3372 8.81123 12.0254 9.13521 11.6408 9.13521C11.2562 9.13521 10.9444 8.81123 10.9444 8.41159V7.50706C10.9444 7.10742 11.2562 6.78345 11.6408 6.78345Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M5.83725 6.78345C6.22188 6.78345 6.53368 7.10742 6.53368 7.50706V8.41159C6.53368 8.81123 6.22188 9.13521 5.83725 9.13521 5.45263 9.13521 5.14082 8.81123 5.14082 8.41159V7.50706C5.14082 7.10742 5.45263 6.78345 5.83725 6.78345zM8.73904 6.78345C9.12366 6.78345 9.43546 7.10742 9.43546 7.50706V8.41159C9.43546 8.81123 9.12366 9.13521 8.73904 9.13521 8.35441 9.13521 8.04261 8.81123 8.04261 8.41159V7.50706C8.04261 7.10742 8.35441 6.78345 8.73904 6.78345zM11.6408 6.78345C12.0254 6.78345 12.3372 7.10742 12.3372 7.50706V8.41159C12.3372 8.81123 12.0254 9.13521 11.6408 9.13521 11.2562 9.13521 10.9444 8.81123 10.9444 8.41159V7.50706C10.9444 7.10742 11.2562 6.78345 11.6408 6.78345z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.84351 3.41861C1.84351 3.01861 2.15531 2.69434 2.53993 2.69434H14.9381C15.3228 2.69434 15.6346 3.01861 15.6346 3.41861V12.4611C15.6346 12.8612 15.3228 13.1854 14.9381 13.1854H8.82117L6.06643 14.6179C5.85054 14.7301 5.59416 14.7181 5.38884 14.5862C5.18352 14.4542 5.05855 14.2211 5.05855 13.9701V13.1854H2.53993C2.15531 13.1854 1.84351 12.8612 1.84351 12.4611L1.84351 3.41861ZM6.45141 12.7982L8.34531 12.0135C8.44201 11.9632 8.54864 11.9371 8.65676 11.9371H14.2417C14.3522 11.9371 14.4417 11.8475 14.4417 11.7371V4.14271C14.4417 4.03225 14.3522 3.94271 14.2417 3.94271H3.23636C3.12591 3.94271 3.03636 4.03225 3.03636 4.14271L3.03636 11.7371C3.03636 11.8475 3.12591 11.9371 3.23636 11.9371L5.75498 11.9371C6.1396 11.9371 6.45141 12.0611 6.45141 12.4611V12.7982Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, Me = _e(function(e, t) {
  return pe(Se, Object.assign({}, e, {
    id: "comment-single",
    ref: t,
    icon: bt
  }));
});
Me.displayName = "CommentSingle";
const xt = {
  [Ye.OTHERS]: {
    [ee.id]: {
      order: 1,
      menuItemFactory: _t
    }
  },
  [Qe.MAIN_AREA]: {
    [Xe.DATA]: {
      [q.id]: {
        order: 1,
        menuItemFactory: Ct
      }
    }
  }
};
var Dt = Object.defineProperty, Rt = Object.getOwnPropertyDescriptor, yt = (e, t, r, n) => {
  for (var i = n > 1 ? void 0 : n ? Rt(t, r) : t, o = e.length - 1, s; o >= 0; o--)
    (s = e[o]) && (i = (n ? s(t, r, i) : s(i)) || i);
  return n && i && Dt(t, r, i), i;
}, Y = (e, t) => (r, n) => t(r, n, e);
let V = class extends B {
  constructor(e, t, r) {
    super(), this._commandService = e, this._menuManagerService = t, this._componentManager = r, this._initCommands(), this._initMenus(), this._initComponents();
  }
  _initCommands() {
    [
      ve,
      ge,
      A,
      q,
      ee
    ].forEach((e) => {
      this.disposeWithMe(this._commandService.registerCommand(e));
    });
  }
  _initMenus() {
    this._menuManagerService.mergeMenu(xt);
  }
  _initComponents() {
    [y].forEach((e) => {
      this.disposeWithMe(this._componentManager.register(e.componentKey, e));
    }), this.disposeWithMe(this._componentManager.register("CommentSingle", Me));
  }
};
V = yt([
  Y(0, M),
  Y(1, et),
  Y(2, v(ke))
], V);
var Tt = Object.defineProperty, Ut = Object.getOwnPropertyDescriptor, Pt = (e, t, r, n) => {
  for (var i = n > 1 ? void 0 : n ? Ut(t, r) : t, o = e.length - 1, s; o >= 0; o--)
    (s = e[o]) && (i = (n ? s(t, r, i) : s(i)) || i);
  return n && i && Tt(t, r, i), i;
}, D = (e, t) => (r, n) => t(r, n, e);
let k = class extends B {
  constructor(e, t, r, n, i, o, s) {
    super(), this._context = e, this._docInterceptorService = t, this._threadCommentPanelService = r, this._docRenderController = n, this._univerInstanceService = i, this._threadCommentModel = o, this._commandService = s, this._interceptorViewModel(), this._initReRender(), this._initSyncComments();
  }
  _initReRender() {
    this.disposeWithMe(this._threadCommentPanelService.activeCommentId$.subscribe((e) => {
      var r;
      if (e) {
        this._docRenderController.reRender(e.unitId);
        return;
      }
      const t = (r = this._univerInstanceService.getCurrentUnitForType(S.UNIVER_DOC)) == null ? void 0 : r.getUnitId();
      t && this._docRenderController.reRender(t);
    })), this.disposeWithMe(this._threadCommentModel.commentUpdate$.subscribe((e) => {
      e.type === "resolve" && this._docRenderController.reRender(e.unitId);
    }));
  }
  _interceptorViewModel() {
    this._docInterceptorService.intercept(Le.CUSTOM_DECORATION, {
      handler: (e, t, r) => {
        if (!e)
          return r(e);
        const { unitId: n, index: i, customDecorations: o } = t, s = this._threadCommentPanelService.activeCommentId, { commentId: d, unitId: a } = s || {}, u = o.find((_) => _.id === d), c = this._threadCommentModel.getComment(n, I, e.id);
        if (!c)
          return r({
            ...e,
            show: !1
          });
        const f = u && i >= u.startIndex && i <= u.endIndex, C = a === n && e.id === d;
        return r({
          ...e,
          active: C || f,
          show: !c.resolved
        });
      }
    });
  }
  _initSyncComments() {
    var i, o, s;
    const e = this._context.unit.getUnitId(), t = I, r = (s = (o = (i = this._context.unit.getBody()) == null ? void 0 : i.customDecorations) == null ? void 0 : o.filter((d) => d.type === X.COMMENT).map((d) => d.id)) != null ? s : [];
    r.forEach((d) => {
      this._threadCommentModel.getComment(e, t, d) || this._threadCommentModel.addComment(e, t, { id: d, threadId: d, ref: "", dT: "", personId: "", text: { dataStream: "" }, unitId: e, subUnitId: t });
    }), r.length && this._threadCommentModel.syncThreadComments(this._context.unit.getUnitId(), I, r);
    let n = r.sort();
    this.disposeWithMe(this._commandService.onCommandExecuted((d) => {
      var a, u, c;
      if (d.id === me.id) {
        if (d.params.unitId !== this._context.unit.getUnitId())
          return;
        const C = (c = (u = (a = this._context.unit.getBody()) == null ? void 0 : a.customDecorations) == null ? void 0 : u.filter((m) => m.type === X.COMMENT).map((m) => m.id)) != null ? c : [], _ = C.sort();
        if (JSON.stringify(n) !== JSON.stringify(_)) {
          const m = new Set(n), l = new Set(_), g = /* @__PURE__ */ new Set(), b = /* @__PURE__ */ new Set();
          C.forEach((h) => {
            m.has(h) || g.add(h);
          }), n.forEach((h) => {
            l.has(h) || b.add(h);
          }), n = _, g.forEach((h) => {
            this._threadCommentModel.getComment(e, t, h) || this._threadCommentModel.addComment(e, t, { id: h, threadId: h, ref: "", dT: "", personId: "", text: { dataStream: "" }, unitId: e, subUnitId: t });
          }), this._threadCommentModel.syncThreadComments(e, t, [...g]);
        }
      }
    }));
  }
};
k = Pt([
  D(1, v(Fe)),
  D(2, v(T)),
  D(3, v(Ke)),
  D(4, F),
  D(5, v(le)),
  D(6, M)
], k);
var be = Object.defineProperty, Et = Object.getOwnPropertyDescriptor, wt = (e, t, r) => t in e ? be(e, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : e[t] = r, Nt = (e, t, r, n) => {
  for (var i = n > 1 ? void 0 : n ? Et(t, r) : t, o = e.length - 1, s; o >= 0; o--)
    (s = e[o]) && (i = (n ? s(t, r, i) : s(i)) || i);
  return n && i && be(t, r, i), i;
}, Q = (e, t) => (r, n) => t(r, n, e), xe = (e, t, r) => wt(e, typeof t != "symbol" ? t + "" : t, r);
let j = class extends we {
  constructor(e = ne, t, r, n) {
    super(), this._config = e, this._injector = t, this._renderManagerSrv = r, this._configService = n;
    const { menu: i, ...o } = Ne(
      {},
      ne,
      this._config
    );
    i && this._configService.setConfig("menu", i, { merge: !0 }), this._configService.setConfig(ct, o);
  }
  onStarting() {
    [
      [V],
      [$],
      [R]
    ].forEach((e) => {
      this._injector.add(e);
    });
  }
  onRendered() {
    this._initRenderModule(), this._injector.get($), this._injector.get(V);
  }
  _initRenderModule() {
    [k].forEach((e) => {
      this._renderManagerSrv.registerRenderModule(S.UNIVER_DOC, e);
    });
  }
};
xe(j, "pluginName", dt);
xe(j, "type", S.UNIVER_DOC);
j = Nt([
  Ee(Ve),
  Q(1, v(de)),
  Q(2, L),
  Q(3, Ae)
], j);
export {
  ve as AddDocCommentComment,
  ge as DeleteDocCommentComment,
  A as ShowCommentPanelOperation,
  q as StartAddCommentOperation,
  j as UniverDocsThreadCommentUIPlugin
};
