// @univerjs/network/lib/es/index.js
var Z = Object.defineProperty;
var P = (e, r, t) => r in e ? Z(e, r, { enumerable: !0, configurable: !0, writable: !0, value: t }) : e[r] = t;
var d = (e, r, t) => P(e, typeof r != "symbol" ? r + "" : r, t);
import { createIdentifier as V, Disposable as z, toDisposable as R, remove as J, ILogService as D, Inject as S, Injector as H, Plugin as C, Quantity as T, LookUp as ee, registerDependencies as re, mergeOverrideWithDependencies as te, DisposableCollection as ne } from "@univerjs/core";
import { of as q, firstValueFrom as se, Observable as m, catchError as oe, throwError as ie } from "rxjs";
import { concatMap as M, share as b, retry as ce } from "rxjs/operators";
const k = "application/json";
class I {
  constructor(r) {
    d(this, "_headers", /* @__PURE__ */ new Map());
    typeof r == "string" ? this._handleHeadersString(r) : r instanceof Headers ? this._handleHeaders(r) : r && this._handleHeadersConstructorProps(r);
  }
  forEach(r) {
    this._headers.forEach((t, n) => r(n, t));
  }
  has(r) {
    return !!this._headers.has(r.toLowerCase());
  }
  get(r) {
    const t = r.toLowerCase();
    return this._headers.has(t) ? this._headers.get(t) : null;
  }
  set(r, t) {
    this._setHeader(r, t);
  }
  toHeadersInit() {
    var t, n;
    const r = {};
    return this._headers.forEach((s, o) => {
      r[o] = s.join(",");
    }), (t = r.accept) != null || (r.accept = "application/json, text/plain, */*"), (n = r["content-type"]) != null || (r["content-type"] = "application/json;charset=UTF-8"), r;
  }
  _setHeader(r, t) {
    const n = r.toLowerCase();
    this._headers.has(n) ? this._headers.get(n).push(t.toString()) : this._headers.set(n, [t.toString()]);
  }
  _handleHeadersString(r) {
    r.split(`
`).forEach((t) => {
      const [n, s] = t.split(":");
      n && s && this._setHeader(n, s);
    });
  }
  _handleHeadersConstructorProps(r) {
    Object.entries(r).forEach(([t, n]) => this._setHeader(t, n));
  }
  _handleHeaders(r) {
    r.forEach((t, n) => this._setHeader(n, t));
  }
}
const K = V("network.http-implementation");
class B {
  constructor(r) {
    this.params = r;
  }
  toString() {
    return this.params ? Object.keys(this.params).map((r) => `${r}=${this.params[r]}`).join("&") : "";
  }
}
let ae = 0;
class G {
  constructor(r, t, n) {
    d(this, "uid", ae++);
    this.method = r, this.url = t, this.requestParams = n;
  }
  get headers() {
    return this.requestParams.headers;
  }
  get withCredentials() {
    return this.requestParams.withCredentials;
  }
  get responseType() {
    return this.requestParams.responseType;
  }
  getUrlWithParams() {
    var t, n;
    const r = (n = (t = this.requestParams) == null ? void 0 : t.params) == null ? void 0 : n.toString();
    return r ? `${this.url}${this.url.includes("?") ? "&" : "?"}${r}` : this.url;
  }
  getBody() {
    var n, s;
    const r = (n = this.headers.get("Content-Type")) != null ? n : k, t = (s = this.requestParams) == null ? void 0 : s.body;
    return r === k && t && typeof t == "object" ? JSON.stringify(t) : t ? `${t}` : null;
  }
  getHeadersInit() {
    return this.headers.toHeadersInit();
  }
}
var le = Object.defineProperty, he = Object.getOwnPropertyDescriptor, pe = (e, r, t, n) => {
  for (var s = n > 1 ? void 0 : n ? he(r, t) : r, o = e.length - 1, i; o >= 0; o--)
    (i = e[o]) && (s = (n ? i(r, t, s) : i(s)) || s);
  return n && s && le(r, t, s), s;
}, de = (e, r) => (t, n) => r(t, n, e);
let x = class extends z {
  constructor(r) {
    super();
    d(this, "_interceptors", []);
    // eslint-disable-next-line ts/no-explicit-any
    d(this, "_pipe");
    this._http = r;
  }
  /**
   * Register an HTTP interceptor.
   *
   * @param interceptor the http interceptor
   * @returns a disposable handler to remove the interceptor
   */
  registerHTTPInterceptor(r) {
    if (this._interceptors.indexOf(r) !== -1)
      throw new Error("[HTTPService]: The interceptor has already been registered!");
    return this._interceptors.push(r), this._interceptors = this._interceptors.sort((t, n) => {
      var s, o;
      return ((s = t.priority) != null ? s : 0) - ((o = n.priority) != null ? o : 0);
    }), this._pipe = null, R(() => J(this._interceptors, r));
  }
  get(r, t) {
    return this._request("GET", r, t);
  }
  post(r, t) {
    return this._request("POST", r, t);
  }
  put(r, t) {
    return this._request("PUT", r, t);
  }
  delete(r, t) {
    return this._request("DELETE", r, t);
  }
  patch(r, t) {
    return this._request("PATCH", r, t);
  }
  getSSE(r, t, n) {
    var c, w;
    const s = new I(n == null ? void 0 : n.headers), o = new B(n == null ? void 0 : n.params), i = new G(r, t, {
      headers: s,
      params: o,
      withCredentials: (c = n == null ? void 0 : n.withCredentials) != null ? c : !1,
      reportProgress: !0,
      responseType: (w = n == null ? void 0 : n.responseType) != null ? w : "json",
      body: ["GET", "DELETE"].includes(r) || n == null ? void 0 : n.body
    });
    return q(i).pipe(M((a) => this._runInterceptorsAndImplementation(a)));
  }
  /** The HTTP request implementations */
  async _request(r, t, n) {
    var a, h;
    const s = new I(n == null ? void 0 : n.headers), o = new B(n == null ? void 0 : n.params), i = new G(r, t, {
      headers: s,
      params: o,
      withCredentials: (a = n == null ? void 0 : n.withCredentials) != null ? a : !1,
      // default value for withCredentials is false by MDN
      responseType: (h = n == null ? void 0 : n.responseType) != null ? h : "json",
      body: ["GET", "DELETE"].includes(r) || n == null ? void 0 : n.body
    }), c = q(i).pipe(
      M((l) => this._runInterceptorsAndImplementation(l))
    );
    return await se(c);
  }
  // eslint-disable-next-line ts/no-explicit-any
  _runInterceptorsAndImplementation(r) {
    return this._pipe || (this._pipe = this._interceptors.map((t) => t.interceptor).reduceRight(
      (t, n) => ue(t, n),
      (t, n) => n(t)
    )), this._pipe(
      r,
      (t) => this._http.send(t)
      /* final handler */
    );
  }
};
x = pe([
  de(0, K)
], x);
function ue(e, r) {
  return (t, n) => r(t, (s) => e(s, n));
}
const fe = 200, ge = 300;
var L = /* @__PURE__ */ ((e) => (e[e.Continue = 100] = "Continue", e[e.SwitchingProtocols = 101] = "SwitchingProtocols", e[e.Processing = 102] = "Processing", e[e.EarlyHints = 103] = "EarlyHints", e[e.Ok = 200] = "Ok", e[e.Created = 201] = "Created", e[e.Accepted = 202] = "Accepted", e[e.NonAuthoritativeInformation = 203] = "NonAuthoritativeInformation", e[e.NoContent = 204] = "NoContent", e[e.ResetContent = 205] = "ResetContent", e[e.PartialContent = 206] = "PartialContent", e[e.MultiStatus = 207] = "MultiStatus", e[e.AlreadyReported = 208] = "AlreadyReported", e[e.ImUsed = 226] = "ImUsed", e[e.MultipleChoices = 300] = "MultipleChoices", e[e.MovedPermanently = 301] = "MovedPermanently", e[e.Found = 302] = "Found", e[e.SeeOther = 303] = "SeeOther", e[e.NotModified = 304] = "NotModified", e[e.UseProxy = 305] = "UseProxy", e[e.Unused = 306] = "Unused", e[e.TemporaryRedirect = 307] = "TemporaryRedirect", e[e.PermanentRedirect = 308] = "PermanentRedirect", e[e.BadRequest = 400] = "BadRequest", e[e.Unauthorized = 401] = "Unauthorized", e[e.PaymentRequired = 402] = "PaymentRequired", e[e.Forbidden = 403] = "Forbidden", e[e.NotFound = 404] = "NotFound", e[e.MethodNotAllowed = 405] = "MethodNotAllowed", e[e.NotAcceptable = 406] = "NotAcceptable", e[e.ProxyAuthenticationRequired = 407] = "ProxyAuthenticationRequired", e[e.RequestTimeout = 408] = "RequestTimeout", e[e.Conflict = 409] = "Conflict", e[e.Gone = 410] = "Gone", e[e.LengthRequired = 411] = "LengthRequired", e[e.PreconditionFailed = 412] = "PreconditionFailed", e[e.PayloadTooLarge = 413] = "PayloadTooLarge", e[e.UriTooLong = 414] = "UriTooLong", e[e.UnsupportedMediaType = 415] = "UnsupportedMediaType", e[e.RangeNotSatisfiable = 416] = "RangeNotSatisfiable", e[e.ExpectationFailed = 417] = "ExpectationFailed", e[e.ImATeapot = 418] = "ImATeapot", e[e.MisdirectedRequest = 421] = "MisdirectedRequest", e[e.UnprocessableEntity = 422] = "UnprocessableEntity", e[e.Locked = 423] = "Locked", e[e.FailedDependency = 424] = "FailedDependency", e[e.TooEarly = 425] = "TooEarly", e[e.UpgradeRequired = 426] = "UpgradeRequired", e[e.PreconditionRequired = 428] = "PreconditionRequired", e[e.TooManyRequests = 429] = "TooManyRequests", e[e.RequestHeaderFieldsTooLarge = 431] = "RequestHeaderFieldsTooLarge", e[e.UnavailableForLegalReasons = 451] = "UnavailableForLegalReasons", e[e.InternalServerError = 500] = "InternalServerError", e[e.NotImplemented = 501] = "NotImplemented", e[e.BadGateway = 502] = "BadGateway", e[e.ServiceUnavailable = 503] = "ServiceUnavailable", e[e.GatewayTimeout = 504] = "GatewayTimeout", e[e.HttpVersionNotSupported = 505] = "HttpVersionNotSupported", e[e.VariantAlsoNegotiates = 506] = "VariantAlsoNegotiates", e[e.InsufficientStorage = 507] = "InsufficientStorage", e[e.LoopDetected = 508] = "LoopDetected", e[e.NotExtended = 510] = "NotExtended", e[e.NetworkAuthenticationRequired = 511] = "NetworkAuthenticationRequired", e))(L || {}), Y = /* @__PURE__ */ ((e) => (e[e.DownloadProgress = 0] = "DownloadProgress", e[e.Response = 1] = "Response", e))(Y || {});
class $ {
  constructor({
    body: r,
    headers: t,
    status: n,
    statusText: s
  }) {
    d(this, "type", 1);
    d(this, "body");
    d(this, "headers");
    d(this, "status");
    d(this, "statusText");
    this.body = r, this.headers = t, this.status = n, this.statusText = s;
  }
}
class we {
  constructor(r, t, n) {
    d(this, "type", 0);
    this.total = r, this.loaded = t, this.partialText = n;
  }
}
class ye {
  constructor(r, t, n) {
    this.headers = r, this.status = t, this.statusText = n;
  }
}
class _ {
  constructor({
    request: r,
    headers: t,
    status: n,
    statusText: s,
    error: o
  }) {
    d(this, "request");
    d(this, "headers");
    d(this, "status");
    d(this, "statusText");
    d(this, "error");
    this.request = r, this.headers = t, this.status = n, this.statusText = s, this.error = o;
  }
}
function Q(e) {
  return {
    method: e.method,
    headers: e.getHeadersInit(),
    body: e.getBody(),
    credentials: e.withCredentials ? "include" : void 0
  };
}
var me = Object.defineProperty, _e = Object.getOwnPropertyDescriptor, ve = (e, r, t, n) => {
  for (var s = n > 1 ? void 0 : n ? _e(r, t) : r, o = e.length - 1, i; o >= 0; o--)
    (i = e[o]) && (s = (n ? i(r, t, s) : i(s)) || s);
  return n && s && me(r, t, s), s;
}, Ee = (e, r) => (t, n) => r(t, n, e);
let N = class {
  constructor(e) {
    this._logService = e;
  }
  send(e) {
    return new m((r) => {
      const t = new AbortController();
      return this._send(e, r, t).catch((n) => {
        r.error(new _({
          error: n,
          request: e
        }));
      }), () => t.abort();
    });
  }
  async _send(e, r, t) {
    var a, h;
    let n;
    try {
      const l = Q(e), y = e.getUrlWithParams(), p = fetch(y, {
        signal: t.signal,
        ...l
      });
      this._logService.debug(`[FetchHTTPImplementation]: sending request to url ${y} with params ${l}`), n = await p;
    } catch (l) {
      const y = new _({
        request: e,
        error: l,
        status: (a = l.status) != null ? a : 0,
        statusText: (h = l.statusText) != null ? h : "Unknown Error",
        headers: l.headers
      });
      this._logService.error("[FetchHTTPImplementation]: network error", y), r.error(y);
      return;
    }
    const s = new I(n.headers), o = n.status, i = n.statusText;
    let c = null;
    if (n.body && (c = await this._readBody(e, n, r)), o >= L.Ok && o < L.MultipleChoices)
      r.next(new $({
        body: c,
        headers: s,
        status: o,
        statusText: i
      }));
    else {
      const l = new _({
        request: e,
        error: c,
        status: o,
        statusText: i,
        headers: s
      });
      this._logService.error("[FetchHTTPImplementation]: network error", l), r.error(l);
    }
    r.complete();
  }
  async _readBody(e, r, t) {
    var y, p;
    const n = [], s = r.body.getReader(), o = r.headers.get("content-length");
    let i = 0;
    const c = (y = e.requestParams) == null ? void 0 : y.reportProgress, w = e.responseType;
    let a, h;
    for (; ; ) {
      const { done: u, value: g } = await s.read();
      if (u) break;
      n.push(g), i += g.length, c && w === "text" && (a = (a != null ? a : "") + (h != null ? h : h = new TextDecoder()).decode(g, { stream: !0 }), t.next(new we(
        o ? Number.parseInt(o, 10) : void 0,
        i,
        a
      )));
    }
    const l = Re(n, i);
    try {
      const u = (p = r.headers.get("content-type")) != null ? p : "";
      return be(e, l, u);
    } catch (u) {
      const g = new _({
        request: e,
        error: u,
        status: r.status,
        statusText: r.statusText,
        headers: new I(r.headers)
      });
      return this._logService.error("[FetchHTTPImplementation]: network error", g), t.error(g), null;
    }
  }
};
N = ve([
  Ee(0, D)
], N);
function Re(e, r) {
  const t = new Uint8Array(r);
  let n = 0;
  for (const s of e)
    t.set(s, n), n += s.length;
  return t;
}
const Ie = /^\)\]\}',?\n/;
function be(e, r, t) {
  switch (e.responseType) {
    case "json":
      const n = new TextDecoder().decode(r).replace(Ie, "");
      return n === "" ? null : JSON.parse(n);
    case "text":
      return new TextDecoder().decode(r);
    case "blob":
      return new Blob([r], { type: t });
    case "arraybuffer":
      return r.buffer;
    default:
      throw new Error(`[FetchHTTPImplementation]: unknown response type: ${e.responseType}.`);
  }
}
var xe = Object.defineProperty, Le = Object.getOwnPropertyDescriptor, Ne = (e, r, t, n) => {
  for (var s = n > 1 ? void 0 : n ? Le(r, t) : r, o = e.length - 1, i; o >= 0; o--)
    (i = e[o]) && (s = (n ? i(r, t, s) : i(s)) || s);
  return n && s && xe(r, t, s), s;
}, Fe = (e, r) => (t, n) => r(t, n, e);
let U = class {
  constructor(e) {
    this._logService = e;
  }
  send(e) {
    return new m((r) => {
      const t = new XMLHttpRequest(), n = e.getUrlWithParams(), s = Q(e);
      t.open(e.method, n), e.withCredentials && (t.withCredentials = !0), s.headers && Object.entries(s.headers).forEach(([a, h]) => t.setRequestHeader(a, h));
      const o = () => {
        const a = t.statusText || "OK", h = new I(t.getAllResponseHeaders());
        return new ye(h, t.status, a);
      }, i = () => {
        const { headers: a, statusText: h, status: l } = o(), { responseType: y } = e;
        let p = null, u = null;
        l !== L.NoContent && (p = typeof t.response > "u" ? t.responseText : t.response);
        let g = l >= fe && l < ge;
        if (y === "json" && typeof p == "string") {
          const f = p;
          try {
            p = p ? JSON.parse(p) : null;
          } catch (v) {
            g = !1, p = f, u = v;
          }
        }
        if (g)
          r.next(
            new $({
              body: p,
              headers: a,
              status: l,
              statusText: h
            })
          );
        else {
          const f = new _({
            request: e,
            error: u,
            headers: a,
            status: l,
            statusText: h
          });
          this._logService.error("[XHRHTTPImplementation]: network error", f), r.error(f);
        }
      }, c = (a) => {
        const h = new _({
          request: e,
          error: a,
          status: t.status || 0,
          statusText: t.statusText || "Unknown Error",
          headers: o().headers
        });
        this._logService.error("[XHRHTTPImplementation]: network error", h), r.error(h);
      };
      t.addEventListener("load", i), t.addEventListener("error", c), t.addEventListener("abort", c), t.addEventListener("timeout", c);
      const w = e.getBody();
      return t.send(w), this._logService.debug(`[XHRHTTPImplementation]: sending request to url ${n} with params ${s}`), () => {
        t.readyState !== t.DONE && t.abort(), t.removeEventListener("load", i), t.removeEventListener("error", c), t.removeEventListener("abort", c), t.removeEventListener("timeout", c);
      };
    });
  }
};
U = Ne([
  Fe(0, D)
], U);
var Oe = Object.defineProperty, Ae = Object.getOwnPropertyDescriptor, Ue = (e, r, t, n) => {
  for (var s = n > 1 ? void 0 : n ? Ae(r, t) : r, o = e.length - 1, i; o >= 0; o--)
    (i = e[o]) && (s = (n ? i(r, t, s) : i(s)) || s);
  return n && s && Oe(r, t, s), s;
}, W = (e, r) => (t, n) => r(t, n, e), A;
let X = (A = class extends C {
  constructor(e = void 0, r, t) {
    super(), this._config = e, this._logger = r, this._injector = t;
  }
  onStarting() {
    var t, n, s;
    if (this._injector.get(x, T.OPTIONAL, ee.SKIP_SELF) && !((t = this._config) != null && t.forceUseNewInstance)) {
      this._logger.warn(
        "[UniverNetworkPlugin]",
        'HTTPService is already registered in an ancestor interceptor. Skipping registration. If you want to force a new instance, set "forceUseNewInstance" to true in the plugin configuration.'
      );
      return;
    }
    const r = (n = this._config) != null && n.useFetchImpl ? N : typeof window < "u" ? U : N;
    re(this._injector, te([
      [x],
      [K, { useClass: r }]
    ], (s = this._config) == null ? void 0 : s.override));
  }
}, d(A, "pluginName", "UNIVER_NETWORK_PLUGIN"), A);
X = Ue([
  W(1, D),
  W(2, S(H))
], X);
const We = V("univer.network.socket.service");
class Xe extends z {
  createSocket(r) {
    try {
      const t = new WebSocket(r), n = new ne();
      return {
        URL: r,
        close: (o, i) => {
          t.close(o, i), n.dispose();
        },
        send: (o) => {
          t.send(o);
        },
        open$: new m((o) => {
          const i = (c) => o.next(c);
          t.addEventListener("open", i), n.add(R(() => t.removeEventListener("open", i)));
        }).pipe(b()),
        close$: new m((o) => {
          const i = (c) => o.next(c);
          t.addEventListener("close", i), n.add(R(() => t.removeEventListener("close", i)));
        }).pipe(b()),
        error$: new m((o) => {
          const i = (c) => o.next(c);
          t.addEventListener("error", i), n.add(R(() => t.removeEventListener("error", i)));
        }).pipe(b()),
        message$: new m((o) => {
          const i = (c) => o.next(c);
          t.addEventListener("message", i), n.add(R(() => t.removeEventListener("message", i)));
        }).pipe(b())
      };
    } catch (t) {
      return console.error(t), null;
    }
  }
}
const De = 3, $e = 1e3, Ve = (e) => {
  var n, s;
  const r = (n = e == null ? void 0 : e.maxRetryAttempts) != null ? n : De, t = (s = e == null ? void 0 : e.delayInterval) != null ? s : $e;
  return (o, i) => i(o).pipe(ce({ delay: t, count: r }));
}, ze = (e) => {
  const r = [], t = /* @__PURE__ */ new Set(), n = () => {
    var s;
    for (; t.size < ((s = e == null ? void 0 : e.maxParallel) != null ? s : 1) && r.length > 0; ) {
      const o = r.shift();
      t.add(o), o();
    }
  };
  return (s, o) => new m((i) => {
    const c = () => o(s).subscribe({
      next: (a) => i.next(a),
      error: (a) => i.error(a),
      complete: () => i.complete()
    }), w = () => {
      t.delete(c), J(r, c), n();
    };
    return r.push(c), n(), w;
  });
}, Je = (e) => {
  const { errorStatusCodes: r, onAuthError: t } = e;
  return (s, o) => o(s).pipe(
    oe((i) => (i instanceof _ && r.some((c) => c === i.status) && t(), ie(() => i)))
  );
}, je = (e = 300) => {
  let t = () => {
  };
  return (n) => new Promise((s) => {
    t();
    const o = setTimeout(() => {
      s(!0);
    }, e);
    t = () => {
      clearTimeout(o), s(!1);
    };
  });
}, qe = () => (e, r) => r.map((t) => ({ config: t, result: e })), Ke = (e, r = {}) => {
  const { isMatch: t, getParamsFromRequest: n, mergeParamsToRequest: s } = e, { fetchCheck: o = je(300), distributeResult: i = qe() } = r, c = [], w = (a) => a.map((h) => h.config);
  return (a, h) => t(a) ? new m((l) => {
    const y = n(a);
    c.push({
      next: (u) => l.next(u),
      error: (u) => l.error(u),
      config: y
    });
    const p = w(c);
    o(a).then((u) => {
      if (u) {
        const g = [];
        p.forEach((f) => {
          const v = c.findIndex((E) => E.config === f);
          if (v >= 0) {
            const [E] = c.splice(v, 1);
            g.push(E);
          }
        }), h(s(p, a)).subscribe({
          next: (f) => {
            if (f.type === Y.Response) {
              const v = f.body, E = i(v, p);
              g.forEach((F) => {
                const j = E.find((O) => O.config === F.config);
                if (j) {
                  const O = new $({
                    body: j.result,
                    headers: f.headers,
                    status: f.status,
                    statusText: f.statusText
                  });
                  F.next(O);
                } else
                  F.error("batch error");
              });
            }
          },
          complete: () => l.complete(),
          error: (f) => l.error(f)
        });
      }
    });
  }) : h(a);
};
export {
  Je as AuthInterceptorFactory,
  N as FetchHTTPImplementation,
  Y as HTTPEventType,
  I as HTTPHeaders,
  we as HTTPProgress,
  G as HTTPRequest,
  $ as HTTPResponse,
  _ as HTTPResponseError,
  x as HTTPService,
  L as HTTPStatusCode,
  K as IHTTPImplementation,
  We as ISocketService,
  Ke as MergeInterceptorFactory,
  ye as ResponseHeader,
  Ve as RetryInterceptorFactory,
  ze as ThresholdInterceptorFactory,
  X as UniverNetworkPlugin,
  Xe as WebSocketService,
  U as XHRHTTPImplementation
};
