// @univerjs/preset-sheets-core/lib/es/index.js
import { UniverDocsPlugin as c } from "@univerjs/docs";
export * from "@univerjs/docs";
import { UniverDocsUIPlugin as x } from "@univerjs/docs-ui";
export * from "@univerjs/docs-ui";
import { UniverFormulaEnginePlugin as g } from "@univerjs/engine-formula";
export * from "@univerjs/engine-formula";
import { UniverRenderEnginePlugin as a } from "@univerjs/engine-render";
export * from "@univerjs/engine-render";
import { UniverRPCMainThreadPlugin as P } from "@univerjs/rpc";
export * from "@univerjs/rpc";
import { UniverSheetsPlugin as v } from "@univerjs/sheets";
export * from "@univerjs/sheets";
import { UniverSheetsFormulaPlugin as h } from "@univerjs/sheets-formula";
export * from "@univerjs/sheets-formula";
import { UniverSheetsFormulaUIPlugin as F } from "@univerjs/sheets-formula-ui";
export * from "@univerjs/sheets-formula-ui";
import { UniverSheetsNumfmtPlugin as S } from "@univerjs/sheets-numfmt";
export * from "@univerjs/sheets-numfmt";
import { UniverSheetsNumfmtUIPlugin as d } from "@univerjs/sheets-numfmt-ui";
export * from "@univerjs/sheets-numfmt-ui";
import { UniverSheetsUIPlugin as R } from "@univerjs/sheets-ui";
export * from "@univerjs/sheets-ui";
import { UniverUIPlugin as E } from "@univerjs/ui";
export * from "@univerjs/ui";
import "@univerjs/sheets/facade";
import "@univerjs/ui/facade";
import "@univerjs/docs-ui/facade";
import "@univerjs/sheets-ui/facade";
import "@univerjs/engine-formula/facade";
import "@univerjs/sheets-formula/facade";
import "@univerjs/sheets-numfmt/facade";
function J(t = {}) {
  const {
    container: i = "app",
    workerURL: e,
    header: n,
    footer: m,
    toolbar: p,
    menu: u,
    contextMenu: f,
    disableAutoFocus: l,
    formula: r,
    customComponents: s
  } = t, o = !!e;
  return {
    plugins: [
      c,
      a,
      [E, {
        container: i,
        header: n,
        footer: m,
        toolbar: p,
        menu: u,
        contextMenu: f,
        disableAutoFocus: l
      }],
      x,
      o ? [P, { workerURL: e }] : null,
      [g, { notExecuteFormula: o, function: r == null ? void 0 : r.function }],
      [v, { notExecuteFormula: o, onlyRegisterFormulaRelatedMutations: !1 }],
      [R, { customComponents: s }],
      S,
      d,
      [h, { notExecuteFormula: o, description: r == null ? void 0 : r.description }],
      F
    ].filter((U) => !!U)
  };
}
export {
  J as UniverSheetsCorePreset
};
