// @univerjs/sheets-sort-ui/lib/es/index.js
var We = Object.defineProperty;
var qe = (e, t, n) => t in e ? We(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n;
var X = (e, t, n) => qe(e, typeof t != "symbol" ? t + "" : t, n);
import { useDependency as O, LocaleService as H, Inject as E, Disposable as Ye, UniverInstanceType as q, IUniverInstanceService as Me, ILogService as Je, ICommandService as be, Tools as Xe, LocaleType as Q, CommandType as b, throttle as Ke, Injector as Ee, RxDisposable as Qe, connectInjector as et, DependentOn as tt, Plugin as nt, merge as rt, IConfigService as ot } from "@univerjs/core";
import { SheetsSortService as Oe, SortType as k, SortRangeCommand as st, UniverSheetsSortPlugin as it } from "@univerjs/sheets-sort";
import { serializeRange as ct } from "@univerjs/engine-formula";
import { getCurrentRangeDisable$ as w, getCurrentExclusiveRangeInterest$ as at, SheetsRenderService as lt, SheetsUIPart as dt } from "@univerjs/sheets-ui";
import { IConfirmService as ut, MenuItemType as _, getMenuHiddenObservable as Te, RibbonStartGroup as mt, ContextMenuPosition as ht, ContextMenuGroup as gt, ComponentManager as Ct, IMenuManagerService as St, IDialogService as ft, ILayoutService as pt, IUIPartsService as vt } from "@univerjs/ui";
import { BehaviorSubject as _t, takeUntil as xt } from "rxjs";
import It, { forwardRef as I, useRef as Pe, createElement as y, useState as B, useCallback as P, useEffect as we } from "react";
import { SheetsSelectionsService as yt, getPrimaryForRange as Rt, SetSelectionsOperation as Mt, expandToContinuousRange as pe, getSheetCommandTarget as bt, WorkbookEditablePermission as N, WorksheetSortPermission as L, WorksheetEditPermission as j, RangeProtectionPermissionEditPoint as D } from "@univerjs/sheets";
import { RadioGroup as Ne, Radio as Z, Checkbox as Et, DraggableList as Ot, Button as ve, DropdownLegacy as Tt } from "@univerjs/design";
const Pt = "sheets-sort-ui.config", _e = {};
var v = function() {
  return v = Object.assign || function(e) {
    for (var t, n = 1, r = arguments.length; n < r; n++) {
      t = arguments[n];
      for (var o in t) Object.prototype.hasOwnProperty.call(t, o) && (e[o] = t[o]);
    }
    return e;
  }, v.apply(this, arguments);
}, wt = function(e, t) {
  var n = {};
  for (var r in e) Object.prototype.hasOwnProperty.call(e, r) && t.indexOf(r) < 0 && (n[r] = e[r]);
  if (e != null && typeof Object.getOwnPropertySymbols == "function")
    for (var o = 0, r = Object.getOwnPropertySymbols(e); o < r.length; o++)
      t.indexOf(r[o]) < 0 && Object.prototype.propertyIsEnumerable.call(e, r[o]) && (n[r[o]] = e[r[o]]);
  return n;
}, R = I(function(e, t) {
  var n = e.icon, r = e.id, o = e.className, s = e.extend, c = wt(e, ["icon", "id", "className", "extend"]), l = "univerjs-icon univerjs-icon-".concat(r, " ").concat(o || "").trim(), d = Pe("_".concat(jt()));
  return Le(n, "".concat(r), { defIds: n.defIds, idSuffix: d.current }, v({ ref: t, className: l }, c), s);
});
function Le(e, t, n, r, o) {
  return y(e.tag, v(v({ key: t }, Nt(e, n, o)), r), (Lt(e, n).children || []).map(function(s, c) {
    return Le(s, "".concat(t, "-").concat(e.tag, "-").concat(c), n, void 0, o);
  }));
}
function Nt(e, t, n) {
  var r = v({}, e.attrs);
  n != null && n.colorChannel1 && r.fill === "colorChannel1" && (r.fill = n.colorChannel1);
  var o = t.defIds;
  return !o || o.length === 0 || (e.tag === "use" && r["xlink:href"] && (r["xlink:href"] = r["xlink:href"] + t.idSuffix), Object.entries(r).forEach(function(s) {
    var c = s[0], l = s[1];
    typeof l == "string" && (r[c] = l.replace(/url\(#(.*)\)/, "url(#$1".concat(t.idSuffix, ")")));
  })), r;
}
function Lt(e, t) {
  var n, r = t.defIds;
  return !r || r.length === 0 ? e : e.tag === "defs" && (!((n = e.children) === null || n === void 0) && n.length) ? v(v({}, e), { children: e.children.map(function(o) {
    return typeof o.attrs.id == "string" && r && r.indexOf(o.attrs.id) > -1 ? v(v({}, o), { attrs: v(v({}, o.attrs), { id: o.attrs.id + t.idSuffix }) }) : o;
  }) }) : e;
}
function jt() {
  return Math.random().toString(36).substring(2, 8);
}
R.displayName = "UniverIcon";
var Dt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M12.4208 14.4667C12.4208 14.798 12.1522 15.0667 11.8208 15.0667C11.4895 15.0667 11.2208 14.798 11.2208 14.4667V2.98193L9.97861 4.22417C9.7443 4.45848 9.3644 4.45848 9.13008 4.22417C8.89577 3.98985 8.89577 3.60995 9.13008 3.37564L11.3967 1.10897C11.6311 0.874657 12.011 0.874657 12.2453 1.10897L14.5119 3.37564C14.7463 3.60995 14.7463 3.98985 14.5119 4.22417C14.2776 4.45848 13.8977 4.45848 13.6634 4.22417L12.4208 2.9816V14.4667Z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.98967 10.2798C1.6583 10.2798 1.38967 10.0112 1.38967 9.67983 1.38967 9.34846 1.6583 9.07983 1.98967 9.07983H6.50138C6.74406 9.07983 6.96284 9.22602 7.05571 9.45022 7.14858 9.67443 7.09725 9.9325 6.92565 10.1041L3.43819 13.5916H6.50138C6.83276 13.5916 7.10138 13.8602 7.10138 14.1916 7.10138 14.5229 6.83276 14.7916 6.50138 14.7916H1.98967C1.74699 14.7916 1.52821 14.6454 1.43534 14.4212 1.34247 14.197 1.3938 13.9389 1.5654 13.7673L5.05286 10.2798H1.98967zM5.1846 1.86439C4.8641.989866 3.62725.989866 3.30674 1.86439L1.34882 7.20672C1.23479 7.51786 1.39458 7.86252 1.70571 7.97655 2.01684 8.09058 2.3615 7.93079 2.47553 7.61966L3.06159 6.02055 3.06338 6.02056H5.42975L6.01581 7.61966C6.12984 7.93079 6.4745 8.09058 6.78563 7.97655 7.09677 7.86252 7.25655 7.51786 7.14252 7.20672L5.1846 1.86439zM4.98996 4.82056L4.24567 2.78971 3.50138 4.82056H4.98996z", fillRule: "evenodd", clipRule: "evenodd" } }] }, te = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "ascending- single",
    ref: t,
    icon: Dt
  }));
});
te.displayName = "AscendingSingle";
var kt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M14.1544 3.75557C14.3887 3.98988 14.3887 4.36978 14.1544 4.6041L6.51409 12.2444C6.40157 12.3569 6.24896 12.4201 6.08983 12.4201C5.9307 12.4201 5.77808 12.3569 5.66556 12.2444L1.84541 8.42425C1.6111 8.18993 1.6111 7.81003 1.84541 7.57572C2.07973 7.34141 2.45963 7.34141 2.69394 7.57572L6.08983 10.9716L13.3059 3.75557C13.5402 3.52126 13.9201 3.52126 14.1544 3.75557Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, je = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "check-mark-single",
    ref: t,
    icon: kt
  }));
});
je.displayName = "CheckMarkSingle";
var Ht = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M9.77445 9.24306L9.77457 11.9573C9.77458 12.2887 9.50597 12.5574 9.17459 12.5574 8.84322 12.5574 8.57458 12.2888 8.57457 11.9574L8.57439 7.79509C8.57438 7.46372 8.84299 7.19508 9.17436 7.19507 9.3281 7.19493 9.48196 7.25351 9.59925 7.3708L11.3465 9.11808C11.5808 9.3524 11.5808 9.7323 11.3465 9.96661 11.1122 10.2009 10.7323 10.2009 10.498 9.96661L9.77445 9.24306zM6.22584 10.5094L6.22572 7.79514C6.22571 7.46377 6.49433 7.19513 6.8257 7.19512 7.15707 7.1951 7.42571 7.46372 7.42572 7.79509L7.4259 11.9574C7.42592 12.2888 7.1573 12.5574 6.82593 12.5574 6.67228 12.5575 6.51828 12.4989 6.40104 12.3817L4.65376 10.6344C4.41945 10.4001 4.41945 10.0202 4.65376 9.78588 4.88808 9.55156 5.26798 9.55156 5.50229 9.78588L6.22584 10.5094z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.14014 3.73989C1.14014 2.30396 2.3042 1.13989 3.74014 1.13989H12.2601C13.6961 1.13989 14.8601 2.30395 14.8601 3.73989V12.2599C14.8601 13.6958 13.6961 14.8599 12.2601 14.8599H3.74014C2.3042 14.8599 1.14014 13.6958 1.14014 12.2599V3.73989ZM2.34014 5.86724V12.2599C2.34014 13.0331 2.96694 13.6599 3.74014 13.6599H12.2601C13.0333 13.6599 13.6601 13.0331 13.6601 12.2599V5.86724H2.34014ZM13.6601 4.66724H2.34014V3.73989C2.34014 2.9667 2.96694 2.33989 3.74014 2.33989H12.2601C13.0333 2.33989 13.6601 2.96669 13.6601 3.73989V4.66724Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, De = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "custom-sort-single",
    ref: t,
    icon: Ht
  }));
});
De.displayName = "CustomSortSingle";
var Ut = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M5.3313 1.4667C5.3313 1.13533 5.59993.866699 5.9313.866699H10.069C10.4004.866699 10.669 1.13533 10.669 1.4667 10.669 1.79807 10.4004 2.0667 10.069 2.0667H5.9313C5.59993 2.0667 5.3313 1.79807 5.3313 1.4667zM1.09985 3.64443C1.09985 3.31306 1.36848 3.04443 1.69985 3.04443H14.2999C14.6312 3.04443 14.8999 3.31306 14.8999 3.64443 14.8999 3.9758 14.6312 4.24443 14.2999 4.24443H1.69985C1.36848 4.24443 1.09985 3.9758 1.09985 3.64443z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M4.75332 5.22217C3.86966 5.22217 3.15332 5.93851 3.15332 6.82217V13.1332C3.15332 14.2377 4.04875 15.1332 5.15332 15.1332H10.8465C11.9511 15.1332 12.8465 14.2377 12.8465 13.1331V6.82217C12.8465 5.93851 12.1302 5.22217 11.2465 5.22217H4.75332ZM4.35332 6.82217C4.35332 6.60125 4.53241 6.42217 4.75332 6.42217H11.2465C11.4674 6.42217 11.6465 6.60125 11.6465 6.82217V13.1331C11.6465 13.575 11.2884 13.9331 10.8465 13.9331H5.15332C4.71149 13.9331 4.35332 13.575 4.35332 13.1332V6.82217Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, ke = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "delete-empty-single",
    ref: t,
    icon: Ut
  }));
});
ke.displayName = "DeleteEmptySingle";
var $t = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M12.4208 1.53335C12.4208 1.20198 12.1522 0.93335 11.8208 0.93335C11.4895 0.93335 11.2208 1.20198 11.2208 1.53335V13.0181L9.97861 11.7758C9.7443 11.5415 9.3644 11.5415 9.13008 11.7758C8.89577 12.0101 8.89577 12.39 9.13008 12.6244L11.3967 14.891C11.6311 15.1253 12.011 15.1253 12.2453 14.891L14.5119 12.6244C14.7463 12.39 14.7463 12.0101 14.5119 11.7758C14.2776 11.5415 13.8977 11.5415 13.6634 11.7758L12.4208 13.0184V1.53335Z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.98967 10.2798C1.6583 10.2798 1.38967 10.0112 1.38967 9.67983 1.38967 9.34846 1.6583 9.07983 1.98967 9.07983H6.50138C6.74406 9.07983 6.96284 9.22602 7.05571 9.45022 7.14858 9.67443 7.09725 9.9325 6.92565 10.1041L3.43819 13.5916H6.50138C6.83276 13.5916 7.10138 13.8602 7.10138 14.1916 7.10138 14.5229 6.83276 14.7916 6.50138 14.7916H1.98967C1.74699 14.7916 1.52821 14.6454 1.43534 14.4212 1.34247 14.197 1.3938 13.9389 1.5654 13.7673L5.05286 10.2798H1.98967zM5.1846 1.86439C4.8641.989866 3.62725.989866 3.30674 1.86439L1.34882 7.20672C1.23479 7.51786 1.39458 7.86252 1.70571 7.97655 2.01684 8.09058 2.3615 7.93079 2.47553 7.61966L3.06159 6.02055 3.06338 6.02056H5.42975L6.01581 7.61966C6.12984 7.93079 6.4745 8.09058 6.78563 7.97655 7.09677 7.86252 7.25655 7.51786 7.14252 7.20672L5.1846 1.86439zM4.98996 4.82056L4.24567 2.78971 3.50138 4.82056H4.98996z", fillRule: "evenodd", clipRule: "evenodd" } }] }, ne = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "descending-single",
    ref: t,
    icon: $t
  }));
});
ne.displayName = "DescendingSingle";
var At = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M1.9064 3.20024C1.57503 3.20024 1.3064 2.93161 1.3064 2.60024C1.3064 2.26887 1.57503 2.00024 1.9064 2.00024H4.57306C4.90443 2.00024 5.17306 2.26887 5.17306 2.60024C5.17306 2.93161 4.90443 3.20024 4.57306 3.20024H1.9064ZM1.3064 6.19985C1.3064 6.53122 1.57503 6.79985 1.9064 6.79985H5.9064C6.23777 6.79985 6.5064 6.53122 6.5064 6.19985C6.5064 5.86848 6.23777 5.59985 5.9064 5.59985H1.9064C1.57503 5.59985 1.3064 5.86848 1.3064 6.19985ZM1.9064 10.3997C1.57503 10.3997 1.3064 10.1311 1.3064 9.79971C1.3064 9.46834 1.57503 9.19971 1.9064 9.19971H7.23973C7.5711 9.19971 7.83973 9.46834 7.83973 9.79971C7.83973 10.1311 7.5711 10.3997 7.23973 10.3997H1.9064ZM1.9064 13.9996C1.57503 13.9996 1.3064 13.7309 1.3064 13.3996C1.3064 13.0682 1.57503 12.7996 1.9064 12.7996H8.57306C8.90443 12.7996 9.17306 13.0682 9.17306 13.3996C9.17306 13.7309 8.90443 13.9996 8.57306 13.9996H1.9064Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M12.4267 14.2667C12.4267 14.5981 12.1581 14.8667 11.8267 14.8667C11.4953 14.8667 11.2267 14.5981 11.2267 14.2667V3.18198L9.98447 4.42422C9.75015 4.65853 9.37026 4.65853 9.13594 4.42422C8.90163 4.1899 8.90163 3.81 9.13594 3.57569L11.4026 1.30902C11.6369 1.07471 12.0168 1.07471 12.2511 1.30902L14.5178 3.57569C14.7521 3.81 14.7521 4.1899 14.5178 4.42422C14.2835 4.65853 13.9036 4.65853 13.6693 4.42422L12.4267 3.18165V14.2667Z" } }] }, He = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "expand-ascending-single",
    ref: t,
    icon: At
  }));
});
He.displayName = "ExpandAscendingSingle";
var Bt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M1.9064 12.7998C1.57503 12.7998 1.3064 13.0684 1.3064 13.3998C1.3064 13.7311 1.57503 13.9998 1.9064 13.9998H4.57306C4.90443 13.9998 5.17306 13.7311 5.17306 13.3998C5.17306 13.0684 4.90443 12.7998 4.57306 12.7998H1.9064ZM1.3064 9.80015C1.3064 9.46878 1.57503 9.20015 1.9064 9.20015H5.9064C6.23777 9.20015 6.5064 9.46878 6.5064 9.80015C6.5064 10.1315 6.23777 10.4001 5.9064 10.4001H1.9064C1.57503 10.4001 1.3064 10.1315 1.3064 9.80015ZM1.9064 5.60029C1.57503 5.60029 1.3064 5.86892 1.3064 6.20029C1.3064 6.53166 1.57503 6.80029 1.9064 6.80029H7.23973C7.5711 6.80029 7.83973 6.53166 7.83973 6.20029C7.83973 5.86892 7.5711 5.60029 7.23973 5.60029H1.9064ZM1.9064 2.00044C1.57503 2.00044 1.3064 2.26907 1.3064 2.60044C1.3064 2.93181 1.57503 3.20044 1.9064 3.20044H8.57306C8.90443 3.20044 9.17306 2.93181 9.17306 2.60044C9.17306 2.26907 8.90443 2.00044 8.57306 2.00044H1.9064Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M12.4267 1.7333C12.4267 1.40193 12.1581 1.1333 11.8267 1.1333C11.4953 1.1333 11.2267 1.40193 11.2267 1.7333V12.818L9.98447 11.5758C9.75015 11.3415 9.37026 11.3415 9.13594 11.5758C8.90163 11.8101 8.90163 12.19 9.13594 12.4243L11.4026 14.691C11.6369 14.9253 12.0168 14.9253 12.2511 14.691L14.5178 12.4243C14.7521 12.19 14.7521 11.8101 14.5178 11.5758C14.2835 11.3415 13.9036 11.3415 13.6693 11.5758L12.4267 12.8183V1.7333Z" } }] }, Ue = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "expand-descending-single",
    ref: t,
    icon: Bt
  }));
});
Ue.displayName = "ExpandDescendingSingle";
var Ft = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M8.6 1.99991C8.60001 1.66854 8.33138 1.39991 8.00001 1.3999C7.66864 1.3999 7.40001 1.66853 7.4 1.9999L7.39996 7.3999H1.9999C1.66853 7.3999 1.3999 7.66853 1.3999 7.9999C1.3999 8.33127 1.66853 8.5999 1.9999 8.5999H7.39995L7.3999 13.9999C7.3999 14.3313 7.66853 14.5999 7.9999 14.5999C8.33127 14.5999 8.5999 14.3313 8.5999 13.9999L8.59995 8.5999H13.9999C14.3313 8.5999 14.5999 8.33127 14.5999 7.9999C14.5999 7.66853 14.3313 7.3999 13.9999 7.3999H8.59996L8.6 1.99991Z" } }] }, ee = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "increase-single",
    ref: t,
    icon: Ft
  }));
});
ee.displayName = "IncreaseSingle";
var Vt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M11.3536 6.14645C11.5488 6.34171 11.5488 6.65829 11.3536 6.85355L8.35355 9.85355C8.15829 10.0488 7.84171 10.0488 7.64645 9.85355L4.64645 6.85355C4.45118 6.65829 4.45118 6.34171 4.64645 6.14645C4.84171 5.95118 5.15829 5.95118 5.35355 6.14645L8 8.79289L10.6464 6.14645C10.8417 5.95118 11.1583 5.95118 11.3536 6.14645Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, $e = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "more-down-single",
    ref: t,
    icon: Vt
  }));
});
$e.displayName = "MoreDownSingle";
var zt = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "mask", attrs: { id: "mask0_622_8", style: { maskType: "alpha" }, width: 16, height: 16, x: 0, y: 0, maskUnits: "userSpaceOnUse" }, children: [{ tag: "path", attrs: { fill: "#D9D9D9", d: "M0 0H16V16H0z" } }] }, { tag: "g", attrs: { fill: "currentColor", mask: "url(#mask0_622_8)" }, children: [{ tag: "path", attrs: { d: "M6 5C6.55228 5 7 4.55228 7 4 7 3.44772 6.55228 3 6 3 5.44772 3 5 3.44772 5 4 5 4.55228 5.44772 5 6 5zM6 9C6.55228 9 7 8.55229 7 8 7 7.44772 6.55228 7 6 7 5.44772 7 5 7.44772 5 8 5 8.55229 5.44772 9 6 9zM7 12C7 12.5523 6.55228 13 6 13 5.44772 13 5 12.5523 5 12 5 11.4477 5.44772 11 6 11 6.55228 11 7 11.4477 7 12zM10 5C10.5523 5 11 4.55228 11 4 11 3.44772 10.5523 3 10 3 9.44771 3 9 3.44772 9 4 9 4.55228 9.44771 5 10 5zM11 8C11 8.55229 10.5523 9 10 9 9.44771 9 9 8.55229 9 8 9 7.44772 9.44771 7 10 7 10.5523 7 11 7.44772 11 8zM10 13C10.5523 13 11 12.5523 11 12 11 11.4477 10.5523 11 10 11 9.44771 11 9 11.4477 9 12 9 12.5523 9.44771 13 10 13z" } }] }] }, Ae = I(function(e, t) {
  return y(R, Object.assign({}, e, {
    id: "sequence-single",
    ref: t,
    icon: zt
  }));
});
Ae.displayName = "SequenceSingle";
var K = { exports: {} }, A = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var xe;
function Zt() {
  if (xe) return A;
  xe = 1;
  var e = It, t = Symbol.for("react.element"), n = Symbol.for("react.fragment"), r = Object.prototype.hasOwnProperty, o = e.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, s = { key: !0, ref: !0, __self: !0, __source: !0 };
  function c(l, d, u) {
    var h, g = {}, p = null, S = null;
    u !== void 0 && (p = "" + u), d.key !== void 0 && (p = "" + d.key), d.ref !== void 0 && (S = d.ref);
    for (h in d) r.call(d, h) && !s.hasOwnProperty(h) && (g[h] = d[h]);
    if (l && l.defaultProps) for (h in d = l.defaultProps, d) g[h] === void 0 && (g[h] = d[h]);
    return { $$typeof: t, type: l, key: p, ref: S, props: g, _owner: o.current };
  }
  return A.Fragment = n, A.jsx = c, A.jsxs = c, A;
}
var Ie;
function Gt() {
  return Ie || (Ie = 1, K.exports = Zt()), K.exports;
}
var i = Gt();
const Wt = "univer-extend-confirm-content", qt = "univer-extend-confirm-radio-group", Yt = "univer-custom-sort-panel-container", Jt = "univer-custom-sort-panel-ext", Xt = "univer-add-condition", Kt = "univer-add-condition-text", Qt = "univer-add-condition-disable", en = "univer-custom-sort-panel-footer", tn = "univer-custom-sort-panel-footer-btn", nn = "univer-custom-sort-panel-item", rn = "univer-custom-sort-panel-item-head", on = "univer-custom-sort-panel-item-handler", sn = "univer-custom-sort-panel-item-remove", cn = "univer-custom-sort-panel-item-column-input", an = "univer-custom-sort-panel-item-column-input-dropdown", ln = "univer-custom-sort-panel-item-column-input-text", dn = "univer-custom-sort-panel-item-order-radio", un = "univer-custom-sort-panel-item-order-radio-cn", mn = "univer-condition-list", hn = "univer-embed-sort-btn-container", gn = "univer-embed-sort-btn", Cn = "univer-embed-sort-btn-asc", Sn = "univer-embed-sort-btn-desc", fn = "univer-embed-sort-btn-icon", pn = "univer-custom-sort-col-menu", vn = "univer-custom-sort-col-menu-item", _n = "univer-custom-sort-col-menu-item-desc", a = {
  extendConfirmContent: Wt,
  extendConfirmRadioGroup: qt,
  customSortPanelContainer: Yt,
  customSortPanelExt: Jt,
  addCondition: Xt,
  addConditionText: Kt,
  addConditionDisable: Qt,
  customSortPanelFooter: en,
  customSortPanelFooterBtn: tn,
  customSortPanelItem: nn,
  customSortPanelItemHead: rn,
  customSortPanelItemHandler: on,
  customSortPanelItemRemove: sn,
  customSortPanelItemColumnInput: cn,
  customSortPanelItemColumnInputDropdown: an,
  customSortPanelItemColumnInputText: ln,
  customSortPanelItemOrderRadio: dn,
  customSortPanelItemOrderRadioCn: un,
  conditionList: mn,
  embedSortBtnContainer: hn,
  embedSortBtn: gn,
  embedSortBtnAsc: Cn,
  embedSortBtnDesc: Sn,
  embedSortBtnIcon: fn,
  customSortColMenu: pn,
  customSortColMenuItem: vn,
  customSortColMenuItemDesc: _n
}, xn = (e) => {
  const [t, n] = B("0"), r = O(H);
  return /* @__PURE__ */ i.jsxs("div", { className: a.extendConfirmContent, children: [
    /* @__PURE__ */ i.jsx("div", { className: "extend-confirm-desc", children: r.t("sheets-sort.dialog.sort-reminder-desc") }),
    /* @__PURE__ */ i.jsxs(
      Ne,
      {
        className: a.extendConfirmRadioGroup,
        value: t,
        direction: "vertical",
        onChange: (o) => {
          n(o), e.onChange(o);
        },
        children: [
          /* @__PURE__ */ i.jsx(
            Z,
            {
              value: "0",
              children: r.t("sheets-sort.dialog.sort-reminder-no")
            }
          ),
          /* @__PURE__ */ i.jsx(Z, { value: "1", children: r.t("sheets-sort.dialog.sort-reminder-ext") })
        ]
      }
    )
  ] });
};
var In = Object.defineProperty, yn = Object.getOwnPropertyDescriptor, Rn = (e, t, n, r) => {
  for (var o = r > 1 ? void 0 : r ? yn(t, n) : t, s = e.length - 1, c; s >= 0; s--)
    (c = e[s]) && (o = (r ? c(t, n, o) : c(o)) || o);
  return r && o && In(t, n, o), o;
}, T = (e, t) => (n, r) => t(n, r, e);
const z = {
  MERGE_ERROR: "sheets-sort.error.merge-size",
  EMPTY_ERROR: "sheets-sort.error.empty",
  SINGLE_ERROR: "sheets-sort.error.single",
  FORMULA_ARRAY: "sheets-sort.error.formula-array"
};
let f = class extends Ye {
  constructor(t, n, r, o, s, c, l) {
    super();
    X(this, "_customSortState$", new _t(null));
    X(this, "customSortState$", this._customSortState$.asObservable());
    this._univerInstanceService = t, this._confirmService = n, this._logService = r, this._selectionManagerService = o, this._sheetsSortService = s, this._localeService = c, this._commandService = l;
  }
  async triggerSortDirectly(t, n, r) {
    const o = r || await this._detectSortLocation(n);
    if (!o || !this._check(o))
      return !1;
    const c = {
      orderRules: [{
        type: t ? k.ASC : k.DESC,
        colIndex: o.colIndex
      }],
      range: o.range
    };
    return this._sheetsSortService.applySort(c, o.unitId, o.subUnitId), !0;
  }
  async triggerSortCustomize() {
    const t = await this._detectSortLocation();
    return !t || !this._check(t) ? !1 : (this.showCustomSortPanel(t), !0);
  }
  customSortState() {
    return this._customSortState$.getValue();
  }
  getTitles(t) {
    var d, u;
    const n = (d = this.customSortState()) == null ? void 0 : d.location;
    if (!n)
      return [];
    const { unitId: r, subUnitId: o, range: s } = n, c = (u = this._univerInstanceService.getUnit(r)) == null ? void 0 : u.getSheetBySheetId(o);
    if (!c)
      return [];
    const l = Mn(this._localeService);
    return Array.from(
      { length: s.endColumn - s.startColumn + 1 },
      (h, g) => {
        var S;
        const p = (S = c.getCell(s.startRow, g + s.startColumn)) == null ? void 0 : S.v;
        return {
          index: g + s.startColumn,
          label: t ? `${p != null ? p : l(g + s.startColumn)}` : l(g + s.startColumn)
        };
      }
    );
  }
  setSelection(t, n, r) {
    var c;
    const o = (c = this._univerInstanceService.getUnit(t)) == null ? void 0 : c.getSheetBySheetId(n);
    if (!o)
      return;
    const s = {
      unitId: t,
      subUnitId: n,
      selections: [{ range: r, primary: Rt(r, o), style: null }]
    };
    this._commandService.executeCommand(Mt.id, s);
  }
  async showCheckError(t) {
    return await this._confirmService.confirm({
      id: "sort-range-check-error",
      title: {
        title: this._localeService.t("info.tooltip")
      },
      children: {
        title: /* @__PURE__ */ i.jsx("div", { children: this._localeService.t(t) })
      }
    });
  }
  async showExtendConfirm() {
    let t = !1;
    return await this._confirmService.confirm({
      id: "extend-sort-range-dialog",
      title: {
        title: this._localeService.t("sheets-sort.dialog.sort-reminder")
      },
      children: {
        title: /* @__PURE__ */ i.jsx(
          xn,
          {
            onChange: (r) => {
              t = r === "1";
            }
          }
        )
      },
      width: 400
    }) ? t ? "extend" : "keep" : "cancel";
  }
  showCustomSortPanel(t) {
    this._customSortState$.next({ location: t, show: !0 });
  }
  closeCustomSortPanel() {
    this._customSortState$.next({ show: !1 });
  }
  _check(t) {
    return this._sheetsSortService.singleCheck(t) ? this._sheetsSortService.mergeCheck(t) ? this._sheetsSortService.formulaCheck(t) ? this._sheetsSortService.emptyCheck(t) ? !0 : (this.showCheckError(z.EMPTY_ERROR), !1) : (this.showCheckError(z.FORMULA_ARRAY), !1) : (this.showCheckError(z.MERGE_ERROR), !1) : (this.showCheckError(z.SINGLE_ERROR), !1);
  }
  async _detectSortLocation(t) {
    const n = this._univerInstanceService.getCurrentUnitForType(q.UNIVER_SHEET), r = n.getActiveSheet(), o = n.getUnitId(), s = r.getSheetId(), c = this._selectionManagerService.getCurrentLastSelection();
    if (!c)
      return null;
    let l;
    if (t === !0)
      l = pe(c.range, { up: !0, down: !0, left: !0, right: !0 }, r), this.setSelection(o, s, l);
    else if (t === !1)
      l = c.range;
    else {
      const d = await this.showExtendConfirm();
      if (d === "cancel")
        return null;
      d === "keep" ? l = c.range : (l = pe(c.range, { up: !0, down: !0, left: !0, right: !0 }, r), this.setSelection(o, s, l));
    }
    return {
      range: l,
      unitId: o,
      subUnitId: s,
      colIndex: c.primary.actualColumn
    };
  }
};
f = Rn([
  T(0, Me),
  T(1, ut),
  T(2, Je),
  T(3, E(yt)),
  T(4, E(Oe)),
  T(5, E(H)),
  T(6, be)
], f);
function Mn(e) {
  return (t) => {
    const n = Xe.chatAtABC(t);
    switch (e.getCurrentLocale()) {
      case Q.ZH_CN:
        return `"${n}"列`;
      case Q.EN_US:
        return `Column "${n}"`;
      default:
        return `Column "${n}"`;
    }
  };
}
const re = {
  id: "sheet.command.sort-range-asc",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!0, !1)
}, oe = {
  id: "sheet.command.sort-range-asc-ext",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!0, !0)
}, se = {
  id: "sheet.command.sort-range-desc",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!1, !1)
}, ie = {
  id: "sheet.command.sort-range-desc-ext",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!1, !0)
}, ce = {
  id: "sheet.command.sort-range-custom",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortCustomize()
}, ae = {
  id: "sheet.command.sort-range-asc-ctx",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!0, !1)
}, le = {
  id: "sheet.command.sort-range-asc-ext-ctx",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!0, !0)
}, de = {
  id: "sheet.command.sort-range-desc-ctx",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!1, !1)
}, ue = {
  id: "sheet.command.sort-range-desc-ext-ctx",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortDirectly(!1, !0)
}, me = {
  id: "sheet.command.sort-range-custom-ctx",
  type: b.COMMAND,
  handler: async (e) => await e.get(f).triggerSortCustomize()
};
function bn() {
  const e = O(f), t = O(Oe), n = O(H), [r, o] = B(!1), [s, c] = B(0), l = Pe(null), d = e.customSortState();
  if (!d || !d.location)
    return null;
  const { range: u, unitId: h, subUnitId: g } = d.location, p = e.getTitles(r), [S, U] = B([
    { type: k.ASC, colIndex: u.startColumn }
  ]), Y = P((m, x) => {
    const $ = [...S];
    x === null ? $.splice(m, 1) : $[m] = x, U($);
  }, [S]), V = P(
    Ke(() => {
      const m = [...S], x = On(u, S);
      x !== null && (m.push({ type: k.ASC, colIndex: x }), U(m));
    }, 200),
    [S, u]
  ), C = P((m, x) => {
    t.applySort({ range: u, orderRules: m, hasTitle: x }), e.closeCustomSortPanel();
  }, [t, e, u]), J = P(() => {
    e.closeCustomSortPanel();
  }, [e]), Ze = P((m) => {
    o(m), m ? e.setSelection(h, g, { ...u, startRow: u.startRow + 1 }) : e.setSelection(h, g, u);
  }, [e, u, g, h]);
  we(() => {
    l.current && S.length > 5 && (l.current.scrollTop = l.current.scrollHeight);
  }, [S]);
  const Ge = S.length < p.length, fe = S.map((m) => ({ ...m, id: `${m.colIndex}` }));
  return /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelContainer, children: [
    /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelContent, onMouseDown: (m) => {
      m.stopPropagation();
    }, children: [
      /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelExt, children: [
        /* @__PURE__ */ i.jsx("div", { className: a.firstRowCheck, children: /* @__PURE__ */ i.jsx(Et, { checked: r, onChange: (m) => Ze(!!m), children: n.t("sheets-sort.dialog.first-row-check") }) }),
        Ge ? /* @__PURE__ */ i.jsxs("div", { className: a.addCondition, onClick: V, children: [
          /* @__PURE__ */ i.jsx(ee, {}),
          /* @__PURE__ */ i.jsx("span", { className: a.addConditionText, children: n.t("sheets-sort.dialog.add-condition") })
        ] }) : /* @__PURE__ */ i.jsxs(
          "div",
          {
            className: `
                              ${a.addCondition}
                              ${a.addConditionDisable}
                            `,
            children: [
              /* @__PURE__ */ i.jsx(ee, {}),
              /* @__PURE__ */ i.jsx("span", { className: a.addConditionText, children: n.t("sheets-sort.dialog.add-condition") })
            ]
          }
        )
      ] }),
      /* @__PURE__ */ i.jsx(
        "div",
        {
          className: a.conditionList,
          onScroll: (m) => {
            const x = m.currentTarget.scrollTop;
            c(x);
          },
          ref: l,
          children: /* @__PURE__ */ i.jsx(
            Ot,
            {
              list: fe,
              onListChange: U,
              idKey: "id",
              draggableHandle: `.${a.customSortPanelItemHandler}`,
              itemRender: (m) => /* @__PURE__ */ i.jsx(
                En,
                {
                  titles: p,
                  list: fe,
                  item: m,
                  onChange: (x, $) => Y($, x),
                  scrollPosition: s
                }
              ),
              rowHeight: 32,
              margin: [0, 12]
            }
          )
        }
      )
    ] }),
    /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelFooter, children: [
      /* @__PURE__ */ i.jsx(ve, { className: a.customSortPanelFooterBtn, type: "default", onClick: () => J(), children: n.t("sheets-sort.dialog.cancel") }),
      /* @__PURE__ */ i.jsx(ve, { className: a.customSortPanelFooterBtn, type: "primary", onClick: () => C(S, r), children: n.t("sheets-sort.dialog.confirm") })
    ] })
  ] });
}
function En(e) {
  var V;
  const { list: t, item: n, titles: r, onChange: o, scrollPosition: s } = e, c = O(H), l = r.filter((C) => !t.some((J) => J.colIndex === C.index) || C.index === n.colIndex), d = t.findIndex((C) => C.colIndex === n.colIndex), u = P((C) => {
    o({ ...n, colIndex: C.index }, d), g(!1);
  }, [d, n, o]), [h, g] = B(!1), p = (C) => {
    g(C);
  };
  we(() => {
    g(!1);
  }, [s]);
  const S = t.length > 1, U = (V = r.find((C) => C.index === n.colIndex)) == null ? void 0 : V.label, Y = c.getCurrentLocale() === Q.ZH_CN ? a.customSortPanelItemOrderRadioCn : a.customSortPanelItemOrderRadio;
  return /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelItem, children: [
    /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelItemHead, children: [
      /* @__PURE__ */ i.jsx("div", { className: a.customSortPanelItemHandler, children: /* @__PURE__ */ i.jsx(Ae, {}) }),
      /* @__PURE__ */ i.jsx("div", { className: a.customSortPanelItemColumn, children: /* @__PURE__ */ i.jsx(
        Tt,
        {
          placement: "bottomLeft",
          trigger: ["click"],
          visible: h,
          onVisibleChange: p,
          overlay: /* @__PURE__ */ i.jsx("ul", { className: a.customSortColMenu, children: l.map((C) => /* @__PURE__ */ i.jsxs(
            "li",
            {
              onClick: () => u(C),
              className: a.customSortColMenuItem,
              children: [
                /* @__PURE__ */ i.jsx("span", { className: a.customSortColMenuItemDesc, children: C.label }),
                /* @__PURE__ */ i.jsx("span", { className: a.customSortColMenuItemCheck, children: C.index === n.colIndex && /* @__PURE__ */ i.jsx(je, {}) })
              ]
            },
            C.index
          )) }),
          children: /* @__PURE__ */ i.jsxs("div", { className: a.customSortPanelItemColumnInput, children: [
            /* @__PURE__ */ i.jsx("span", { className: a.customSortPanelItemColumnInputText, children: U }),
            /* @__PURE__ */ i.jsx($e, { className: a.customSortPanelItemColumnInputDropdown })
          ] })
        }
      ) })
    ] }),
    /* @__PURE__ */ i.jsx("div", { className: a.customSortPanelItemOrder, children: /* @__PURE__ */ i.jsxs(
      Ne,
      {
        className: Y,
        value: n.type,
        onChange: (C) => {
          o({ ...n, type: C }, d);
        },
        children: [
          /* @__PURE__ */ i.jsx(Z, { value: k.ASC, children: c.t("sheets-sort.general.sort-asc") }),
          /* @__PURE__ */ i.jsx(Z, { value: k.DESC, children: c.t("sheets-sort.general.sort-desc") })
        ]
      }
    ) }),
    /* @__PURE__ */ i.jsx("div", { className: a.customSortPanelItemRemove, children: S && /* @__PURE__ */ i.jsx(ke, { onClick: () => o(null, d) }) })
  ] });
}
function On(e, t) {
  const { startColumn: n, endColumn: r } = e, o = new Set(t.map((s) => s == null ? void 0 : s.colIndex));
  for (let s = n; s <= r; s++)
    if (!o.has(s))
      return s;
  return null;
}
function Tn(e) {
  const { range: t, colIndex: n, onClose: r } = e, o = O(f), s = O(Me), c = O(H), l = P((d) => {
    const { unitId: u, subUnitId: h } = bt(s) || {};
    if (t && u && h) {
      const g = { ...t, startRow: t.startRow + 1 };
      o.triggerSortDirectly(d, !1, { unitId: u, subUnitId: h, range: g, colIndex: n });
    } else
      console.warn(`Cannot find the target to sort. unitId: ${u}, subUnitId: ${h}, range: ${t}, colIndex: ${n}`);
    r();
  }, [t, n, o, s, r]);
  return /* @__PURE__ */ i.jsxs("div", { className: a.embedSortBtnContainer, children: [
    /* @__PURE__ */ i.jsxs(
      "div",
      {
        className: `
                  ${a.embedSortBtn}
                  ${a.embedSortBtnAsc}
                `,
        onClick: () => l(!0),
        children: [
          /* @__PURE__ */ i.jsx(te, { className: a.embedSortBtnIcon }),
          c.t("sheets-sort.general.sort-asc")
        ]
      }
    ),
    /* @__PURE__ */ i.jsxs(
      "div",
      {
        className: `
                  ${a.embedSortBtn}
                  ${a.embedSortBtnDesc}
                `,
        onClick: () => l(!1),
        children: [
          /* @__PURE__ */ i.jsx(ne, { className: a.embedSortBtnIcon }),
          c.t("sheets-sort.general.sort-desc")
        ]
      }
    )
  ] });
}
const Be = "sheet.menu.sheets-sort", Fe = "sheet.menu.sheets-sort-ctx", F = "AscendingSingle", he = "ExpandAscendingSingle", ge = "DescendingSingle", Ce = "ExpandDescendingSingle", Se = "CustomSortSingle";
function Pn(e) {
  return {
    id: Be,
    type: _.SUBITEMS,
    icon: F,
    tooltip: "sheets-sort.general.sort",
    hidden$: Te(e, q.UNIVER_SHEET),
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
function wn(e) {
  return {
    id: re.id,
    icon: F,
    title: "sheets-sort.general.sort-asc-cur",
    type: _.BUTTON,
    hidden$: at(e)
  };
}
function Nn(e) {
  return {
    id: oe.id,
    title: "sheets-sort.general.sort-asc-ext",
    icon: he,
    type: _.BUTTON
  };
}
function Ln(e) {
  return {
    id: se.id,
    title: "sheets-sort.general.sort-desc-cur",
    icon: ge,
    type: _.BUTTON
  };
}
function jn(e) {
  return {
    id: ie.id,
    title: "sheets-sort.general.sort-desc-ext",
    icon: Ce,
    type: _.BUTTON
  };
}
function Dn(e) {
  return {
    id: ce.id,
    title: "sheets-sort.general.sort-custom",
    type: _.BUTTON,
    icon: Se
  };
}
function kn(e) {
  return {
    id: Fe,
    title: "sheets-sort.general.sort",
    type: _.SUBITEMS,
    icon: F,
    hidden$: Te(e, q.UNIVER_SHEET),
    disabled$: w(e, {
      workbookTypes: [N],
      worksheetTypes: [L, j],
      rangeTypes: [D]
    })
  };
}
function Hn(e) {
  return {
    id: ae.id,
    title: "sheets-sort.general.sort-asc-cur",
    type: _.BUTTON,
    icon: F,
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
function Un(e) {
  return {
    id: le.id,
    title: "sheets-sort.general.sort-asc-ext",
    type: _.BUTTON,
    icon: he,
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
function $n(e) {
  return {
    id: de.id,
    title: "sheets-sort.general.sort-desc-cur",
    type: _.BUTTON,
    icon: ge,
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
function An(e) {
  return {
    id: ue.id,
    title: "sheets-sort.general.sort-desc-ext",
    type: _.BUTTON,
    icon: Ce,
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
function Bn(e) {
  return {
    id: me.id,
    title: "sheets-sort.general.sort-custom",
    type: _.BUTTON,
    icon: Se,
    disabled$: w(e, { workbookTypes: [N], worksheetTypes: [L, j], rangeTypes: [D] })
  };
}
const Fn = {
  [mt.FORMULAS_INSERT]: {
    [Be]: {
      order: 2,
      menuItemFactory: Pn,
      [re.id]: {
        order: 0,
        menuItemFactory: wn
      },
      [oe.id]: {
        order: 1,
        menuItemFactory: Nn
      },
      [se.id]: {
        order: 2,
        menuItemFactory: Ln
      },
      [ie.id]: {
        order: 3,
        menuItemFactory: jn
      },
      [ce.id]: {
        order: 4,
        menuItemFactory: Dn
      }
    }
  },
  [ht.MAIN_AREA]: {
    [gt.DATA]: {
      [Fe]: {
        order: 0,
        menuItemFactory: kn,
        [ae.id]: {
          order: 0,
          menuItemFactory: Hn
        },
        [le.id]: {
          order: 1,
          menuItemFactory: Un
        },
        [de.id]: {
          order: 2,
          menuItemFactory: $n
        },
        [ue.id]: {
          order: 3,
          menuItemFactory: An
        },
        [me.id]: {
          order: 4,
          menuItemFactory: Bn
        }
      }
    }
  }
};
var Vn = Object.defineProperty, zn = Object.getOwnPropertyDescriptor, Zn = (e, t, n, r) => {
  for (var o = r > 1 ? void 0 : r ? zn(t, n) : t, s = e.length - 1, c; s >= 0; s--)
    (c = e[s]) && (o = (r ? c(t, n, o) : c(o)) || o);
  return r && o && Vn(t, n, o), o;
}, M = (e, t) => (n, r) => t(n, r, e);
const ye = "custom-sort-dialog", Gn = 560;
let G = class extends Qe {
  constructor(e, t, n, r, o, s, c, l, d, u) {
    super(), this._commandService = e, this._menuManagerService = t, this._dialogService = n, this._layoutService = r, this._uiPartsService = o, this._sheetRenderService = s, this._localeService = c, this._sheetsSortUIService = l, this._injector = d, this._componentManager = u, this._initCommands(), this._initMenu(), this._initUI();
  }
  _initMenu() {
    this._menuManagerService.mergeMenu(Fn);
  }
  _initCommands() {
    [
      re,
      oe,
      se,
      ie,
      ce,
      ae,
      le,
      de,
      ue,
      me
    ].forEach((e) => this.disposeWithMe(this._commandService.registerCommand(e))), this.disposeWithMe(this._sheetRenderService.registerSkeletonChangingMutations(st.id));
  }
  _initUI() {
    this.disposeWithMe(this._componentManager.register("CustomSortPanel", bn)), this.disposeWithMe(this._uiPartsService.registerComponent(dt.FILTER_PANEL_EMBED_POINT, () => et(Tn, this._injector))), this.disposeWithMe(this._componentManager.register(F, te)), this.disposeWithMe(this._componentManager.register(he, He)), this.disposeWithMe(this._componentManager.register(ge, ne)), this.disposeWithMe(this._componentManager.register(Ce, Ue)), this.disposeWithMe(this._componentManager.register(Se, De)), this._sheetsSortUIService.customSortState$.pipe(xt(this.dispose$)).subscribe((e) => {
      e && e.show && e.location ? this._openCustomSortPanel(e.location) : e && !(e != null && e.show) && this._closePanel();
    });
  }
  _openCustomSortPanel(e) {
    this._dialogService.open({
      id: ye,
      draggable: !0,
      width: Gn,
      title: { title: `${this._localeService.t("sheets-sort.general.sort-custom")}: ${ct(e.range)}` },
      children: { label: "CustomSortPanel" },
      destroyOnClose: !0,
      defaultPosition: Wn(),
      preservePositionOnDestroy: !1,
      onClose: () => this._closePanel(),
      mask: !0
    });
  }
  _closePanel() {
    this._dialogService.close(ye), queueMicrotask(() => this._layoutService.focus());
  }
};
G = Zn([
  M(0, be),
  M(1, St),
  M(2, ft),
  M(3, pt),
  M(4, vt),
  M(5, E(lt)),
  M(6, E(H)),
  M(7, E(f)),
  M(8, E(Ee)),
  M(9, E(Ct))
], G);
function Wn() {
  return { x: 0, y: 0 };
}
var Ve = Object.defineProperty, qn = Object.getOwnPropertyDescriptor, Yn = (e, t, n) => t in e ? Ve(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, Jn = (e, t, n, r) => {
  for (var o = r > 1 ? void 0 : r ? qn(t, n) : t, s = e.length - 1, c; s >= 0; s--)
    (c = e[s]) && (o = (r ? c(t, n, o) : c(o)) || o);
  return r && o && Ve(t, n, o), o;
}, Re = (e, t) => (n, r) => t(n, r, e), ze = (e, t, n) => Yn(e, typeof t != "symbol" ? t + "" : t, n);
const Xn = "SHEET_SORT_UI_PLUGIN";
let W = class extends nt {
  constructor(e = _e, t, n) {
    super(), this._config = e, this._injector = t, this._configService = n;
    const { ...r } = rt(
      {},
      _e,
      this._config
    );
    this._configService.setConfig(Pt, r);
  }
  onStarting() {
    [
      [f],
      [G]
    ].forEach((e) => this._injector.add(e));
  }
  onRendered() {
    this._injector.get(G);
  }
};
ze(W, "type", q.UNIVER_SHEET);
ze(W, "pluginName", Xn);
W = Jn([
  tt(it),
  Re(1, E(Ee)),
  Re(2, ot)
], W);
export {
  re as SortRangeAscCommand,
  oe as SortRangeAscExtCommand,
  le as SortRangeAscExtInCtxMenuCommand,
  ae as SortRangeAscInCtxMenuCommand,
  ce as SortRangeCustomCommand,
  me as SortRangeCustomInCtxMenuCommand,
  se as SortRangeDescCommand,
  ie as SortRangeDescExtCommand,
  ue as SortRangeDescExtInCtxMenuCommand,
  de as SortRangeDescInCtxMenuCommand,
  W as UniverSheetsSortUIPlugin
};
