// @univerjs/docs-hyper-link-ui/lib/es/index.js
var Ke = Object.defineProperty;
var Fe = (n, e, t) => e in n ? Ke(n, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[e] = t;
var P = (n, e, t) => Fe(n, typeof e != "symbol" ? e + "" : e, t);
import { CommandType as E, ICommandService as x, generateRandomId as Ve, CustomRangeType as D, IUniverInstanceService as O, UniverInstanceType as I, getBodySlice as Pe, useDependency as S, LocaleService as Re, useObservable as Ye, Tools as Ze, BuildTextUtils as _e, Disposable as N, Inject as v, DOCS_ZEN_EDITOR_UNIT_ID_KEY as We, DOCS_NORMAL_EDITOR_UNIT_ID_KEY as Ge, DependentOn as qe, Injector as Je, Plugin as ze, merge as Qe, IConfigService as Xe } from "@univerjs/core";
import { UniverDocsHyperLinkPlugin as en } from "@univerjs/docs-hyper-link";
import { IRenderManagerService as nn } from "@univerjs/engine-render";
import { addCustomRangeBySelectionFactory as tn, DocSelectionManagerService as b, replaceSelectionFactory as rn, deleteCustomRangeFactory as on, SetTextSelectionsOperation as sn, DocSkeletonManagerService as cn, DocInterceptorService as dn, DOC_INTERCEPTOR_POINT as an } from "@univerjs/docs";
import { DocCanvasPopManagerService as ln, DocEventManagerService as un, DocRenderController as pn, whenDocAndEditorFocused as fn } from "@univerjs/docs-ui";
import { BehaviorSubject as ve, distinctUntilChanged as gn, pairwise as hn, Observable as mn, debounceTime as _n } from "rxjs";
import vn, { useState as J, useEffect as In, forwardRef as j, useRef as Cn, createElement as U } from "react";
import { FormLayout as Ie, Input as Ce, Button as Se, Tooltip as B, MessageType as Sn } from "@univerjs/design";
import { KeyCode as te, IMessageService as Ln, useObservable as yn, MetaKeys as kn, MenuItemType as xn, getMenuHiddenObservable as On, RibbonStartGroup as Pn, ContextMenuPosition as Rn, ContextMenuGroup as En, ComponentManager as bn, IMenuManagerService as Mn, IShortcutService as wn } from "@univerjs/ui";
const Dn = "docs-hyper-link-ui.config", Le = {};
var z = { exports: {} }, M = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var ye;
function Nn() {
  if (ye) return M;
  ye = 1;
  var n = vn, e = Symbol.for("react.element"), t = Symbol.for("react.fragment"), i = Object.prototype.hasOwnProperty, r = n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, o = { key: !0, ref: !0, __self: !0, __source: !0 };
  function s(c, d, p) {
    var l, h = {}, f = null, a = null;
    p !== void 0 && (f = "" + p), d.key !== void 0 && (f = "" + d.key), d.ref !== void 0 && (a = d.ref);
    for (l in d) i.call(d, l) && !o.hasOwnProperty(l) && (h[l] = d[l]);
    if (c && c.defaultProps) for (l in d = c.defaultProps, d) h[l] === void 0 && (h[l] = d[l]);
    return { $$typeof: e, type: c, key: f, ref: a, props: h, _owner: r.current };
  }
  return M.Fragment = t, M.jsx = s, M.jsxs = s, M;
}
var ke;
function jn() {
  return ke || (ke = 1, z.exports = Nn()), z.exports;
}
var u = jn();
const Ee = {
  type: E.COMMAND,
  id: "docs.command.add-hyper-link",
  async handler(n, e) {
    if (!e)
      return !1;
    const { payload: t, unitId: i, selections: r } = e, o = n.get(x), s = Ve(), c = tn(
      n,
      {
        rangeId: s,
        rangeType: D.HYPERLINK,
        properties: {
          url: t
        },
        unitId: i,
        selections: r
      }
    );
    return c ? o.syncExecuteCommand(c.id, c.params) : !1;
  }
}, be = {
  id: "docs.command.update-hyper-link",
  type: E.COMMAND,
  handler(n, e) {
    var m;
    if (!e)
      return !1;
    const { unitId: t, payload: i, segmentId: r, linkId: o } = e, s = n.get(x), c = n.get(O), p = n.get(b).getActiveTextRange(), l = c.getUnit(t, I.UNIVER_DOC);
    if (!p || !l)
      return !1;
    const f = (m = Pe(l.getSelfOrHeaderFooterModel(r).getBody(), p.startOffset, p.endOffset).textRuns) == null ? void 0 : m[0];
    f && (f.ed = e.label.length + 1);
    const a = rn(n, {
      unitId: t,
      body: {
        dataStream: `${e.label}`,
        customRanges: [{
          rangeId: o,
          rangeType: D.HYPERLINK,
          startIndex: 0,
          endIndex: e.label.length + 1,
          properties: {
            url: i
          }
        }],
        textRuns: f ? [f] : void 0
      },
      selection: {
        startOffset: p.startOffset,
        endOffset: p.endOffset,
        collapsed: !1,
        segmentId: r
      }
    });
    return a ? s.syncExecuteCommand(a.id, a.params) : !1;
  }
}, Un = "univer-docs-link-edit", $n = "univer-docs-link-edit-title", Tn = "univer-docs-link-edit-close", Hn = "univer-docs-link-edit-buttons", An = "univer-docs-link-edit-button", K = {
  docsLinkEdit: Un,
  docsLinkEditTitle: $n,
  docsLinkEditClose: Tn,
  docsLinkEditButtons: Hn,
  docsLinkEditButton: An
};
function Bn(n) {
  return /^[a-zA-Z]+:\/\//.test(n);
}
function Kn(n) {
  return /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(n);
}
function Fn(n) {
  return Bn(n) ? n : Kn(n) ? `mailto://${n}` : `https://${n}`;
}
const V = () => {
  const n = S(L), e = S(Re), t = Ye(n.editingLink$), i = S(x), r = S(O), o = S(b), [s, c] = J(""), [d, p] = J(""), [l, h] = J(!1), f = Ze.isLegalUrl(s), a = t ? r.getUnit(t.unitId, I.UNIVER_DOC) : r.getCurrentUnitForType(I.UNIVER_DOC);
  In(() => {
    var de, ae, le, ue, pe, fe, ge, he, me;
    const g = o.getActiveTextRange();
    if (!g)
      return;
    if (t) {
      const H = (de = a == null ? void 0 : a.getSelfOrHeaderFooterModel(t.segmentId)) == null ? void 0 : de.getBody(), A = (ae = H == null ? void 0 : H.customRanges) == null ? void 0 : ae.find((q) => (t == null ? void 0 : t.linkId) === q.rangeId && q.startIndex === t.startIndex && q.endIndex === t.endIndex);
      a && A && (c((ue = (le = A.properties) == null ? void 0 : le.url) != null ? ue : ""), p(_e.transform.getPlainText(Pe(H, A.startIndex, A.endIndex + 1).dataStream)));
      return;
    }
    const y = (pe = a == null ? void 0 : a.getSelfOrHeaderFooterModel(g.segmentId)) == null ? void 0 : pe.getBody(), ce = y ? g : null, T = ce && ((ge = _e.customRange.getCustomRangesInterestsWithSelection(ce, (fe = y == null ? void 0 : y.customRanges) != null ? fe : [])) == null ? void 0 : ge[0]);
    a && T && c((me = (he = T == null ? void 0 : T.properties) == null ? void 0 : he.url) != null ? me : "");
  }, [a, t, o, r]);
  const m = () => {
    n.hideEditPopup();
  }, C = () => {
    if (h(!0), !f || !a)
      return;
    const g = Fn(s);
    if (!t)
      i.executeCommand(Ee.id, {
        unitId: a.getUnitId(),
        payload: g
      });
    else {
      if (!d)
        return;
      i.executeCommand(be.id, {
        unitId: a.getUnitId(),
        payload: g,
        linkId: t.linkId,
        label: d,
        segmentId: t.segmentId
      });
    }
    n.hideEditPopup();
  };
  if (a)
    return /* @__PURE__ */ u.jsxs("div", { className: K.docsLinkEdit, children: [
      /* @__PURE__ */ u.jsxs("div", { children: [
        t ? /* @__PURE__ */ u.jsx(
          Ie,
          {
            label: e.t("docLink.edit.label"),
            error: l && !d ? e.t("docLink.edit.labelError") : "",
            children: /* @__PURE__ */ u.jsx(
              Ce,
              {
                value: d,
                onChange: p,
                autoFocus: !0,
                onKeyDown: (g) => {
                  g.keyCode === te.ENTER && C();
                }
              }
            )
          }
        ) : null,
        /* @__PURE__ */ u.jsx(
          Ie,
          {
            label: e.t("docLink.edit.address"),
            error: l && !f ? e.t("docLink.edit.addressError") : "",
            children: /* @__PURE__ */ u.jsx(
              Ce,
              {
                value: s,
                onChange: c,
                autoFocus: !0,
                onKeyDown: (g) => {
                  g.keyCode === te.ENTER && C();
                }
              }
            )
          }
        )
      ] }),
      /* @__PURE__ */ u.jsxs("div", { className: K.docsLinkEditButtons, children: [
        /* @__PURE__ */ u.jsx(
          Se,
          {
            className: K.docsLinkEditButton,
            onClick: m,
            children: e.t("docLink.edit.cancel")
          }
        ),
        /* @__PURE__ */ u.jsx(
          Se,
          {
            disabled: !s,
            className: K.docsLinkEditButton,
            type: "primary",
            onClick: C,
            children: e.t("docLink.edit.confirm")
          }
        )
      ] })
    ] });
};
V.componentKey = "docs-hyper-link-edit";
var _ = function() {
  return _ = Object.assign || function(n) {
    for (var e, t = 1, i = arguments.length; t < i; t++) {
      e = arguments[t];
      for (var r in e) Object.prototype.hasOwnProperty.call(e, r) && (n[r] = e[r]);
    }
    return n;
  }, _.apply(this, arguments);
}, Vn = function(n, e) {
  var t = {};
  for (var i in n) Object.prototype.hasOwnProperty.call(n, i) && e.indexOf(i) < 0 && (t[i] = n[i]);
  if (n != null && typeof Object.getOwnPropertySymbols == "function")
    for (var r = 0, i = Object.getOwnPropertySymbols(n); r < i.length; r++)
      e.indexOf(i[r]) < 0 && Object.prototype.propertyIsEnumerable.call(n, i[r]) && (t[i[r]] = n[i[r]]);
  return t;
}, $ = j(function(n, e) {
  var t = n.icon, i = n.id, r = n.className, o = n.extend, s = Vn(n, ["icon", "id", "className", "extend"]), c = "univerjs-icon univerjs-icon-".concat(i, " ").concat(r || "").trim(), d = Cn("_".concat(Wn()));
  return Me(t, "".concat(i), { defIds: t.defIds, idSuffix: d.current }, _({ ref: e, className: c }, s), o);
});
function Me(n, e, t, i, r) {
  return U(n.tag, _(_({ key: e }, Yn(n, t, r)), i), (Zn(n, t).children || []).map(function(o, s) {
    return Me(o, "".concat(e, "-").concat(n.tag, "-").concat(s), t, void 0, r);
  }));
}
function Yn(n, e, t) {
  var i = _({}, n.attrs);
  t != null && t.colorChannel1 && i.fill === "colorChannel1" && (i.fill = t.colorChannel1);
  var r = e.defIds;
  return !r || r.length === 0 || (n.tag === "use" && i["xlink:href"] && (i["xlink:href"] = i["xlink:href"] + e.idSuffix), Object.entries(i).forEach(function(o) {
    var s = o[0], c = o[1];
    typeof c == "string" && (i[s] = c.replace(/url\(#(.*)\)/, "url(#$1".concat(e.idSuffix, ")")));
  })), i;
}
function Zn(n, e) {
  var t, i = e.defIds;
  return !i || i.length === 0 ? n : n.tag === "defs" && (!((t = n.children) === null || t === void 0) && t.length) ? _(_({}, n), { children: n.children.map(function(r) {
    return typeof r.attrs.id == "string" && i && i.indexOf(r.attrs.id) > -1 ? _(_({}, r), { attrs: _(_({}, r.attrs), { id: r.attrs.id + e.idSuffix }) }) : r;
  }) }) : n;
}
function Wn() {
  return Math.random().toString(36).substring(2, 8);
}
$.displayName = "UniverIcon";
var Gn = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M4.1302 12.4251C4.25802 13.7417 5.36779 14.7708 6.71792 14.7708H11.7179C13.1539 14.7708 14.3179 13.6067 14.3179 12.1708V6.1708C14.3179 4.78586 13.2351 3.65383 11.8698 3.57517C11.742 2.25858 10.6323 1.22949 9.28213 1.22949H4.28213C2.84619 1.22949 1.68213 2.39355 1.68213 3.82949V9.82949C1.68213 11.2144 2.76497 12.3465 4.1302 12.4251ZM10.6583 3.5708H6.71792C5.28198 3.5708 4.11792 4.73486 4.11792 6.1708V11.22C3.4221 11.1387 2.88213 10.5471 2.88213 9.82949V3.82949C2.88213 3.05629 3.50893 2.42949 4.28213 2.42949H9.28213C9.96695 2.42949 10.5369 2.92119 10.6583 3.5708ZM13.1179 6.1708C13.1179 5.3976 12.4911 4.7708 11.7179 4.7708H6.71792C5.94472 4.7708 5.31792 5.3976 5.31792 6.1708V12.1708C5.31792 12.944 5.94472 13.5708 6.71792 13.5708H11.7179C12.4911 13.5708 13.1179 12.944 13.1179 12.1708V6.1708Z", fillRule: "evenodd", clipRule: "evenodd" } }] }, we = j(function(n, e) {
  return U($, Object.assign({}, n, {
    id: "copy-single",
    ref: e,
    icon: Gn
  }));
});
we.displayName = "CopySingle";
var qn = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M9.8816 1.97978C11.0177 0.843607 12.862 0.884962 14.0004 2.02342C15.1389 3.16188 15.1803 5.00612 14.0441 6.14228L11.399 8.78737C11.1608 9.02559 10.7746 9.02559 10.5363 8.78737C10.2981 8.54915 10.2981 8.16292 10.5363 7.9247L13.1814 5.2796C13.8195 4.64155 13.8217 3.57006 13.1378 2.8861C12.4538 2.20211 11.3823 2.20438 10.7443 2.84245L7.6976 5.88911L7.69317 5.89349C7.05959 6.53211 7.05894 7.60014 7.74132 8.28252C7.97954 8.52074 7.97954 8.90697 7.74132 9.14519C7.5031 9.38341 7.11687 9.38341 6.87865 9.14519C5.74016 8.00671 5.69884 6.16251 6.83497 5.02633L6.84021 5.02116L9.8816 1.97978Z" } }, { tag: "path", attrs: { fill: "currentColor", d: "M4.61426 7.2364C4.85248 6.99818 5.23871 6.99818 5.47693 7.2364C5.71515 7.47462 5.71515 7.86085 5.47693 8.09907L2.83183 10.7442C2.19375 11.3823 2.1915 12.4537 2.87547 13.1377C3.55945 13.8217 4.6309 13.8194 5.26899 13.1813L8.31566 10.1347C8.32262 10.1277 8.32971 10.121 8.33691 10.1144C8.34408 10.1064 8.3515 10.0986 8.35916 10.091C8.99721 9.45291 8.99949 8.38145 8.3155 7.69746C8.07728 7.45924 8.07728 7.07301 8.3155 6.83479C8.55372 6.59657 8.93995 6.59657 9.17817 6.83479C10.3166 7.97327 10.358 9.81748 9.22183 10.9536C9.21487 10.9606 9.20779 10.9673 9.20058 10.9739C9.19341 10.9819 9.18599 10.9897 9.17833 10.9973L6.13166 14.044C4.99548 15.1802 3.15127 15.1389 2.01279 14.0004C0.874362 12.8619 0.83297 11.0177 1.96916 9.8815L4.61426 7.2364Z" } }] }, se = j(function(n, e) {
  return U($, Object.assign({}, n, {
    id: "link-single",
    ref: e,
    icon: qn
  }));
});
se.displayName = "LinkSingle";
var Jn = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 16 17", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M12.5935 3.47302C11.6354 2.51492 10.082 2.51492 9.12388 3.47302L7.83534 4.76157C7.60102 4.99588 7.22112 4.99588 6.98681 4.76157 6.75249 4.52725 6.75249 4.14735 6.98681 3.91304L8.27535 2.62449C9.70209 1.19776 12.0153 1.19776 13.442 2.62449 14.8688 4.05123 14.8688 6.36442 13.442 7.79116L12.1535 9.0797C11.9192 9.31402 11.5393 9.31402 11.3049 9.0797 11.0706 8.84539 11.0706 8.46549 11.3049 8.23117L12.5935 6.94263C13.5516 5.98452 13.5516 4.43113 12.5935 3.47302zM3.40637 12.6606C2.44827 11.7025 2.44827 10.1491 3.40637 9.19102L4.69492 7.90248C4.92923 7.66816 4.92923 7.28826 4.69492 7.05395 4.4606 6.81963 4.0807 6.81963 3.84639 7.05395L2.55784 8.34249C1.13111 9.76923 1.13111 12.0824 2.55784 13.5092 3.98458 14.9359 6.29777 14.9359 7.72451 13.5092L9.01305 12.2206C9.24737 11.9863 9.24737 11.6064 9.01305 11.3721 8.77874 11.1378 8.39884 11.1378 8.16452 11.3721L6.87598 12.6606C5.91787 13.6187 4.36448 13.6187 3.40637 12.6606zM3.5852 2.80332C3.35088 2.569 2.97098 2.569 2.73667 2.80332 2.50235 3.03763 2.50235 3.41753 2.73667 3.65185L12.4151 13.3302C12.6494 13.5646 13.0293 13.5646 13.2636 13.3302 13.4979 13.0959 13.4979 12.716 13.2636 12.4817L3.5852 2.80332z" } }] }, De = j(function(n, e) {
  return U($, Object.assign({}, n, {
    id: "unlink-single",
    ref: e,
    icon: Jn
  }));
});
De.displayName = "UnlinkSingle";
var zn = { tag: "svg", attrs: { fill: "none", viewBox: "0 0 17 16", width: "1em", height: "1em" }, children: [{ tag: "path", attrs: { fill: "currentColor", d: "M12.6551 1.98906C11.7476 1.08113 10.2757 1.08149 9.3686 1.98987L4.82542 6.53955C4.75087 6.61421 4.69336 6.70411 4.65682 6.80309L3.2461 10.625C3.16506 10.8446 3.21909 11.0912 3.3845 11.2568C3.54991 11.4224 3.79651 11.4767 4.01616 11.3959L7.85031 9.98517C7.94979 9.94856 8.04014 9.89077 8.11508 9.81579L12.6552 5.27327C13.5618 4.36621 13.5618 2.89607 12.6551 1.98906ZM10.2177 2.83779C10.6562 2.39869 11.3677 2.39851 11.8064 2.8374C12.2447 3.27584 12.2447 3.9865 11.8065 4.42497L7.3392 8.89457L4.82213 9.82068L5.74706 7.31487L10.2177 2.83779Z", fillRule: "evenodd", clipRule: "evenodd" } }, { tag: "path", attrs: { fill: "currentColor", d: "M1.79238 13.2999C1.46101 13.2999 1.19238 13.5685 1.19238 13.8999C1.19238 14.2313 1.46101 14.4999 1.79238 14.4999H14.4924C14.8238 14.4999 15.0924 14.2313 15.0924 13.8999C15.0924 13.5685 14.8238 13.2999 14.4924 13.2999H1.79238Z" } }] }, Ne = j(function(n, e) {
  return U($, Object.assign({}, n, {
    id: "write-single",
    ref: e,
    icon: zn
  }));
});
Ne.displayName = "WriteSingle";
function je(n) {
  var e, t, i = "";
  if (typeof n == "string" || typeof n == "number") i += n;
  else if (typeof n == "object") if (Array.isArray(n)) {
    var r = n.length;
    for (e = 0; e < r; e++) n[e] && (t = je(n[e])) && (i && (i += " "), i += t);
  } else for (t in n) n[t] && (i && (i += " "), i += t);
  return i;
}
function xe() {
  for (var n, e, t = 0, i = "", r = arguments.length; t < r; t++) (n = arguments[t]) && (e = je(n)) && (i && (i += " "), i += e);
  return i;
}
const Ue = {
  type: E.COMMAND,
  id: "docs.command.delete-hyper-link",
  async handler(n, e) {
    if (!e)
      return !1;
    const { unitId: t, linkId: i, segmentId: r } = e, o = n.get(x), s = on(n, { unitId: t, rangeId: i, segmentId: r });
    return s ? await o.syncExecuteCommand(s.id, s.params) : !1;
  }
}, $e = (n) => {
  const e = n.get(b), t = n.get(O), i = e.getTextRanges();
  if (!(i != null && i.length))
    return !0;
  const r = i[0];
  return !!(!t.getCurrentUnitForType(I.UNIVER_DOC) || !r || r.collapsed);
}, R = {
  type: E.OPERATION,
  id: "doc.operation.show-hyper-link-edit-popup",
  handler(n, e) {
    var s;
    const t = e == null ? void 0 : e.link, i = n.get(O);
    if ($e(n) && !t)
      return !1;
    const r = n.get(L), o = (t == null ? void 0 : t.unitId) || ((s = i.getCurrentUnitForType(I.UNIVER_DOC)) == null ? void 0 : s.getUnitId());
    return o ? (r.showEditPopup(o, t), !0) : !1;
  }
}, ie = {
  type: E.OPERATION,
  id: "doc.operation.toggle-hyper-link-info-popup",
  handler(n, e) {
    const t = n.get(L);
    return e ? (t.showInfoPopup(e), !0) : (t.hideInfoPopup(), !0);
  }
}, Te = {
  type: E.OPERATION,
  id: "doc.operation.click-hyper-link",
  handler(n, e) {
    var p, l, h;
    if (!e)
      return !1;
    const { unitId: t, linkId: i, segmentId: r } = e, s = n.get(O).getUnit(t, I.UNIVER_DOC), c = s == null ? void 0 : s.getSelfOrHeaderFooterModel(r).getBody(), d = (h = (l = (p = c == null ? void 0 : c.customRanges) == null ? void 0 : p.find((f) => f.rangeId === i && f.rangeType === D.HYPERLINK)) == null ? void 0 : l.properties) == null ? void 0 : h.url;
    return d && window.open(d, "_blank", "noopener noreferrer"), !0;
  }
}, Qn = "univer-doc-link", Xn = "univer-doc-link-type", et = "univer-doc-link-content", nt = "univer-doc-link-content-error", tt = "univer-doc-link-url", it = "univer-doc-link-operations", rt = "univer-doc-link-operation", ot = "univer-doc-link-operation-error", k = {
  docLink: Qn,
  docLinkType: Xn,
  docLinkContent: et,
  docLinkContentError: nt,
  docLinkUrl: tt,
  docLinkOperations: it,
  docLinkOperation: rt,
  docLinkOperationError: ot
}, Y = () => {
  var C, g;
  const n = S(L), e = S(x), t = S(Ln), i = S(Re), r = yn(n.showingLink$), o = S(O);
  if (!r)
    return null;
  const { unitId: s, linkId: c, segmentId: d, startIndex: p, endIndex: l } = r, h = o.getUnit(s, I.UNIVER_DOC), f = h == null ? void 0 : h.getSelfOrHeaderFooterModel(d).getBody(), a = (C = f == null ? void 0 : f.customRanges) == null ? void 0 : C.find((y) => y.rangeId === c && y.rangeType === D.HYPERLINK && y.startIndex === p && y.endIndex === l);
  if (!a)
    return null;
  const m = (g = a.properties) == null ? void 0 : g.url;
  return /* @__PURE__ */ u.jsxs(
    "div",
    {
      className: k.docLink,
      onClick: () => {
        n.hideInfoPopup();
      },
      children: [
        /* @__PURE__ */ u.jsxs("div", { className: xe(k.docLinkContent), onClick: () => window.open(m), children: [
          /* @__PURE__ */ u.jsx("div", { className: k.docLinkType, children: /* @__PURE__ */ u.jsx(se, {}) }),
          /* @__PURE__ */ u.jsx(B, { showIfEllipsis: !0, title: m, children: /* @__PURE__ */ u.jsx("span", { className: k.docLinkUrl, children: m }) })
        ] }),
        /* @__PURE__ */ u.jsxs("div", { className: k.docLinkOperations, children: [
          /* @__PURE__ */ u.jsx(
            "div",
            {
              className: xe(k.docLinkOperation),
              onClick: () => {
                navigator.clipboard.writeText(m), t.show({
                  content: i.t("docLink.info.coped"),
                  type: Sn.Info
                });
              },
              children: /* @__PURE__ */ u.jsx(B, { placement: "bottom", title: i.t("docLink.info.copy"), children: /* @__PURE__ */ u.jsx(we, {}) })
            }
          ),
          /* @__PURE__ */ u.jsx(
            "div",
            {
              className: k.docLinkOperation,
              onClick: () => {
                e.executeCommand(R.id, {
                  link: r
                });
              },
              children: /* @__PURE__ */ u.jsx(B, { placement: "bottom", title: i.t("docLink.info.edit"), children: /* @__PURE__ */ u.jsx(Ne, {}) })
            }
          ),
          /* @__PURE__ */ u.jsx(
            "div",
            {
              className: k.docLinkOperation,
              onClick: () => {
                e.executeCommand(Ue.id, {
                  unitId: s,
                  linkId: a.rangeId,
                  segmentId: d
                });
              },
              children: /* @__PURE__ */ u.jsx(B, { placement: "bottom", title: i.t("docLink.info.cancel"), children: /* @__PURE__ */ u.jsx(De, {}) })
            }
          )
        ] })
      ]
    }
  );
};
Y.componentKey = "univer.doc.link-info-popup";
var st = Object.defineProperty, ct = Object.getOwnPropertyDescriptor, dt = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? ct(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && st(e, t, r), r;
}, Q = (n, e) => (t, i) => e(t, i, n);
let L = class extends N {
  constructor(e, t, i) {
    super();
    P(this, "_editingLink$", new ve(null));
    P(this, "_showingLink$", new ve(null));
    P(this, "editingLink$", this._editingLink$.asObservable());
    P(this, "showingLink$", this._showingLink$.asObservable());
    P(this, "_editPopup", null);
    P(this, "_infoPopup", null);
    this._docCanvasPopupManagerService = e, this._textSelectionManagerService = t, this._univerInstanceService = i, this.disposeWithMe(() => {
      this._editingLink$.complete(), this._showingLink$.complete();
    });
  }
  get editing() {
    return this._editingLink$.value;
  }
  get showing() {
    return this._showingLink$.value;
  }
  showEditPopup(e, t) {
    this._editPopup && this._editPopup.dispose(), this._editingLink$.next(t);
    const i = this._textSelectionManagerService.getTextRanges({ unitId: e, subUnitId: e });
    let r = i == null ? void 0 : i[i.length - 1];
    if (t) {
      const { segmentId: o, segmentPage: s, startIndex: c, endIndex: d } = t;
      r = {
        collapsed: !1,
        startOffset: c,
        endOffset: d + 1,
        segmentId: o,
        segmentPage: s
      }, this._textSelectionManagerService.replaceDocRanges([{
        startOffset: c,
        endOffset: d + 1
      }]);
    }
    return r ? (this._editPopup = this._docCanvasPopupManagerService.attachPopupToRange(
      r,
      {
        componentKey: V.componentKey,
        direction: "bottom"
      },
      e
    ), this._editPopup) : null;
  }
  hideEditPopup() {
    var e;
    this._editingLink$.next(null), (e = this._editPopup) == null || e.dispose();
  }
  showInfoPopup(e) {
    var p, l, h, f, a, m;
    const { linkId: t, unitId: i, segmentId: r, segmentPage: o, startIndex: s, endIndex: c } = e;
    if (!(((p = this.showing) == null ? void 0 : p.linkId) === t && ((l = this.showing) == null ? void 0 : l.unitId) === i && ((h = this.showing) == null ? void 0 : h.segmentId) === r && ((f = this.showing) == null ? void 0 : f.segmentPage) === o && ((a = this.showing) == null ? void 0 : a.startIndex) === s && ((m = this.showing) == null ? void 0 : m.endIndex) === c || (this._infoPopup && this._infoPopup.dispose(), !this._univerInstanceService.getUnit(i, I.UNIVER_DOC))))
      return this._showingLink$.next({ unitId: i, linkId: t, segmentId: r, segmentPage: o, startIndex: s, endIndex: c }), this._infoPopup = this._docCanvasPopupManagerService.attachPopupToRange(
        {
          collapsed: !1,
          startOffset: s,
          endOffset: c + 1,
          segmentId: r,
          segmentPage: o
        },
        {
          componentKey: Y.componentKey,
          direction: "top-center",
          multipleDirection: "top",
          onClickOutside: () => {
            this.hideInfoPopup();
          }
        },
        i
      ), this._infoPopup;
  }
  hideInfoPopup() {
    var e;
    this._showingLink$.next(null), (e = this._infoPopup) == null || e.dispose();
  }
};
L = dt([
  Q(0, v(ln)),
  Q(1, v(b)),
  Q(2, O)
], L);
var at = Object.defineProperty, lt = Object.getOwnPropertyDescriptor, ut = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? lt(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && at(e, t, r), r;
}, X = (n, e) => (t, i) => e(t, i, n);
let Z = class extends N {
  constructor(n, e, t) {
    super(), this._commandService = n, this._univerInstanceService = e, this._docHyperLinkService = t, this._initSelectionChange();
  }
  _initSelectionChange() {
    this.disposeWithMe(
      this._commandService.onCommandExecuted((n) => {
        var e, t, i;
        if (n.id === sn.id) {
          const r = n.params, { unitId: o, ranges: s, segmentId: c } = r, d = this._univerInstanceService.getUnit(o, I.UNIVER_DOC), p = s[0];
          if (p && d) {
            const { startOffset: l, endOffset: h, collapsed: f, segmentPage: a } = p, m = (t = (e = d.getSelfOrHeaderFooterModel(c)) == null ? void 0 : e.getBody()) == null ? void 0 : t.customRanges;
            if (f) {
              const C = (i = m == null ? void 0 : m.findIndex((g) => g.startIndex < l && g.endIndex > h - 1)) != null ? i : -1;
              if (C > -1) {
                const g = m[C];
                this._docHyperLinkService.showInfoPopup({ unitId: o, linkId: g.rangeId, segmentId: c, segmentPage: a, startIndex: g.startIndex, endIndex: g.endIndex });
                return;
              }
            } else if (m == null ? void 0 : m.find((g) => g.startIndex <= l && g.endIndex >= h - 1))
              return;
          }
          this._docHyperLinkService.hideInfoPopup(), this._docHyperLinkService.hideEditPopup();
        }
      })
    );
  }
};
Z = ut([
  X(0, x),
  X(1, O),
  X(2, v(L))
], Z);
var pt = Object.defineProperty, ft = Object.getOwnPropertyDescriptor, gt = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? ft(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && pt(e, t, r), r;
}, w = (n, e) => (t, i) => e(t, i, n);
let re = class extends N {
  constructor(n, e, t, i, r, o) {
    super(), this._context = n, this._docEventManagerService = e, this._commandService = t, this._hyperLinkPopupService = i, this._docSkeletonManagerService = r, this._docSelectionManagerService = o, !(this._context.unitId === We || this._context.unitId === Ge) && (this._initHover(), this._initClick());
  }
  get _skeleton() {
    return this._docSkeletonManagerService.getSkeleton();
  }
  _hideInfoPopup() {
    this._hyperLinkPopupService.showing && this._commandService.executeCommand(
      ie.id
    );
  }
  _initHover() {
    this.disposeWithMe(
      this._docEventManagerService.hoverCustomRanges$.subscribe((n) => {
        var r;
        const e = n.find((o) => o.range.rangeType === D.HYPERLINK), t = this._docSelectionManagerService.getTextRanges(), i = t == null ? void 0 : t[0].segmentId;
        if (((r = e == null ? void 0 : e.segmentId) != null ? r : "") !== i) {
          this._hideInfoPopup();
          return;
        }
        e ? this._commandService.executeCommand(
          ie.id,
          {
            unitId: this._context.unitId,
            linkId: e.range.rangeId,
            segmentId: e.segmentId,
            segmentPage: e.segmentPageIndex,
            rangeId: e.range.rangeId,
            startIndex: e.range.startIndex,
            endIndex: e.range.endIndex
          }
        ) : this._hideInfoPopup();
      })
    );
  }
  _initClick() {
    this.disposeWithMe(
      this._docEventManagerService.clickCustomRanges$.subscribe((n) => {
        const e = n.range;
        e && this._commandService.executeCommand(
          Te.id,
          {
            unitId: this._context.unitId,
            linkId: e.rangeId,
            segmentId: n.segmentId
          }
        );
      })
    );
  }
};
re = gt([
  w(1, v(un)),
  w(2, x),
  w(3, v(L)),
  w(4, v(cn)),
  w(5, v(b))
], re);
var ht = Object.defineProperty, mt = Object.getOwnPropertyDescriptor, _t = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? mt(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && ht(e, t, r), r;
}, ee = (n, e) => (t, i) => e(t, i, n);
let oe = class extends N {
  constructor(n, e, t, i) {
    super(), this._context = n, this._docInterceptorService = e, this._hyperLinkService = t, this._docRenderController = i, this._init(), this._initReRender();
  }
  _init() {
    this._docInterceptorService.intercept(an.CUSTOM_RANGE, {
      handler: (n, e, t) => {
        if (!n)
          return t(n);
        const { unitId: i, index: r } = e, o = this._hyperLinkService.showing;
        if (!o)
          return t({
            ...n,
            active: !1
          });
        const { linkId: s, unitId: c, startIndex: d, endIndex: p } = o, l = c === i && n.rangeId === s && r >= d && r <= p;
        return t({
          ...n,
          active: l
        });
      }
    });
  }
  _initReRender() {
    this.disposeWithMe(this._hyperLinkService.showingLink$.pipe(
      gn((n, e) => (n == null ? void 0 : n.linkId) === (e == null ? void 0 : e.linkId) && (n == null ? void 0 : n.unitId) === (e == null ? void 0 : e.unitId) && (n == null ? void 0 : n.startIndex) === (e == null ? void 0 : e.startIndex)),
      hn()
    ).subscribe(([n, e]) => {
      e ? e.unitId === this._context.unitId && this._docRenderController.reRender(e.unitId) : n && n.unitId === this._context.unitId && this._docRenderController.reRender(n.unitId);
    }));
  }
};
oe = _t([
  ee(1, v(dn)),
  ee(2, v(L)),
  ee(3, v(pn))
], oe);
const He = "doc-hyper-link-icon";
function Oe(n) {
  return {
    id: R.id,
    type: xn.BUTTON,
    icon: He,
    title: "docLink.menu.tooltip",
    tooltip: "docLink.menu.tooltip",
    hidden$: On(n, I.UNIVER_DOC),
    disabled$: new mn(function(e) {
      const i = n.get(b).textSelection$.pipe(_n(16)).subscribe(() => {
        e.next($e(n));
      });
      return () => {
        i.unsubscribe();
      };
    })
  };
}
const vt = {
  id: R.id,
  binding: kn.CTRL_COMMAND | te.K,
  description: "docLink.menu.tooltip",
  preconditions: fn
}, It = {
  [Pn.OTHERS]: {
    [R.id]: {
      order: 0,
      menuItemFactory: Oe
    }
  },
  [Rn.MAIN_AREA]: {
    [En.DATA]: {
      [R.id]: {
        order: 0,
        menuItemFactory: Oe
      }
    }
  }
};
var Ct = Object.defineProperty, St = Object.getOwnPropertyDescriptor, Lt = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? St(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && Ct(e, t, r), r;
}, F = (n, e) => (t, i) => e(t, i, n);
let W = class extends N {
  constructor(n, e, t, i) {
    super(), this._componentManager = n, this._commandService = e, this._menuManagerService = t, this._shortcutService = i, this._initComponents(), this._initCommands(), this._initMenus(), this._initShortcut();
  }
  _initComponents() {
    [
      [V, V.componentKey],
      [Y, Y.componentKey],
      [se, He]
    ].forEach(([n, e]) => {
      this._componentManager.register(e, n);
    });
  }
  _initCommands() {
    [
      Ee,
      be,
      Ue,
      R,
      ie,
      Te
    ].forEach((n) => {
      this._commandService.registerCommand(n);
    });
  }
  _initShortcut() {
    [vt].forEach((n) => {
      this._shortcutService.registerShortcut(n);
    });
  }
  _initMenus() {
    this._menuManagerService.mergeMenu(It);
  }
};
W = Lt([
  F(0, v(bn)),
  F(1, x),
  F(2, Mn),
  F(3, wn)
], W);
const yt = "DOC_HYPER_LINK_UI_PLUGIN";
var Ae = Object.defineProperty, kt = Object.getOwnPropertyDescriptor, xt = (n, e, t) => e in n ? Ae(n, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[e] = t, Ot = (n, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? kt(e, t) : e, o = n.length - 1, s; o >= 0; o--)
    (s = n[o]) && (r = (i ? s(e, t, r) : s(r)) || r);
  return i && r && Ae(e, t, r), r;
}, ne = (n, e) => (t, i) => e(t, i, n), Be = (n, e, t) => xt(n, typeof e != "symbol" ? e + "" : e, t);
let G = class extends ze {
  constructor(n = Le, e, t, i) {
    super(), this._config = n, this._injector = e, this._renderManagerSrv = t, this._configService = i;
    const { menu: r, ...o } = Qe(
      {},
      Le,
      this._config
    );
    r && this._configService.setConfig("menu", r, { merge: !0 }), this._configService.setConfig(Dn, o);
  }
  onStarting() {
    [
      [L],
      [W],
      [Z]
    ].forEach((e) => {
      this._injector.add(e);
    }), this._injector.get(W);
  }
  onReady() {
    this._injector.get(Z);
  }
  onRendered() {
    this._initRenderModule();
  }
  _initRenderModule() {
    [
      [oe],
      [re]
    ].forEach((n) => {
      this._renderManagerSrv.registerRenderModule(I.UNIVER_DOC, n);
    });
  }
};
Be(G, "pluginName", yt);
Be(G, "type", I.UNIVER_DOC);
G = Ot([
  qe(en),
  ne(1, v(Je)),
  ne(2, nn),
  ne(3, Xe)
], G);
export {
  G as UniverDocsHyperLinkUIPlugin
};
