// @univerjs/slides/lib/es/index.js
var be = Object.defineProperty;
var Te = (n, e, r) => e in n ? be(n, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : n[e] = r;
var h = (n, e, r) => Te(n, typeof e != "symbol" ? e + "" : e, r);
import { Plugin as Oe, merge as ye, UniverInstanceType as De, Inject as H, Injector as q, IConfigService as _e, Registry as me, LocaleService as ne, PageElementType as Y, DocumentDataModel as je, getColorStyle as J, BasicShapes as he, sortRules as Ee, SlideDataModel as Me, IContextService as Ve, Styles as Xe, Worksheet as ze } from "@univerjs/core";
import { IRenderingEngine as Ae, IRenderManagerService as Re, Liquid as Le, DocumentViewModel as Ne, DocumentSkeleton as He, Documents as We, SceneViewer as we, Scene as se, Viewport as $, ScrollBar as ve, Image as Se, PageLayoutType as pe, RichText as Ce, Rect as B, Slide as $e, SpreadsheetSkeleton as Ye, Spreadsheet as Fe, SpreadsheetRowHeader as Ke, SpreadsheetColumnHeader as ke, getColor as ue } from "@univerjs/engine-render";
const Be = "slides.config", ge = {};
var qe = Object.defineProperty, Ue = Object.getOwnPropertyDescriptor, Ge = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? Ue(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && qe(e, r, t), t;
}, Z = (n, e) => (r, i) => e(r, i, n);
const Ze = "slides";
var k;
let fe = (k = class extends Oe {
  // private _canvasView: CanvasView | null = null;
  constructor(e = ge, r, i, t) {
    super();
    h(this, "_canvasEngine", null);
    this._config = e, this._injector = r, this._renderManagerService = i, this._configService = t;
    const { ...a } = ye(
      {},
      ge,
      this._config
    );
    this._configService.setConfig(Be, a), this._initializeDependencies(this._injector);
  }
  initialize() {
    this.initCanvasEngine();
  }
  onReady() {
  }
  getConfig() {
    return this._config;
  }
  initCanvasEngine() {
    this._canvasEngine = this._injector.get(Ae);
  }
  onRendered() {
    this.initialize();
  }
  getCanvasEngine() {
    return this._canvasEngine;
  }
  _initializeDependencies(e) {
    [
      // [CanvasView],
      // [DocSelectionManagerService],
    ].forEach((i) => {
      e.add(i);
    });
  }
}, h(k, "pluginName", Ze), h(k, "type", De.UNIVER_SLIDE), k);
fe = Ge([
  Z(1, H(q)),
  Z(2, Re),
  Z(3, _e)
], fe);
class F {
  constructor() {
    h(this, "zIndex", 0);
    h(this, "viewKey", null);
  }
  check(e) {
    if (e === this.viewKey)
      return this;
  }
  create(e) {
  }
}
const W = me.create();
var Je = Object.defineProperty, Qe = Object.getOwnPropertyDescriptor, et = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? Qe(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && Je(e, r, t), t;
}, tt = (n, e) => (r, i) => e(r, i, n);
let Q = class extends F {
  constructor(e) {
    super();
    h(this, "zIndex", 5);
    h(this, "viewKey", Y.DOCUMENT);
    h(this, "_liquid", new Le());
    this._localeService = e;
  }
  check(e) {
    if (e === this.viewKey)
      return this;
  }
  // eslint-disable-next-line max-lines-per-function
  convert(e, r) {
    var ae, ce;
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      title: D,
      description: M,
      document: b
    } = e;
    if (b == null)
      return;
    const P = new je(b), p = new Ne(P), S = He.create(p, this._localeService), I = new We("__DocsRender__", S);
    S.calculate();
    const T = new we("__DocsViewer__" + i, {
      top: s,
      left: a,
      width: o,
      height: c,
      zIndex: t,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g
    }), f = new se("__DocsScene__" + i, T), x = new $("__DocsViewPort_" + i, f, {
      left: 0,
      top: 0,
      bottom: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      explicitViewportHeightSet: !1,
      isWheelPreventDefaultX: !0
    });
    f.attachControl(), f.onMouseWheel$.subscribeEvent((j, C) => {
      const y = j;
      if (y.ctrlKey) {
        const z = Math.abs(y.deltaX);
        let E = z < 40 ? 0.2 : z < 80 ? 0.4 : 0.2;
        E *= y.deltaY > 0 ? -1 : 1, f.scaleX < 1 && (E /= 2), f.scaleX + E > 4 ? f.scale(4, 4) : f.scaleX + E < 0.1 ? f.scale(0.1, 0.1) : (y.deltaY > 0, y.preventDefault());
      } else
        x.onMouseWheel(y, C);
    }), new ve(x, {
      mainScene: r
    }), f.addObject(I);
    const O = S.getActualSize();
    I.resize(O.actualWidth, O.actualHeight), f.resize(O.actualWidth, O.actualHeight + 200), (ae = I.getSkeleton()) == null || ae.getPageSize(), I.pageRender$.subscribe((j) => {
      const { page: C, pageLeft: y, pageTop: z, ctx: E } = j;
      E.save(), E.translate(y - 0.5, z - 0.5), E.restore();
    });
    const { left: X, top: A } = I, L = S.getSkeletonData();
    if (L == null)
      return;
    const { pages: N } = L, m = [], R = /* @__PURE__ */ new Map();
    this._recalculateSizeBySkeleton(I, f, S);
    for (let j = 0, C = N.length; j < C; j++) {
      const y = N[j], { skeDrawings: z, marginLeft: E, marginTop: oe, pageWidth: le, pageHeight: de } = y;
      this._liquid.translatePagePadding(y), z.forEach((V) => {
        const { aLeft: U, aTop: G, height: Ie, width: Pe, drawingOrigin: Dt } = V, xe = new Se(V.drawingId, {
          // url: docTransform.imageProperties?.contentUrl || '',
          left: U + X + this._liquid.x,
          top: G + A + this._liquid.y,
          width: Pe,
          height: Ie,
          zIndex: 11
        });
        R.set(V.drawingId, {
          marginLeft: this._liquid.x,
          marginTop: this._liquid.y
        }), m.push(xe);
      }), this._liquid.translatePage(
        y,
        I.pageLayoutType,
        I.pageMarginLeft,
        I.pageMarginTop
      );
    }
    return f.addObjects(m), m.forEach((j) => {
      f.attachTransformerTo(j);
    }), (ce = f.getTransformer()) == null || ce.changing$.subscribe((j) => {
      const { objects: C } = j;
      C.forEach((y) => {
        const { oKey: z, left: E, top: oe, height: le, width: de } = y, V = R.get(z), U = (V == null ? void 0 : V.marginLeft) || 0, G = (V == null ? void 0 : V.marginTop) || 0;
        S == null || S.getViewModel().getDataModel().updateDrawing(z, {
          left: E - X - U,
          top: oe - A - G,
          height: le,
          width: de
        });
      }), S == null || S.calculate();
    }), this._calculatePagePosition(I, f, x), T;
  }
  _recalculateSizeBySkeleton(e, r, i) {
    var o;
    const t = (o = i.getSkeletonData()) == null ? void 0 : o.pages;
    if (t == null)
      return;
    let a = 0, s = 0;
    for (let c = 0, w = t.length; c < w; c++) {
      const v = t[c], { pageWidth: u, pageHeight: l } = v;
      e.pageLayoutType === pe.VERTICAL ? (s += l, s += e.pageMarginTop, c === w - 1 && (s += e.pageMarginTop), a = Math.max(a, u)) : e.pageLayoutType === pe.HORIZONTAL && (a += u, c !== w - 1 && (a += e.pageMarginLeft), s = Math.max(s, l));
    }
    e.resize(a, s), r.resize(a, s);
  }
  _calculatePagePosition(e, r, i, t = 1) {
    const a = r == null ? void 0 : r.getParent(), { width: s, height: o, pageMarginLeft: c, pageMarginTop: w } = e;
    if (a == null || s === Number.POSITIVE_INFINITY || o === Number.POSITIVE_INFINITY)
      return;
    const { width: v, height: u } = a;
    let l = 0, d = 0, _ = 0, g = 0, D = Number.POSITIVE_INFINITY;
    if (v > (s + c * 2) * t ? (l = v / 2 - s * t / 2, l /= t, _ = (v - c * 2) / t, D = 0) : (l = c, _ = s + c * 2, D = (_ - v / t) / 2), u > o ? (d = u / 2 - o / 2, g = (u - w * 2) / t) : (d = w, g = o + w * 2), r.resize(_, g + 200), e.translate(l, d), D !== Number.POSITIVE_INFINITY && i != null) {
      const M = i.transScroll2ViewportScrollValue(D, 0).x;
      i.scrollToBarPos({
        x: M
      });
    }
    return this;
  }
};
Q = et([
  tt(0, H(ne))
], Q);
class rt {
  constructor() {
    h(this, "zIndex", 5);
  }
  create(e) {
    return e.createInstance(Q);
  }
}
W.add(new rt());
class it extends F {
  constructor() {
    super(...arguments);
    h(this, "zIndex", 1);
    h(this, "viewKey", Y.IMAGE);
  }
  check(r) {
    if (r === this.viewKey)
      return this;
  }
  convert(r) {
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      title: D,
      description: M,
      image: b = {}
    } = r, { imageProperties: P, placeholder: p, link: S } = b, I = (P == null ? void 0 : P.contentUrl) || "";
    return new Se(i, {
      url: I,
      top: s,
      left: a,
      width: o,
      height: c,
      zIndex: t,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      forceRender: !0
    });
  }
}
class nt {
  constructor() {
    h(this, "zIndex", 4);
  }
  create(e) {
    return e.createInstance(it);
  }
}
W.add(new nt());
var st = Object.defineProperty, at = Object.getOwnPropertyDescriptor, ct = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? at(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && st(e, r, t), t;
}, ot = (n, e) => (r, i) => e(r, i, n);
let ee = class extends F {
  constructor(e) {
    super();
    h(this, "zIndex", 2);
    h(this, "viewKey", Y.TEXT);
    this._localeService = e;
  }
  check(e) {
    if (e === this.viewKey)
      return this;
  }
  convert(e, r) {
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      title: D,
      description: M,
      richText: b = {}
    } = e, { text: P, ff: p, fs: S, it: I, bl: T, ul: f, st: x, ol: O, bg: X, bd: A, cl: L, rich: N } = b;
    let m = {
      top: s,
      left: a,
      width: o,
      height: c,
      zIndex: t,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      forceRender: !0
    }, R = !1;
    if (P != null ? (m = { ...m, text: P, ff: p, fs: S, it: I, bl: T, ul: f, st: x, ol: O, bg: X, bd: A, cl: L }, R = !0) : N != null && (m = { ...m, richText: N }, R = !0), R !== !1)
      return new Ce(this._localeService, i, m);
  }
};
ee = ct([
  ot(0, H(ne))
], ee);
class lt {
  constructor() {
    h(this, "zIndex", 0);
  }
  create(e) {
    return e.createInstance(ee);
  }
}
W.add(new lt());
class dt extends F {
  constructor() {
    super(...arguments);
    h(this, "zIndex", 2);
    h(this, "viewKey", Y.SHAPE);
  }
  check(r) {
    if (r === this.viewKey)
      return this;
  }
  convert(r) {
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      title: D,
      description: M
    } = r, { shapeType: b, text: P, shapeProperties: p, placeholder: S, link: I } = r.shape || {}, T = p == null ? "" : J(p.shapeBackgroundFill) || "rgba(255,255,255,1)", f = p == null ? void 0 : p.outline, x = {};
    if (f) {
      const { outlineFill: O, weight: X } = f;
      x.strokeWidth = X, x.stroke = J(O) || "rgba(0,0,0,1)";
    }
    if (b === he.Rect)
      return new B(i, {
        fill: T,
        top: s,
        left: a,
        width: o,
        height: c,
        zIndex: t,
        angle: w,
        scaleX: v,
        scaleY: u,
        skewX: l,
        skewY: d,
        flipX: _,
        flipY: g,
        forceRender: !0,
        ...x
      });
    if (b === he.RoundRect) {
      const O = (p == null ? void 0 : p.radius) || 0;
      return new B(i, {
        fill: T,
        top: s,
        left: a,
        width: o,
        height: c,
        zIndex: t,
        angle: w,
        scaleX: v,
        scaleY: u,
        skewX: l,
        skewY: d,
        flipX: _,
        flipY: g,
        forceRender: !0,
        radius: O,
        ...x
      });
    }
  }
}
class ht {
  constructor() {
    h(this, "zIndex", 2);
  }
  create(e) {
    return e.createInstance(dt);
  }
}
W.add(new ht());
var pt = Object.defineProperty, ut = Object.getOwnPropertyDescriptor, gt = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? ut(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && pt(e, r, t), t;
}, ft = (n, e) => (r, i) => e(r, i, n);
let te = class {
  constructor(n) {
    h(this, "_adaptors", []);
    this._injector = n, this._adaptorLoader();
  }
  convertToRenderObjects(n, e) {
    const r = Object.keys(n), i = [];
    return r.forEach((t) => {
      const a = n[t], s = this._executor(a, e);
      s != null && i.push(s);
    }), i;
  }
  convertToRenderObject(n, e) {
    return this._executor(n, e);
  }
  _executor(n, e) {
    var t;
    const { id: r, type: i } = n;
    for (const a of this._adaptors) {
      const s = (t = a.check(i)) == null ? void 0 : t.convert(n, e);
      if (s != null)
        return s;
    }
  }
  _adaptorLoader() {
    W.getData().sort(Ee).forEach((n) => {
      this._adaptors.push(n.create(this._injector));
    });
  }
};
te = gt([
  ft(0, H(q))
], te);
var _t = Object.defineProperty, wt = Object.getOwnPropertyDescriptor, vt = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? wt(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && _t(e, r, t), t;
}, St = (n, e) => (r, i) => e(r, i, n), It = /* @__PURE__ */ ((n) => (n.MAIN = "__SLIDERender__", n.SCENE_VIEWER = "__SLIDEViewer__", n.SCENE = "__SLIDEScene__", n.VIEWPORT = "__SLIDEViewPort_", n))(It || {});
let re = class extends F {
  constructor(e) {
    super();
    h(this, "zIndex", 6);
    h(this, "viewKey", Y.SLIDE);
    h(this, "_ObjectProvider", null);
    this._injector = e;
  }
  check(e) {
    if (e === this.viewKey)
      return this;
  }
  convert(e, r) {
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      title: D,
      description: M,
      slide: b
    } = e;
    if (b == null)
      return;
    const P = new Me(b), p = new $e("__SLIDERender__" + i, {
      top: s,
      left: a,
      width: o,
      height: c,
      zIndex: t,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      forceRender: !0
    });
    p.enableNav(), p.enableSelectedClipElement();
    const S = P.getPageOrder(), I = P.getPages();
    if (!S || !I)
      return p;
    this._ObjectProvider = new te(this._injector);
    for (let T = 0, f = S.length; T < f; T++) {
      const x = I[S[T]], { id: O } = x;
      p.addPageScene(this._createScene(O, p, x, r, P));
    }
    return p.activeFirstPage(), p;
  }
  _createScene(e, r, i, t, a) {
    var d;
    const { width: s, height: o } = r, c = new se(e, r, {
      width: s,
      height: o
    });
    new $(`PageViewer_${e}`, c, {
      left: 0,
      top: 0,
      bottom: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      explicitViewportHeightSet: !1
    }).closeClip();
    const { pageElements: v, pageBackgroundFill: u } = i, l = (d = this._ObjectProvider) == null ? void 0 : d.convertToRenderObjects(v, t);
    return this._addBackgroundRect(c, u, a), c.addObjects(l), l == null || l.forEach((_) => {
      c.attachTransformerTo(_);
    }), c;
  }
  _addBackgroundRect(e, r, i) {
    const t = i.getPageSize(), { width: a = 0, height: s = 0 } = t, o = new B("canvas", {
      left: 0,
      top: 0,
      width: a,
      height: s,
      strokeWidth: 1,
      stroke: "rgba(198,198,198, 1)",
      fill: J(r) || "rgba(255,255,255, 1)",
      zIndex: 0,
      evented: !1
    });
    e.addObject(o, 0);
  }
};
re = vt([
  St(0, H(q))
], re);
class Pt {
  constructor() {
    h(this, "zIndex", 6);
  }
  create(e) {
    return e.createInstance(re);
  }
}
W.add(new Pt());
var xt = Object.defineProperty, bt = Object.getOwnPropertyDescriptor, Tt = (n, e, r, i) => {
  for (var t = i > 1 ? void 0 : i ? bt(e, r) : e, a = n.length - 1, s; a >= 0; a--)
    (s = n[a]) && (t = (i ? s(e, r, t) : s(t)) || t);
  return i && t && xt(e, r, t), t;
}, K = (n, e) => (r, i) => e(r, i, n);
let ie = class extends F {
  constructor(e, r, i, t) {
    super();
    h(this, "zIndex", 4);
    h(this, "viewKey", Y.SPREADSHEET);
    this._localeService = e, this._contextService = r, this._configService = i, this._injector = t;
  }
  check(e) {
    if (e === this.viewKey)
      return this;
  }
  convert(e, r) {
    const {
      id: i,
      zIndex: t,
      left: a = 0,
      top: s = 0,
      width: o,
      height: c,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      spreadsheet: D
    } = e;
    if (D == null)
      return;
    const { worksheet: M, styles: b } = D, P = new Xe(b), p = new Ye(
      new ze(i, M, P),
      // FIXME: worksheet in slide doesn't has a Worksheet object
      P,
      this._localeService,
      this._contextService,
      this._configService,
      this._injector
    ), { rowTotalHeight: S, columnTotalWidth: I, rowHeaderWidth: T, columnHeaderHeight: f } = p, x = I + M.rowHeader.width || 0, O = S + M.columnHeader.height || 0, X = new we("spreadInSlideSceneViewer" + i, {
      top: s,
      left: a,
      width: o,
      height: c,
      zIndex: t,
      angle: w,
      scaleX: v,
      scaleY: u,
      skewX: l,
      skewY: d,
      flipX: _,
      flipY: g,
      forceRender: !0
    }), A = new se("spreadInSlideScene" + i, X, {
      width: x,
      height: O
    });
    this._updateViewport(i, T, f, A, r);
    const L = new Fe("testSheetViewer", p, !1), N = new Ke("spreadInSlideRow", p), m = new ke("spreadInSlideColumn", p), R = new B("spreadInSlideLeftTop", {
      zIndex: 2,
      left: -1,
      top: -1,
      width: T,
      height: f,
      fill: ue([248, 249, 250]),
      stroke: ue([217, 217, 217]),
      strokeWidth: 1
    });
    return L.zIndex = 10, A.addObjects([L], 1), A.addObjects([N, m, R], 2), X;
  }
  // eslint-disable-next-line max-lines-per-function
  _updateViewport(e, r, i, t, a) {
    if (a == null)
      return;
    const s = r * t.scaleX, o = i * t.scaleY, c = new $("spreadInSlideViewMain" + e, t, {
      left: s,
      top: o,
      bottom: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      explicitViewportHeightSet: !1,
      isWheelPreventDefaultX: !0
    }), w = new $("spreadInSlideViewTop" + e, t, {
      left: s,
      height: o,
      top: 0,
      right: 0,
      explicitViewportWidthSet: !1,
      isWheelPreventDefaultX: !0
    }), v = new $("spreadInSlideViewLeft" + e, t, {
      left: 0,
      bottom: 0,
      top: o,
      width: s,
      explicitViewportHeightSet: !1,
      isWheelPreventDefaultX: !0
    });
    new $("spreadInSlideViewLeftTop" + e, t, {
      left: 0,
      top: 0,
      width: s,
      height: o,
      isWheelPreventDefaultX: !0
    }), c.onScrollAfter$.subscribeEvent((u) => {
      const { scrollX: l, scrollY: d, viewportScrollX: _, viewportScrollY: g } = u;
      w.updateScrollVal({
        scrollX: l,
        viewportScrollX: _
      }), v.updateScrollVal({
        scrollY: d,
        viewportScrollY: g
      });
    }), t.attachControl(), new ve(c, {
      mainScene: a
    }), t.onMouseWheel$.subscribeEvent((u, l) => {
      const d = u;
      if (d.ctrlKey) {
        const _ = Math.abs(d.deltaX);
        let g = _ < 40 ? 0.05 : _ < 80 ? 0.02 : 0.01;
        g *= d.deltaY > 0 ? -1 : 1, t.scaleX < 1 && (g /= 2), t.scaleX + g > 4 ? t.scale(4, 4) : t.scaleX + g < 0.1 ? t.scale(0.1, 0.1) : (t.scaleBy(g, g), d.preventDefault());
      } else
        c.onMouseWheel(d, l);
    });
  }
};
ie = Tt([
  K(0, H(ne)),
  K(1, Ve),
  K(2, _e),
  K(3, H(q))
], ie);
class Ot {
  constructor() {
    h(this, "zIndex", 4);
  }
  create(e) {
    return e.createInstance(ie);
  }
}
W.add(new Ot());
var yt = /* @__PURE__ */ ((n) => (n.COMPONENT = "__slideRender__", n.SCENE = "__mainScene__", n.VIEW = "__mainView__", n))(yt || {});
export {
  W as CanvasObjectProviderRegistry,
  F as ObjectAdaptor,
  te as ObjectProvider,
  yt as SLIDE_KEY,
  It as SLIDE_VIEW_KEY,
  fe as UniverSlidesPlugin
};
