// @univerjs/rpc-node/lib/es/index.js
var j = Object.defineProperty;
var O = (r, e, n) => e in r ? j(r, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : r[e] = n;
var i = (r, e, n) => O(r, typeof e != "symbol" ? e + "" : e, n);
import { fork as E } from "node:child_process";
import a from "node:process";
import { Inject as P, Injector as h, Plugin as u, IConfigService as w, ILogService as y } from "@univerjs/core";
import { DataSyncPrimaryController as f, DataSyncReplicaController as p, IRPCChannelService as v, ChannelService as I, IRemoteInstanceService as M, WebWorkerRemoteInstanceService as U, IRemoteSyncService as D, RemoteSyncPrimaryService as b } from "@univerjs/rpc";
import { Observable as N, shareReplay as S } from "rxjs";
var x = Object.defineProperty, W = Object.getOwnPropertyDescriptor, C = (r, e, n, o) => {
  for (var s = o > 1 ? void 0 : o ? W(e, n) : e, t = r.length - 1, c; t >= 0; t--)
    (c = r[t]) && (s = (o ? c(e, n, s) : c(s)) || s);
  return o && s && x(e, n, s), s;
}, l = (r, e) => (n, o) => e(n, o, r);
const G = "node-rpc.main.config";
var _;
let m = (_ = class extends u {
  constructor(r, e, n) {
    super(), this._config = r, this._injector = e, this._configService = n, this._configService.setConfig(G, this._config);
  }
  onStarting() {
    const { workerSrc: r } = this._config, e = L(this._injector, r);
    [
      [v, {
        useFactory: () => new I(e)
      }],
      [f],
      [D, { useClass: b }]
    ].forEach((o) => this._injector.add(o)), this._injector.get(f);
  }
}, i(_, "pluginName", "UNIVER_RPC_NODE_MAIN_PLUGIN"), _);
m = C([
  l(1, P(h)),
  l(2, w)
], m);
var g;
let d = (g = class extends u {
  constructor(r, e) {
    super(), this._config = r, this._injector = e;
  }
  onStarting() {
    [
      [p],
      [v, {
        useFactory: () => new I(F())
      }],
      [M, { useClass: U }]
    ].forEach((r) => this._injector.add(r)), this._injector.get(p);
  }
}, i(g, "pluginName", "UNIVER_RPC_NODE_WORKER_PLUGIN"), g);
d = C([
  l(1, P(h))
], d);
function L(r, e) {
  const n = r.get(y), o = E(e);
  return o.on("spawn", () => n.log("Child computing process spawned!")), o.on("error", (t) => n.error(t)), {
    send(t) {
      o.send(t);
    },
    onMessage: new N((t) => {
      const c = (R) => {
        t.next(R);
      };
      return o.on("message", c), () => o.off("message", c);
    }).pipe(S(1))
  };
}
function F() {
  return {
    send(r) {
      a.send(r);
    },
    onMessage: new N((r) => {
      const e = (n) => {
        r.next(n);
      };
      return a.on("message", e), () => a.off("message", e);
    }).pipe(S(1))
  };
}
export {
  m as UniverRPCNodeMainPlugin,
  d as UniverRPCNodeWorkerPlugin
};
