// @univerjs/preset-sheets-advanced/lib/es/index.js
import { UniverProFormulaEnginePlugin as i } from "@univerjs-pro/engine-formula";
export * from "@univerjs-pro/engine-formula";
import { UniverExchangeClientPlugin as l } from "@univerjs-pro/exchange-client";
export * from "@univerjs-pro/exchange-client";
import { UniverLicensePlugin as u } from "@univerjs-pro/license";
export * from "@univerjs-pro/license";
import { UniverSheetsChartPlugin as f } from "@univerjs-pro/sheets-chart";
export * from "@univerjs-pro/sheets-chart";
import { UniverSheetsChartUIPlugin as a } from "@univerjs-pro/sheets-chart-ui";
export * from "@univerjs-pro/sheets-chart-ui";
import { UniverSheetsExchangeClientPlugin as s } from "@univerjs-pro/sheets-exchange-client";
export * from "@univerjs-pro/sheets-exchange-client";
import { UniverSheetsPivotTablePlugin as t } from "@univerjs-pro/sheets-pivot";
export * from "@univerjs-pro/sheets-pivot";
import { UniverSheetsPivotTableUIPlugin as v } from "@univerjs-pro/sheets-pivot-ui";
export * from "@univerjs-pro/sheets-pivot-ui";
import { UniverSheetsPrintPlugin as x } from "@univerjs-pro/sheets-print";
export * from "@univerjs-pro/sheets-print";
import { UniverSheetSparklinePlugin as U } from "@univerjs-pro/sheets-sparkline";
export * from "@univerjs-pro/sheets-sparkline";
import { UniverSheetSparklineUIPlugin as g } from "@univerjs-pro/sheets-sparkline-ui";
export * from "@univerjs-pro/sheets-sparkline-ui";
import { UniverNetworkPlugin as c } from "@univerjs/network";
export * from "@univerjs/network";
import "@univerjs/network/facade";
import "@univerjs-pro/exchange-client/facade";
import "@univerjs-pro/sheets-pivot/facade";
import "@univerjs-pro/engine-formula/facade";
import "@univerjs-pro/sheets-print/facade";
import "@univerjs-pro/sheets-chart-ui/facade";
import "@univerjs-pro/sheets-sparkline/facade";
export * from "@univerjs/sheets-graphics";
function j(n = {
  license: "",
  universerEndpoint: ""
}) {
  const {
    license: p,
    universerEndpoint: e,
    useWorker: o
  } = n, r = e != null ? e : `${window.location.protocol}//${window.location.host}`;
  return {
    plugins: [
      c,
      [u, { license: p }],
      // TODO: @wzhudev: if we use worker, we need to add different configurations to SheetsPivotTable
      o ? [t, { notExecuteFormula: !0 }] : [t],
      v,
      o ? [i, { notExecuteFormula: !0 }] : i,
      x,
      f,
      a,
      U,
      g,
      [l, {
        uploadFileServerUrl: `${r}/universer-api/stream/file/upload`,
        getTaskServerUrl: `${r}/universer-api/exchange/task/{taskID}`,
        signUrlServerUrl: `${r}/universer-api/file/{fileID}/sign-url`,
        importServerUrl: `${r}/universer-api/exchange/{type}/import`,
        exportServerUrl: `${r}/universer-api/exchange/{type}/export`,
        downloadEndpointUrl: `${r}/`
      }],
      s
    ].filter((m) => !!m)
  };
}
export {
  j as UniverSheetsAdvancedPreset
};
